---
title: "Family Code on Homeo Domain SELEX data"
output: html_notebook
---
### Load motifs
```{r}
library(FamilyCode)
workDir <- 'SELEX_HD/'
motifPoses <- c('N1','N2','T3','D4','A5','Y6','N7','N8')
#loadMotifs
proBound_DIR <- paste0(workDir, 'consensus_fits/')
template <- paste0(workDir, 'consensus_fits/$modelFile$.consensus.json')
gene_symbol_factor_ids <- read.csv(paste0(workDir, 'gene_symbol_factor_ids.csv'), row.names=1)
files <- list.files(proBound_DIR, pattern = 'consensus')
files <- unlist(lapply(strsplit(files, '\\.'), function(x) x[[1]][1]))
gene_symbol <- unlist(lapply(strsplit(files, '_'), function(x) x[[1]][1]))
study <- unlist(lapply(strsplit(files, '_'), function(x) x[[2]][1]))
info <- data.frame(gene_symbol, study)
info <- info[as.numeric(info$gene_symbol) < 10000,]

HD_all_motifs <- loadFromIndex(info, template)
motifPositions <- motifPoses
HDseed <- matrix(nrow = 4,ncol = 8,data = c(0,0,0,0,
                                            0,0,0,0,
                                            0,0,0,1,
                                            1,0,1,1,
                                            1,0,0,0,
                                            0,1,0,1,
                                            0,0,0,0,
                                            0,0,0,0))
rownames(HDseed) <- DNA()

HD_motifScores <- scoreMotifList(HD_all_motifs, HDseed, weight = c(0,0,1,1,1,1,0,0))

HDids<- unique(HD_motifScores$gene_symbol)
HDfiltered_table <- data.frame(NULL)
for(i in 1:length(HDids)){
  fDT <- HD_motifScores[HD_motifScores$gene_symbol == HDids[i],]
  HDfiltered_table <- rbind.data.frame(HDfiltered_table, fDT[which.min(fDT$score),])
}
plot.ecdf(as.numeric(HDfiltered_table$score))
HDfiltered_table <- HDfiltered_table[HDfiltered_table$score < 1,]
```

### Load Alignment
```{r}
#prepare alignment
HD_ali <- read.table(paste0(workDir,"HD_hmm.sto"), quote="\"", fill = T)
HD_ali$V1 <- apply(HD_ali, 1, function(x) strsplit(x[1],'/')[[1]][1])
colnames(HD_ali) <- c('name','alignment')

#change names
geneName <- apply(HDfiltered_table, 1, function(x) gene_symbol_factor_ids$gene_symbol[gene_symbol_factor_ids$factor_ids == x[1]])
HDfiltered_table$gene_symbol <- geneName
HD_motifs <- filterMotifList(HDfiltered_table, HD_all_motifs, 8,motifPositions)


#quality filter
HD_alignment <- matchAliMotif(HD_motifs, HD_ali)
del <- which(substr(HD_alignment$alignment, 50,50) != 'N')
HD_alignment <- HD_alignment[-del,]
HD_motifs <- HD_motifs[-del]


del <- which(duplicated(HD_alignment$name))
HD_alignment <- HD_alignment[-del,]
HD_motifs <- HD_motifs[-del]
```

### Make p-val table
```{r}
#genreate p-val tablle
pvalTable <- getPvalTable(HD_motifs, Alignment = HD_alignment, pos_index = motifPositions)
#apply -log(10)
pvalTable <- -log(pvalTable,10)
#set NAs to 0
pvalTable[is.na(pvalTable)] <- 0
#plot heatmap
gplots::heatmap.2(as.matrix(as.data.frame(lapply(pvalTable, as.numeric))),dendrogram='none',
                  Rowv=FALSE, Colv=FALSE,trace='none',col = rev(heat.colors(12)), key.title = '-logP-val')

AAPosition <- 49
motifPosition <- 'N7'
HD_alignment <- matchAliMotif(HD_motifs, HD_alignment)
AAbp_Matrix <- AAbpCombination(HD_motifs, HD_alignment, AAPosition, motifPosition)
plot_tetrahedron(AAbp_Matrix, color = T, vertex = F, size = 10)
```

### Prepare Base line
```{r}
#get base line data set
HDidstds <- unique(HD_motifScores[,1:2])
HDfiltered_table_baseline <- data.frame(NULL)
for(i in 1:nrow(HDidstds)){
  fDT <- HD_motifScores[HD_motifScores$gene_symbol == HDidstds[i,1],]
  fDT <- fDT[fDT$study == HDidstds[i,2],]
  HDfiltered_table_baseline <- rbind.data.frame(HDfiltered_table_baseline, fDT[which.min(fDT$score),])
}
plot.ecdf(as.numeric(HDfiltered_table_baseline$score))
```

```{r}
HDfiltered_table_baseline <- HDfiltered_table_baseline[HDfiltered_table_baseline$score < 1,]
#Vsx1 have two gene_symbol
HDfiltered_table_baseline[HDfiltered_table_baseline$gene_symbol == 2708,]$gene_symbol <- 386
HD_motifs_baseline <- filterMotifList(HDfiltered_table_baseline, HD_all_motifs, 8, motifPositions)

```

### Show Base Line 
```{r}
motifPoses <- motifPositions
evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
for(pos in motifPoses){
  ScreenPoses <- c(pos,pos)
  BaseLines.ddG <- getBaseLineAccuracy(HD_motifs_baseline, pos = ScreenPoses, randomSample = F)
  colnames(BaseLines.ddG) <- c('pred','true')
  BaseLines.PFM <- predTrue.ddG2frequency(BaseLines.ddG, PFM = T)
  plotPredTrue(BaseLines.ddG, xlab = 'Experiment 1', ylab = 'Experiment 2', main = paste0('Pair-wise experiment comparison at ',pos))
  boxplot(groupedR2(BaseLines.ddG$pred, BaseLines.ddG$true, member = length(ScreenPoses)*4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(BaseLines.ddG$pred, BaseLines.ddG$true, 
                                                          member = length(ScreenPoses)*4, mean= T),4)))
  plotPredTrue(BaseLines.PFM, xlab = 'Experiment 1', ylab = 'Experiment 2', main = paste0('Base Line PFM at ',pos))
  hit <- sum(round(groupedR2(BaseLines.PFM$pred, BaseLines.PFM$true, member = 4, mean= F)^(1/2),4) > 0.5)/(nrow(BaseLines.PFM)/4)
  boxplot(groupedR2(BaseLines.PFM$pred, BaseLines.PFM$true, member = length(ScreenPoses)*4, mean= F)^(1/2), ylim = c(0,1),
        main = paste0(pos, 'Mean PCC = ', round(groupedR2(BaseLines.PFM$pred, BaseLines.PFM$true, 
                                                     member = length(ScreenPoses)*4, mean= T)^(1/2),4), ' Accuracy = ', round(hit,4)))
  
  evaMatrix[,pos] <- c(round(groupedR2(BaseLines.ddG$pred, 
        BaseLines.ddG$true, nrow(BaseLines.ddG), throughZero = T), 
        4), round(RMSD(BaseLines.ddG$pred, BaseLines.ddG$true), 
        4), round(groupedR2(BaseLines.PFM$pred, 
        BaseLines.PFM$true, nrow(BaseLines.PFM), throughZero = T), 
        4)^(1/2), round(hit,4))
}
evaMatrix_BL <- evaMatrix
```

### leave-one-out test for all available samples

Perform leave-one-out testing for accuracy with all available training samples.

```{r}
#make predictions
CV_alignment <- HD_alignment 
CV_motif <- HD_motifs
FCpredTrueList <- list()
for(pos in motifPoses){
  predTrue <- SVDregression.Iterative.CV(CV_motif,CV_alignment, pos = pos,Ftest_pVal = 0.001, useSimilarAA = T)
  #delete unsuccessful predictions
  predTrue <- predTrue[!is.na(predTrue$pred),]
  FCpredTrueList[[pos]] <- predTrue
  print(pos)
  
}

#plot results
evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
for(pos in motifPoses){
  predTrue <- FCpredTrueList[[pos]]
  #delete unsuccessful predictions
  #predTrue <- predTrue[!is.na(predTrue$pred),]
  attr(predTrue, 'confidence') <- attr(predTrue, 'confidence')[attr(predTrue, 'confidence') <= 1]
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('FamilyCode prediction at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= T),4)))
  #plot confidance
  plot(attr(predTrue, 'confidence'),groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F), ylab = 'Predict-True R^2', xlab = 'Confidence score', main = paste0('Confidence at ', pos))
  legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
                                                     attr(predTrue, 'confidence'), ),4)), bty = 'n')
  dt <- data.frame(conf = attr(predTrue, 'confidence'), R2 = groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F))
  lm <- lm(R2~conf, data= dt)
  abline(lm, col= 'red')
  #plot frequency

  plotPredTrue(predTrue.PFM, main = paste0('FamilyCode prediction PFM at ',pos))
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F)^(1/2), ylim = c(0,1),main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= T)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = T), 
        4)^(1/2), round(hit,4))
}
evaMatrix_FC <- evaMatrix
```

### Show Nearest neighbor prediction 
```{r}
CSpredTrueList <- list()
for(pos in motifPoses){
  predTrue <- closestSeqPred(CV_motif,CV_alignment, pos = pos)
  #delete unsuccessful predictions
  predTrue <- predTrue[!is.na(predTrue$pred),]
  CSpredTrueList[[pos]] <- predTrue
  print(pos)
  
}
evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
for(pos in motifPoses){
  predTrue <- CSpredTrueList[[pos]]
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  
  plotPredTrue(predTrue, main = paste0('Closest sequence prediction at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  
  plotPredTrue(predTrue.PFM, main = paste0('Closest sequence prediction PFM at ',pos))
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F)^(1/2), ylim = c(0,1),
        main = paste0(pos, 'Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
      predTrue$true, nrow(predTrue), throughZero = T), 
      4), round(RMSD(predTrue$pred, predTrue$true), 
      4), round(groupedR2(predTrue.PFM$pred, 
      predTrue.PFM$true, nrow(predTrue.PFM), throughZero = T), 
      4)^(1/2), round(hit,4))
}
evaMatrix_CS <- evaMatrix

```

### Comparing prediction result with Similarity regression
from paper:
Lambert, S. A., Yang, A. W., Sasse, A., Cowley, G., Albu, M., Caddick, M. X., ... & Hughes, T. R. (2019). Similarity regression predicts evolution of transcription factor sequence specificity. Nature genetics, 51(6), 981-989.
Figure S6/S7

```{r}
options(warn=0)
SRpredTrueList <- list()
for(pos in motifPoses){
  predTrue <- SRpred(CV_motif,CV_alignment, pos = pos, weightfile = 'cisBP_HD/F223_1.97d.json')
  #delete unsuccessful predictions
  predTrue <- predTrue[!is.na(predTrue$pred),]
  SRpredTrueList[[pos]] <- predTrue
  print(pos)
}
evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
for(pos in motifPoses){
  predTrue <- SRpredTrueList[[pos]]
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  
  plot.ecdf(predTrue$similarity)
  plotPredTrue(predTrue, main = paste0('Similarity regression at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  
  plotPredTrue(predTrue.PFM, main = paste0('Similarity regression prediction PFM at ',pos))
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F)^(1/2), ylim = c(0,1),
        main = paste0(pos, 'Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
    predTrue$true, nrow(predTrue), throughZero = T), 
    4), round(RMSD(predTrue$pred, predTrue$true), 
    4), round(groupedR2(predTrue.PFM$pred, 
    predTrue.PFM$true, nrow(predTrue.PFM), throughZero = T), 
    4)^(1/2), round(hit,4))
}
evaMatrix_SR <- evaMatrix
```

```{r}
evaMatrix <- rbind.data.frame(evaMatrix_BL, evaMatrix_FC,  evaMatrix_SR, evaMatrix_CS)
evaMatrix <- evaMatrix[c((1:4)*4-3, (1:4)*4-2, (1:4)*4-1,(1:4)*4), ]
write.csv(evaMatrix, 'evaMatrix_HD_SELEX.csv')
```
### systemetic test of prediction methods with subsampling
```{r}
options(warn=0)
CV_alignment <- HD_alignment
CV_motif <- HD_motifs
nset <- nrow(CV_alignment)
results <- list()
for(i in 2:10){
  FCR2List <- c()
  FCrmsdList <- c()
  CSR2List <- c()
  CSrmsdList <- c()
  SRR2List <- c()
  SRrmsdList <- c()
  for(j in 1:40){
    trainID <- sample(1:nset, floor(nset-nset/(i)))
    train_alignment <- CV_alignment[trainID,]
    train_motifs <- CV_motif[trainID]
    test_alignment <- CV_alignment[-trainID,]
    test_motifs <- CV_motif[-trainID]
    #FC
    svdModel <- makeSVDModel(train_motifs, train_alignment, Positions = c('N1','N2','T3','D4','A5','Y6','N7','N8'))
    pred <- lapply(test_alignment$alignment, function(x) SVDmodelPredPSAM(svdModel, x))
    pred <- lapply(pred, frequency2ddG)
    true <- lapply(test_motifs, function(x) x$matrix)
    true <- lapply(true, frequency2ddG)
    predTrue <- data.frame(pred = as.numeric(unlist(pred)), true = as.numeric(unlist(true)))
    FCR2 <- round(groupedR2(predTrue$true, predTrue$pred),4)
    FCrmsd <- round(RMSD(predTrue$true,predTrue$pred),4)
    #CS
    pred_motifs <- CSpredPSAM(test_alignment, train_motifs, train_alignment)
    pred <- lapply(pred_motifs, function(x) x$matrix)
    pred <- lapply(pred, frequency2ddG)
    true <- lapply(test_motifs, function(x) x$matrix)
    true <- lapply(true, frequency2ddG)
    predTrue <- data.frame(pred = as.numeric(unlist(pred)), true = as.numeric(unlist(true)))
    CSR2 <- round(groupedR2(predTrue$true, predTrue$pred),4)
    CSrmsd <- round(RMSD(predTrue$true,predTrue$pred),4)
    #SR
    pred_motifs <- SRpredPSAM(test_alignment, train_motifs, train_alignment, weightfile = 'cisBP_HD/F223_1.97d.json')
    pred <- lapply(pred_motifs, function(x) x$matrix)
    pred <- lapply(pred, frequency2ddG)
    true <- lapply(test_motifs, function(x) x$matrix)
    true <- lapply(true, frequency2ddG)
    predTrue <- data.frame(pred = as.numeric(unlist(pred)), true = as.numeric(unlist(true)))
    SRR2 <- round(groupedR2(predTrue$true, predTrue$pred),4)
    SRrmsd <- round(RMSD(predTrue$true,predTrue$pred),4)
    FCR2List <- c(FCR2List, FCR2)
    FCrmsdList <- c(FCrmsdList, FCrmsd)
    CSR2List <- c(CSR2List, CSR2)
    CSrmsdList <- c(CSrmsdList, CSrmsd)
    SRR2List <- c(SRR2List, SRR2)
    SRrmsdList <- c(SRrmsdList, SRrmsd)
  }
  add <- list()
  add$FCR2 <- FCR2List
  add$FCrmsd <- FCrmsdList
  add$CSR2 <- CSR2List
  add$CSrmsd <- CSrmsdList
  add$SRR2 <- SRR2List
  add$SRrmsd <- SRrmsdList
  results[[length(results)+1]] <- add
}
HD_SELEX_results <- results
FCR2 <- unlist(lapply(results, function(x) mean(x$FCR2)))
FCrmsd <- unlist(lapply(results, function(x) mean(x$FCrmsd)))
CSR2 <- unlist(lapply(results, function(x) mean(x$CSR2)))
CSrmsd <- unlist(lapply(results, function(x) mean(x$CSrmsd)))
SRR2 <- unlist(lapply(results, function(x) mean(x$SRR2)))
SRrmsd <- unlist(lapply(results, function(x) mean(x$SRrmsd)))

matplot(1-1/2:10, cbind(FCR2, CSR2, SRR2), type = "l", lty = 1, 
        col = c("red", "blue", "green"), xlab = "X", 
        ylab = "Y", main = "Multiple Lines Plot")
matplot(1-1/2:10, cbind(FCrmsd, CSrmsd, SRrmsd), type = "l", lty = 1, 
        col = c("red", "blue", "green"), xlab = "X", 
        ylab = "Y", main = "Multiple Lines Plot")
```

### inspecting Ubx
```{r}
svdModel <- makeSVDModel(CV_motif, CV_alignment, Positions = c('N1','N2','T3','D4','A5','Y6','N7','N8'))
svdModel$N7$keyPos

```

### predict for BARHL2
```{r}
BARHL2id <- which(CV_alignment$name == 'BARHL2')
train_alignment <- CV_alignment#[-BARHL2id,]
train_motif <- CV_motif#[-BARHL2id]

svdModel <- makeSVDModel(train_motif, train_alignment, Positions = motifPoses)

mutants <- c('TT','NR_6C','TA_6T','IN_6T')
BARHL2seq <- CV_alignment$alignment[BARHL2id]
BARHL2seqs <- rep(BARHL2seq,length(mutants))
substr(BARHL2seqs, 46,46) <- substr(mutants,1,1)
substr(BARHL2seqs, 53,53) <- substr(mutants,2,2)
test_alignment <- data.frame(name = paste0('BARHL2_', mutants), alignment = BARHL2seqs)

SVDmodelPredPSAM(svdModel, BARHL2seqs[1])
BARHL2.pred <- lapply(test_alignment$alignment, function(x) SVDmodelPredPSAM(svdModel, x))

mononucleotide_logo(frequency2ddG(CV_motif[[BARHL2id]]$matrix))
for(i in 2:length(BARHL2.pred)){
  plot <- mononucleotide_logo((frequency2ddG(BARHL2.pred[[i]])))
  plot <- plot + ggplot2::ggtitle(paste0(test_alignment$name[i]))
  plot(plot)
}
```
