---
title: "Family Code on bHLH SELEX data"
output: html_notebook
---

```{r}
#devtools::install_github('ShaoxunLiu/FCpackage')
library(FamilyCode)
```
### Load motifs
```{r}
library(FamilyCode)
workDir <- 'intermediateData/bHLH/SELEX_bHLH/'
rawData <- 'rawData/bHLH/'
motifPoses <- c('P-3','P-2','P-1','P1','P2','P3')
```
### Input data sets

1: Probound_runDir: Directory containing the ProBound results each in a folder.
2: bHLH_info: CSV table containing experiment information. 
3: bHLH_ali: Alignment file

```{r}
proBound_runDIR <- paste0(rawData, 'HTdata/')
bHLH_info <- read.csv(file = paste0(rawData, "bHLH_info.csv"), row.names = 1)
```

### Preparing variables for subsequent processes
```{r}
proBound_files <- list.files(proBound_runDIR)
modelFile_Template <- paste0(proBound_runDIR, '$modelFile$/result/fit.models.consensus.json')
rows <- c("A", "C", "G", "T")
blank <- data.frame(P00 = c(1,1,1,1), P0 = c(1,1,1,1))

#getting expeirment identifier from bHLH_info, may vary according to different naming scheme
bHLH_index <- bHLH_info[,1:2]
```

### Entering quality measurements for each model

This step can be ignored if no quality measurements are performed, or when each sample protein is only appearing once.

```{r}
bHLH_index$quality <- 0
for(i in 1:nrow(bHLH_index)){
  tryCatch({
    dir <- paste0(proBound_runDIR,bHLH_index$gene_symbol[i],'_',bHLH_index$study[i])
    suppressWarnings(quality <- readChar(paste0(dir,'//fitEval.txt'),nchars = 100))
    quality <- strsplit(quality, ', ')[[1]][1]
    quality <- strsplit(quality, ' = ')[[1]][2]
    bHLH_index$quality[i] <- quality
  }, error = function(e){})
}
```

### Reading motif model data from ProBound results

Reads in all data of bHLH samples and compares the consensus sequence recognition score between the binding models output from ProBound. For bHLH, a logical value is given for symmetry. The motif with the highest consensus sequence recognition score is saved in the list of mono_motifs. The mono_motifs contains binding motifs in frequency format.

```{r}
Motif_Table <- loadMono_motifs(bHLH_index, modelFile_Template, rec_seq = 'CANNTG', pos_index = motifPoses, checkSymmetry = T, withTable = T)

all_mono_motifs <- Motif_Table$motifs
bHLH_model_info <- Motif_Table$table
```

### Loading the Alignment

Input Alignment file generated with Clustal Omega and select the aligned DNA-binding Domain of the family.

```{r}
bHLH_pbAlignment <- read.table(paste0(rawData, 'bHLH.sto'), quote="\"", fill = T)
name <- apply(bHLH_pbAlignment,1,function(x) strsplit(x[1],'/')[[1]][1])
colnames(bHLH_pbAlignment) <- c('name','alignment')
bHLH_pbAlignment$name <- name
```

For a different sample set, likely a different family, You can start from this point with properly structured list of mono_motifs and data frame of Alignment

### Filtering and reloading for saticfactory data samples

Family code uses the bHLH_index data frame to load models. Filter rows according to recorded parameters and reload mono_motifs. Here, the data frame: bHLH_model_info has contains parameters and the first two rows are in the same format as the bHLH_index data frame.

```{r}
#filter for symmetry
bHLH_sym <- bHLH_model_info[bHLH_model_info$symmetry == 1,]
#filter for best model of each protein
bHLH_names <- unique(bHLH_sym$gene_symbol)
bHLH_index <- data.frame(NULL)
for(i in 1:length(bHLH_names)){
  qualities <- bHLH_sym[bHLH_sym$gene_symbol == bHLH_names[i],]
  add <- qualities[qualities$quality == max(qualities$quality),]
  bHLH_index <- rbind.data.frame(bHLH_index, add[1,])
}
#filter for quality > 0.15
bHLH_index <- bHLH_index[bHLH_index$quality > 0.15,]


#reload mono_motifs
bHLH_motifs <- loadMono_motifs(bHLH_index, modelFile_Template, rec_seq = 'CANNTG', pos_index = motifPoses, checkSymmetry = F, withTable = F)
#uniform variable name
CV_motif <- bHLH_motifs
CV_alignment <- matchAliMotif(CV_motif, bHLH_pbAlignment)
saveRDS(CV_alignment, file = paste0('intermediateData/bHLH/bHLH_SELEX_CV_alignment.rsd'))
saveRDS(CV_motif, file =  paste0('intermediateData/bHLH/bHLH_SELEX_CV_motif.rsd'))
```

### Make p-val table
```{r}
#genreate p-val tablle
pvalTable <- getPvalTable(CV_motif, Alignment = CV_alignment, pos_index = motifPoses)
#apply -log(10)
pvalTable <- -log(pvalTable,10)
#set NAs to 0
pvalTable[is.na(pvalTable)] <- 0
#plot heatmap
gplots::heatmap.2(as.matrix(as.data.frame(lapply(pvalTable, as.numeric))),dendrogram='none',
                  Rowv=FALSE, Colv=FALSE,trace='none',col = rev(heat.colors(12)), key.title = '-logP-val')
```


### Show Base Line 
```{r}
#550/600
evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
R2s_BL <- list()
for(pos in motifPoses){
  BaseLines.ddG <- getBaseLineAccuracy(all_mono_motifs, pos = pos, randomSample = F)
  colnames(BaseLines.ddG) <- c('pred','true')
  BaseLines.PFM <- predTrue.ddG2frequency(BaseLines.ddG, PFM = T)
  plotPredTrue(BaseLines.ddG, xlab = 'Experiment 1 -ΔΔG/RT', ylab = 'Experiment 2 -ΔΔG/RT', main = paste0('Pair-wise experiment comparison at ',pos))
  boxplot(groupedR2(BaseLines.ddG$pred, BaseLines.ddG$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(BaseLines.ddG$pred, BaseLines.ddG$true, 
                                                          member = 4, mean= T),4)))
  plotPredTrue(BaseLines.PFM, xlab = 'Experiment 1', ylab = 'Experiment 2', main = paste0('Base Line PFM at ',pos))
  hit <- sum(round(groupedR2(BaseLines.PFM$pred, BaseLines.PFM$true, member = 4, mean= F, throughZero = F)^(1/2),4) > 0.5)/(nrow(BaseLines.PFM)/4)
  boxplot(groupedR2(BaseLines.PFM$pred, BaseLines.PFM$true, member = 4, mean= F, throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, 'Mean PCC = ', round(groupedR2(BaseLines.PFM$pred, BaseLines.PFM$true, 
                                                     member = 4, mean= T, throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(BaseLines.ddG$pred, 
        BaseLines.ddG$true, nrow(BaseLines.ddG), throughZero = T), 
        4), round(RMSD(BaseLines.ddG$pred, BaseLines.ddG$true), 
        4), round(groupedR2(BaseLines.PFM$pred, 
        BaseLines.PFM$true, nrow(BaseLines.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  R2s <- groupedR2(BaseLines.ddG$pred, BaseLines.ddG$true, member = 4, mean= F)
  R2s_BL[[pos]] <- R2s
}
evaMatrix_BL <- evaMatrix

```



### leave-one-out test for all available samples
Perform leave-one-out testing for accuracy with all available training samples.

```{r}
#make predictions
FCpredTrueList <- list()

for(pos in motifPoses){
  predTrue <- SVDregression.Iterative.CV(CV_motif,CV_alignment, pos = pos,Ftest_pVal = 0.001, useSimilarAA = T)
  #delete unsuccessful predictions
  predTrue <- predTrue[!is.na(predTrue$pred),]
  FCpredTrueList[[pos]] <- predTrue
  print(pos)
  
}

#plot results
#mkdirs(paste0(workDir, 'plots'))
evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
for(pos in motifPoses){
  predTrue <- FCpredTrueList[[pos]]
  #delete unsuccessful predictions
  #predTrue <- predTrue[!is.na(predTrue$pred),]
  attr(predTrue, 'confidence') <- attr(predTrue, 'confidence')[attr(predTrue, 'confidence') <= 1]
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('FamilyCode at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= T),4)))
  #plot confidance
  plot(attr(predTrue, 'confidence'),groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F), ylab = 'Predict-True R^2', xlab = 'Confidence score', main = paste0('Confidence at ', pos))
  legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
                                                     attr(predTrue, 'confidence'), ),4)), bty = 'n')
  dt <- data.frame(conf = attr(predTrue, 'confidence'), R2 = groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F))
  lm <- lm(R2~conf, data= dt)
  abline(lm, col= 'red')
  #plot frequency

  plotPredTrue(predTrue.PFM, main = paste0('FamilyCode prediction PFM at ',pos))
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F, throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F, throughZero = F)^(1/2), ylim = c(0,1),main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= T, throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
}
evaMatrix_FC <- evaMatrix
```

### Compair with closest paralog prediction 
from paper:
Weirauch, M. T., Yang, A., Albu, M., Cote, A. G., Montenegro-Montero, A., Drewe, P., ... & Hughes, T. R. (2014). Determination and inference of eukaryotic transcription factor sequence specificity. Cell, 158(6), 1431-1443.
```{r}
CSpredTrueList <- list()
for(pos in motifPoses){
  predTrue <- closestSeqPred(CV_motif,CV_alignment, pos = pos)
  #delete unsuccessful predictions
  predTrue <- predTrue[!is.na(predTrue$pred),]
  CSpredTrueList[[pos]] <- predTrue
  print(pos)
  
}
R2s_CS <- list()
evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
for(pos in motifPoses){
  predTrue <- CSpredTrueList[[pos]]
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  
  plotPredTrue(predTrue, main = paste0('Closest sequence prediction at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  
  plotPredTrue(predTrue.PFM, main = paste0('Closest sequence prediction PFM at ',pos))
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F, throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F, throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, 'Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T, throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
      predTrue$true, nrow(predTrue), throughZero = T), 
      4), round(RMSD(predTrue$pred, predTrue$true), 
      4), round(groupedR2(predTrue.PFM$pred, 
      predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
      4)^(1/2), round(hit,4))
  R2s <- groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F)
  R2s_CS[[pos]] <- R2s
}
evaMatrix_CS <- evaMatrix

```

### Comparing prediction result with Similarity regression
from paper:
Lambert, S. A., Yang, A. W., Sasse, A., Cowley, G., Albu, M., Caddick, M. X., ... & Hughes, T. R. (2019). Similarity regression predicts evolution of transcription factor sequence specificity. Nature genetics, 51(6), 981-989.
Figure S6/S7

```{r}
options(warn=0)
SRpredTrueList <- list()
for(pos in motifPoses){
  mod_alignment <- CV_alignment
  mod_alignment$alignment <- paste0(substr(mod_alignment$alignment,1,34), '-','-',substr(mod_alignment$alignment,35,53))
  predTrue <- SRpred(CV_motif,mod_alignment, pos = pos, weightfile = paste0(rawData,'SimilarityRegression/F082_1.97d.json'))
  #delete unsuccessful predictions
  predTrue <- predTrue[!is.na(predTrue$pred),]
  SRpredTrueList[[pos]] <- predTrue
  print(pos)
}
R2s_SR <- list()
evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
for(pos in motifPoses){
  predTrue <- SRpredTrueList[[pos]]
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  
  plot.ecdf(predTrue$similarity)
  plotPredTrue(predTrue, main = paste0('Similarity regression at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  
  plotPredTrue(predTrue.PFM, main = paste0('Similarity regression prediction PFM at ',pos))
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F, throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F, throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, 'Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T, throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
    predTrue$true, nrow(predTrue), throughZero = T), 
    4), round(RMSD(predTrue$pred, predTrue$true), 
    4), round(groupedR2(predTrue.PFM$pred, 
    predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
    4)^(1/2), round(hit,4))
  R2s <- groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F)
  R2s_SR[[pos]] <- R2s
}
evaMatrix_SR <- evaMatrix
```
Write Evaluation matrix
```{r}
evaMatrix <- rbind.data.frame(evaMatrix_BL, evaMatrix_FC,  evaMatrix_SR, evaMatrix_CS)
evaMatrix <- evaMatrix[c((1:4)*4-3, (1:4)*4-2, (1:4)*4-1,(1:4)*4), ]
```

## SVD prediction for single and double mutants

load SELEX motifs
```{r}
groups <- c('HWT', 'H5', 'H13','HD', 'AWT', 'A5', 'A13', 'AD')
SELEXFile_Template <- paste0(rawData,'SELEXdata/$modelFile$/fit.models.consensus.json')
SELEX_info <- data.frame(gene_symbol = groups, study = 'pb')
SELEX_motifs <- loadMono_motifs(SELEX_info, SELEXFile_Template, rec_seq = 'CANNTG', pos_index = c('P-3','P-2','P-1','P1','P2','P3'), checkSymmetry = T, withTable = F, useMode = c(1,2,1,1,1,1,1,1))
```


```{r 5}
result <- readLines(paste0(rawData,"SELEXdata/HWT_pb/fit.models.consensus.json"))
mono <- JSON2Matrix(result, mode = 1)
mononucleotide_logo(frequency2ddG(mono))
result <- readLines(paste0(rawData,"SELEXdata/AWT_pb/fit.models.consensus.json"))
mono <- JSON2Matrix(result, mode = 1)
mononucleotide_logo(frequency2ddG(mono))
test <- data.frame(pred = as.numeric(frequency2ddG(JSON2Matrix(readLines(paste0(rawData,"SELEXdata/HWT_pb/fit.models.consensus.json"))))[,4:6]), true = as.numeric(frequency2ddG(JSON2Matrix(readLines(paste0(rawData,"HTdata/HES2_Yin2017/result/fit.models.consensus.json"))))[,4:6]))
plotPredTrue(test,  xlab = 'HT-SELEX -ΔΔG/RT', ylab = 'Experimental -ΔΔG/RT', main = paste0('HES2 replicate'))
test <- data.frame(pred = as.numeric(frequency2ddG(JSON2Matrix(readLines(paste0(rawData,"SELEXdata/AWT_pb/fit.models.consensus.json"))))[,4:6]), true = as.numeric(frequency2ddG(JSON2Matrix(readLines(paste0(rawData,"HTdata/ASCL2_Yin2017/result/fit.models.consensus.json"))))[,4:6]))
plotPredTrue(test,  xlab = 'HT-SELEX -ΔΔG/RT', ylab = 'Experimental -ΔΔG/RT', main = paste0('ASCL2 replicate'))
```

```{r}
library(readxl)
library(ggplot2)
library(ggsignif)
bandIntensity <- read_excel(paste0(rawData,"bandIntensity.xlsx"))

proportions <- bandIntensity$...6[c(1:6,13:18)]
proportions <- c(proportions[1:3], exp(1.788), 8.09,
                 proportions[4:6],exp(1.956), 28.41,
                 proportions[7:9], exp(-3.948),0.023,
                 proportions[10:12], exp(-2.15), 0.135)
names <- c(rep('WT',5),rep('R5K',5),rep('V13R',5),rep('Double',5))
names <- as.factor(names)
cols <- c(rep('#6cab42',1),rep('#87eff2',1),rep('#f9d85b',1),rep('#c4560c',1))
dt <- data.frame(name = names, prop = proportions)
dt$prop <- log(dt$prop)
dt$name <- factor(dt$name, levels = c('WT', 'R5K', 'V13R', 'Double'))


p <- ggplot(dt, aes(x = name, y = prop, fill = name, color = name)) +
  geom_dotplot(binaxis = 'y', stackdir = 'center', dotsize = 1) +
  scale_fill_manual(values = cols, name = "Name") +
  scale_color_manual(values = cols, name = "Name") +
  labs(
    x = "",
    y = "-ΔΔG/RT (GC->CG)",
    size =15
  ) +
  theme(legend.position = "none",
    axis.text.x = element_text(face = 'bold',color=cols, size = 15),
    panel.background = element_blank(),     # Remove panel background
    panel.grid.major = element_blank(),     # Remove major gridlines
    panel.grid.minor = element_blank(),     # Remove minor gridlines
    plot.background = element_blank(),
    panel.border = element_rect(colour = "black", fill=NA, size=0.5)# Remove plot background
  ) +
  geom_signif(comparisons = list(c("R5K", "WT"),
                                 c("V13R", "Double"),
                                 c("R5K", "V13R"),
                                 c("WT", "V13R"),
                                 c("WT", "Double"),
                                 c("R5K", "Double")
                                 ),
              map_signif_level=TRUE, color = 'black',  step_increase = 0.1)
plot(p)
```
Applying the SVD-regression model to predict the binding motifs of the single and double mutants of HES2 and ASCL2. The key Position used are the mutated positions, and the number of positions to use for each PC is inferred from the line plot of Principle Component-AA position Association.


### tetrahedron dddG
```{r}
svd <- matrixSVD(gene2pos(CV_motif, pos = 'P-1'))
svdModelpreTrain <- trainSVD.Iterative(svd, CV_alignment, Ftest_pVal = 0.001)
keyPos <- svdModelpreTrain$keyPos
keyPos <- lapply(keyPos, function(x) c(x,5))
#svdModel <- trainSVD(svd, CV_alignment, no.keyPos = unlist(lapply(keyPos,length)), keyPos = keyPos)
svdModel <- trainSVD(svd, CV_alignment, no.keyPos = c(2,1,1), keyPos = list(c(13,5), c(13), c(13)))
#Predict binding motifs for test set
test_alignment <- data.frame(name = paste0(groups, '_pb')[-c(1,5)], alignment =c(
                             'RKSLRPLLEKRRRARINQSLSQLKGLLPENSNCSKLEKADVLEMTVRFLQEL-',
                             'RKSLKPLLEKRRVARINQSLSQLKGLLPENSNCSKLEKADVLEMTVRFLQEL-',
                             'RKSLRPLLEKRRVARINQSLSQLKGLLPENSNCSKLEKADVLEMTVRFLQEL-',
                             '--ARKNERERNRVKLVNLGFQALRQHVPHGGASKKLSKVETLRSAVEYIRALQ',
                             '--ARRNERERNRRKLVNLGFQALRQHVPHGGASKKLSKVETLRSAVEYIRALQ',
                             '--ARKNERERNRRKLVNLGFQALRQHVPHGGASKKLSKVETLRSAVEYIRALQ')
                             )
test_motifs <- SELEX_motifs[-c(1,5)]
pred_motifs <- predict(svdModel, test_alignment[], zero = 0.01, useSimilarAA = T)
#Comparing between true and predicted testing set motifs 

predT <- matrix2tetrahedron(pred_motifs)

true <- lapply(test_motifs, function(x) x$matrix[,c('P-1','P-1')])
trueT <- lapply(true, matrix2tetrahedron)
trueT <- matrix(nrow = 6, ncol = 6, data = unlist(trueT), byrow = T)
trueT <- trueT[,c(1,3,5)]
wt <- lapply(SELEX_motifs[c(1,5)], function(x) x$matrix[,c('P-1','P-1')])
wtT <- lapply(wt, matrix2tetrahedron)
wtT <- matrix(nrow = 2, ncol = 6, data = unlist(wtT), byrow = T)
wtT <- wtT[,c(1,3,5)]
pred <- predT[1:6,] - wtT[c(1,1,1,2,2,2),]
true <- trueT[1:6,] - wtT[c(1,1,1,2,2,2),]
pred <- tetrahedron2matrix(pred)
true <- tetrahedron2matrix(true)
pred[pred < 0.01] <- 0.01
true[true < 0.01] <- 0.01

pred <- frequency2ddG(pred)
true <- frequency2ddG(true)
predTrueHes2 <- data.frame(pred = as.numeric(unlist(pred[,1:3])), true = as.numeric(unlist(true[,1:3])))
plotPredTrue(predTrueHes2,  xlab = 'Experimental -ΔΔΔG/RT', ylab = 'Predicted -ΔΔΔG/RT', main = paste0('FamilyCode on HES2 mutants'))
predTrueAscl2 <- data.frame(pred = as.numeric(unlist(pred[,4:6])), true = as.numeric(unlist(true[,4:6])))
plotPredTrue(predTrueAscl2,  xlab = 'Experimental -ΔΔΔG/RT', ylab = 'Predicted -ΔΔΔG/RT', main = paste0('FamilyCode on ASCL2 mutants'))
```

### checking with EMSA measured ddGs
```{r}
dm <- matrix(nrow = 5, ncol = 4, data = dt$prop)
meanddGs <- apply(dm, 2, mean)
predddG <- pred[3,4:6] - pred[2,4:6]
EMSAdddG <- meanddGs[2:4] - meanddGs[1]
predTrue <- data.frame(pred = predddG, true = EMSAdddG)
plot(EMSAdddG, predddG, xlab = 'EMSA measrued -ΔΔΔG', ylab = 'family code predicted -ΔΔΔG', main = 'EMSA measured vs. family code predicted -ΔΔΔG')
abline(lm(pred~true, predTrue), col = '#666666', lty = 2)
legend('topleft', paste0('R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, nrow(predTrue)),4), '\n',
                         'RMSD = ', round(RMSD(predTrue$pred,predTrue$true),4)), bty = 'n')
```
#R2 p-value
Calculating bootstrap p-values for R2s
```{r}
mod_alignment <- CV_alignment
mod_alignment$alignment <- paste0(substr(mod_alignment$alignment,1,34), '-','-',substr(mod_alignment$alignment,35,53))
bootstrap <- matrix(nrow = 52, ncol = 6, data = 0)
for(i in 1:52){
  #FC
  predTrue <- SVDregression.Iterative.CV(CV_motif[-i],CV_alignment[-i,], pos = 'P-1',Ftest_pVal = 0.001, useSimilarAA = T)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  bootstrap[i,1] <- round(groupedR2(predTrue$pred, predTrue$true, nrow(predTrue), throughZero = T), 4)
  bootstrap[i,2] <- round(RMSD(predTrue$pred, predTrue$true), 4)
  #CS
  predTrue <- closestSeqPred(CV_motif[-i],CV_alignment[-i,], pos = 'P-1')
  predTrue <- predTrue[!is.na(predTrue$pred),]
  bootstrap[i,3] <- round(groupedR2(predTrue$pred, predTrue$true, nrow(predTrue), throughZero = T), 4)
  bootstrap[i,4] <- round(RMSD(predTrue$pred, predTrue$true), 4)
  #SR
  predTrue <- SRpred(CV_motif[-i],mod_alignment[-i,], pos = 'P-1', weightfile = 'cisBP_bHLH/F082_1.97d.json')
  #delete unsuccessful predictions
  predTrue <- predTrue[!is.na(predTrue$pred),]
  bootstrap[i,5] <- round(groupedR2(predTrue$pred, predTrue$true, nrow(predTrue), throughZero = T), 4)
  bootstrap[i,6] <- round(RMSD(predTrue$pred, predTrue$true), 4)
  print(i)
}

t.test(bootstrap[,1], bootstrap[,3], paired = T)$p.value
t.test(bootstrap[,2], bootstrap[,4], paired = T)$p.value
t.test(bootstrap[,1], bootstrap[,5], paired = T)$p.value
t.test(bootstrap[,2], bootstrap[,6], paired = T)$p.value

```

# Supplemental data S2
```{r SD1}
library(R.utils)
mkdirs('S2')
for(i in 1:nrow(bHLH_index)){
  result <- readLines(paste0(rawData,"HTdata/",
                                     bHLH_index$gene_symbol[i], '_', bHLH_index$study[i],'/result/fit.models.consensus.json'))
  mono <- JSON2Matrix(result, mode = bHLH_index$motif_model[i])
  plot <- mononucleotide_logo(mono)
  plot <- plot + ggplot2::ggtitle(paste0(gsub('_','',bHLH_index$gene_symbol[i]), '\nR^2: ', bHLH_index$quality[i]))
  png(paste0('S2/',gsub('_','',bHLH_index$gene_symbol[i]),'.png'))
  plot(plot)
  dev.off()
}

```

# Supplemental data S3
```{r SD2}
library(plotly)
mkdirs('S3')
pos_index <- c('P-3','P-2','P-1','P1','P2','P3')
for(i in 1:6){
  for(j in c(5,13)){
    AAPosition <- j
    motifPosition <- pos_index[i]
    AAbp_Matrix <- AAbpCombination(CV_motif, CV_alignment, AAPosition, motifPosition)
    plot <- plot_tetrahedron(AAbp_Matrix, color = T, vertex = F, size = 10)
    scene = list(camera = list(eye = list(x = 1.25, y = 1.25, z = 1.25)))
    plot <- plot%>%layout(scene=scene)
    htmlwidgets::saveWidget(as_widget(plot), paste0("S3/AA",j,"_",motifPosition,".html"))
  }
}
```

# Supplemental data S4
```{r SD3}
mkdirs('S4')
posMatrix <- gene2pos(CV_motif, pos = 'P-1')
svd <- matrixSVD(posMatrix)
#ANOVA test on the level of association between SVD Principal components and residue type
svd.pval <- svdANOVA(svd, CV_alignment)
for(j in 1:56){
  for(u in 1:3){
    pos <- j
    svd <- matrixSVD(gene2pos(CV_motif, pos = 'P-1'))
    x <- svd$u[,u]
    y <- substr(CV_alignment$alignment, pos,pos)
    dt <- data.frame(AA = y, u = x)
    dt$col <- AAcolor(dt$AA)
    dt <- arrange(dt,u)
    colScheme <- unique(dt[,c(1,3)])
    png(paste0('S4/AA',pos,'_PC',u,'.png'))
    plot(x = dt$u, y = 1:52/52, col = dt$col, pch = 19, main = paste0('eCDF for PC',u,' at position ', pos), xlab = paste0('PC',u), ylab = 'Frequency', cex = 1, cex.lab = 1.2, cex.axis = 1.2)
    legend(x = "topleft", legend = colScheme$AA, col =colScheme$col, pch = 19, cex = 1, bty = 'n')
    legend(x = "bottomright", legend = paste0('-log(p-Value): ',round(svd.pval[pos,u],2)), bty = 'n')
    dev.off()
  }
}

```

# Supplemental data S5
```{r SD4}
#1400/400
mkdirs('S5')
pos_index <- c('P1','P2','P3')
for(i in 1:3){
  posMatrix <- gene2pos(CV_motif, pos = pos_index[i])
  svd <- matrixSVD(posMatrix)
  #ANOVA test on the level of association between SVD Principal components and residue type
  svd.pval <- svdANOVA(svd, CV_alignment)
  #Plot p-vals 
  png(paste0('S5/PC_',i,'.png'), width = 1400, height = 400)
  plot(svd.pval$X1, pch = 1, col = "red", type = 'l', lwd = 3, xlab = 'AA Position', ylab = '-log p-val', ylim = c(0, (1.1*max(svd.pval[,1:3]))) ,main = paste0('Statistical significance of ANOVA test for PCs at DNA position -', i, '/+',i))
  lines(svd.pval$X2, pch = 1, col = "green", lwd = 3)
  lines(svd.pval$X3, pch = 1, col = "blue", lwd = 3)
  legend(l, x = "topright", legend = c('PC1', 'PC2', 'PC3'), lwd = 2, col =c('red','green','blue'), lty=1, cex = 0.8, bty = 'n')
  dev.off()
}

```
