---
title: "FamilyCode on SELEX and PBM data with pyProBound"
output: html_notebook
---

### Load SELEX motifs
```{r}
library(FamilyCode)
workDir <- 'intermediateData/HD/PBM_HD/'
rawData <- 'rawData/HD/'
motifPoses <- c('N1','N2','T3','D4','A5','Y6','N7','N8')
#loadMotifs
proBound_DIR <- paste0(rawData, 'consensus_fits/')
template <- paste0(rawData, 'consensus_fits/$modelFile$.consensus.json')
gene_symbol_factor_ids <- read.csv(paste0(rawData, 'gene_symbol_factor_ids.csv'), row.names=1)
files <- list.files(proBound_DIR, pattern = 'consensus')
files <- unlist(lapply(strsplit(files, '\\.'), function(x) x[[1]][1]))
gene_symbol <- unlist(lapply(strsplit(files, '_'), function(x) x[[1]][1]))
study <- unlist(lapply(strsplit(files, '_'), function(x) x[[2]][1]))
info <- data.frame(gene_symbol, study)
info <- info[as.numeric(info$gene_symbol) < 10000,]

HD_all_motifs <- loadFromIndex(info, template)
motifPositions <- motifPoses
HDseed <- matrix(nrow = 4,ncol = 8,data = c(0,0,0,0,
                                            0,0,0,0,
                                            0,0,0,1,
                                            1,0,1,1,
                                            1,0,0,0,
                                            0,1,0,1,
                                            0,0,0,0,
                                            0,0,0,0))
rownames(HDseed) <- DNA()

HD_motifScores <- scoreMotifList(HD_all_motifs, HDseed, weight = c(0,0,1,1,1,1,0,0))

HDids<- unique(HD_motifScores$gene_symbol)
HDfiltered_table <- data.frame(NULL)
for(i in 1:length(HDids)){
  fDT <- HD_motifScores[HD_motifScores$gene_symbol == HDids[i],]
  HDfiltered_table <- rbind.data.frame(HDfiltered_table, fDT[which.min(fDT$score),])
}
plot.ecdf(as.numeric(HDfiltered_table$score))
HDfiltered_table <- HDfiltered_table[HDfiltered_table$score < 1,]

#prepare alignment
HD_ali <- read.table(paste0(rawData,"HD_hmm.sto"), quote="\"", fill = T)
HD_ali$V1 <- apply(HD_ali, 1, function(x) strsplit(x[1],'/')[[1]][1])
colnames(HD_ali) <- c('name','alignment')

#change names
geneName <- apply(HDfiltered_table, 1, function(x) gene_symbol_factor_ids$gene_symbol[gene_symbol_factor_ids$factor_ids == x[1]])
HDfiltered_table$gene_symbol <- geneName
HD_motifs <- filterMotifList(HDfiltered_table, HD_all_motifs, 8,motifPositions)


#quality filter
HD_alignment <- matchAliMotif(HD_motifs, HD_ali)
del <- which(substr(HD_alignment$alignment, 50,50) != 'N')
HD_alignment <- HD_alignment[-del,]
HD_motifs <- HD_motifs[-del]


del <- which(duplicated(HD_alignment$name))
HD_alignment_SELEX <- HD_alignment[-del,]
HD_motif_SELEX <- HD_motifs[-del]
```

### Prepare Base line
```{r}
#get base line data set
HDidstds <- unique(HD_motifScores[,1:2])
HDfiltered_table_baseline <- data.frame(NULL)
for(i in 1:nrow(HDidstds)){
  fDT <- HD_motifScores[HD_motifScores$gene_symbol == HDidstds[i,1],]
  fDT <- fDT[fDT$study == HDidstds[i,2],]
  HDfiltered_table_baseline <- rbind.data.frame(HDfiltered_table_baseline, fDT[which.min(fDT$score),])
}
plot.ecdf(as.numeric(HDfiltered_table_baseline$score))

HDfiltered_table_baseline <- HDfiltered_table_baseline[HDfiltered_table_baseline$score < 1,]
HD_motifs_baseline <- filterMotifList(HDfiltered_table_baseline, HD_all_motifs, 8, motifPositions)

```

### Import library and data of PBM data
```{r}
library(R.utils)
#download from cisBP 'bulk download' tab, inducing Z-score and TF_Information
TF_Information <- read.delim(paste0(rawData, "TF_Information.txt"))
Zscores <- read.delim(paste0(rawData,"Zscores.txt"))
#filter for PBM data with Z-scores
TFDirect <- TF_Information[TF_Information$TF_Status == 'D',]
rawPBM <- substr(colnames(Zscores)[-1], 1, 11)
#deal with duplicated experiment
substr(rawPBM[duplicated(rawPBM)],11,11) <- '1'
colnames(Zscores)[-1] <- rawPBM
TFDirect.duplicated <- TFDirect
substr(TFDirect.duplicated$Motif_ID,11,11) <- '1'
TFDirect <- rbind.data.frame(TFDirect, TFDirect.duplicated)
#get cisBP entries with PBM Z-scores
TFDirect <- TFDirect[match(intersect(TFDirect$Motif_ID, rawPBM),TFDirect$Motif_ID),]

geneInfoUniprot <- read.delim(paste0(workDir,"geneInfoUniprot_all.tsv"))
geneInfoUniprot$Gene.Names <- paste0(geneInfoUniprot$Gene.Names, ' ')

gene_ids <- unique(TFDirect[,c('TF_Name', 'TF_Species')])
geneInfo <- data.frame()
for(i in 1:nrow(gene_ids)){
  add <- geneInfoUniprot[grep(paste0(gene_ids$TF_Name[i],' '), geneInfoUniprot$Gene.Names),]
  if(nrow(add) != 0){
    add <- add[nchar(add$Sequence) == max(nchar(add$Sequence)),][1,]
    addSeq <- add$Sequence
  }else{
    addSeq <- NA 
  }
  Motif <- unique(TFDirect[TFDirect$TF_Name == gene_ids$TF_Name[i],]$Motif_ID)
  addLine <- data.frame(name = gene_ids$TF_Name[i], seq = addSeq, motif = Motif)
  geneInfo <- rbind.data.frame(geneInfo, addLine)
}

geneInfo <- geneInfo[!is.na(geneInfo$seq),]


```

```{r}
ali <- read.table(paste0(workDir,"TFhmm.sto"), quote="\"", fill = T)
ali <- ali[-nrow(ali),]
ali$V1 <- apply(ali, 1, function(x) strsplit(x[1],'/')[[1]][1])
ali$V1 <- apply(ali, 1, function(x) strsplit(x[1],'\\|')[[1]][2])
colnames(ali) <- c('name','alignment')
keep <- which(substr(ali$alignment, 50,50) == 'N')
cisBP_alignment <- ali[keep,]
geneInfo <- geneInfo[keep,]
#Check if cisBP_alignment and geneInfo matches
sum(cisBP_alignment$name != geneInfo$name) == 0
```
```{r}
cisBP_motifs <- list()
options(warn=0)
for(i in 1:nrow(geneInfo)){
  motif <- geneInfo$motif[i]
  motif <- read.table(paste0(workDir, 'pwms/', motif, '.txt'), header = F)
  motif <- ddG2frequency(motif)
  motif[motif < 0.001] <- 0.001#add psuedo count
  rownames(motif) <- DNA()
  add <- list()
  add$gene_symbol <- geneInfo$name[i]
  add$study <- i
  add$mode <- 1
  add$matrix <- as.matrix(motif)
  cisBP_motifs[[length(cisBP_motifs)+1]] <- add
  add <- list()
  #add reverse compliment motif
  add$gene_symbol <- geneInfo$name[i]
  add$study <- i
  add$mode <- -1
  if(ncol(motif) == 0){
    rev <- motif
  }else{
    rev <- as.matrix(motif)[nrow(motif):1, ncol(motif):1]
  }
  rownames(rev) <- DNA()
  add$matrix <- rev
  cisBP_motifs[[length(cisBP_motifs)+1]] <- add
}
#check if length of cisBP_motif is equal to 2 times geneInfo
length(cisBP_motifs) == nrow(geneInfo)*2

```
```{r}
#align to NNTDAYNN motif
HDseed <- matrix(nrow = 4,ncol = 8,data = c(0,0,0,0,
                                            0,0,0,0,
                                            0,0,0,1,
                                            1,0,1,1,
                                            1,0,0,0,
                                            0,1,0,1,
                                            0,0,0,0,
                                            0,0,0,0))
rownames(HDseed) <- DNA()

motifScores <- scoreMotifList(cisBP_motifs, HDseed, weight = c(0,0,1,1,1,1,0,0))
```

### Filter and reorder motif information according to motif matching score.
The motif (forward or reverse compliment) with the lowest motif matching score of each entry will be kept. For the same protein, entries are arranged from lower score to higher score
```{r}
idstds <- unique(motifScores[,1:2])
all_motifs <- data.frame(NULL)
for(i in 1:nrow(idstds)){
  fDT <- motifScores[motifScores$gene_symbol == idstds[i,1],]
  fDT <- fDT[fDT$study == idstds[i,2],]
  all_motifs <- rbind.data.frame(all_motifs, fDT[which.min(fDT$score),])
}
motifs_arranged <- data.frame(NULL)
motifScore <- c()
for(i in 1:length(unique(all_motifs$gene_symbol))){
  add <- all_motifs[all_motifs$gene_symbol == unique(all_motifs$gene_symbol)[i],]
  add <- dplyr::arrange(add, score)
  motifScore <- c(motifScore,add$score[1])
  motifs_arranged <- rbind.data.frame(motifs_arranged, add)
}
#Plot distribution of lowest motif matching score
plot.ecdf(as.numeric(motifScore))
```
```{r}
#Filter out motifs that do not match well with the seed
motifTable_baseline <- motifs_arranged[motifs_arranged$score < 1,]
#Number of unique HD protiens kept:
length(unique(motifTable_baseline$gene_symbol))
```

```{r}
motifPoses <- c('N1','N2','T3','D4','A5','Y6','N7','N8')
cisBP_motifs_baseline <- filterMotifList(motifTable_baseline, cisBP_motifs, 8, motifPoses)
cisBP_alignment <- cisBP_alignment[match(motifTable_baseline$gene_symbol, cisBP_alignment$name),]
#check if motif List has same length as alignment
length(cisBP_motifs_baseline) == nrow(cisBP_alignment)
```

```{r}
del <- which(duplicated(cisBP_alignment$alignment))
HD_alignment_PBM <- cisBP_alignment[-del,]
HD_motif_PBM <- cisBP_motifs_baseline[-del]
```

###Concatnate two data sets 
```{r}
intersect <- intersect(HD_alignment_PBM$alignment, HD_alignment_SELEX$alignment)
PBM_int <- match(intersect, HD_alignment_PBM$alignment)
SELEX_int <- match(intersect, HD_alignment_SELEX$alignment)
all_motif <- c(HD_motif_SELEX, HD_motif_PBM)
CV_alignment <- rbind.data.frame(HD_alignment_SELEX, HD_alignment_PBM[-PBM_int,])
CV_motif <- c(HD_motif_SELEX, HD_motif_PBM[-PBM_int])
nrow(CV_alignment) == length(CV_motif)
saveRDS(CV_alignment, file = paste0('intermediateData/HD/HD_SELEX_PBM_CV_alignment.rsd'))
saveRDS(CV_motif, file =  paste0('intermediateData/HD/HD_SELEX_PBM_CV_motif.rsd'))
CV_alignment <- readRDS( paste0('intermediateData/HD/HD_SELEX_PBM_CV_alignment.rsd'))
CV_motif <- readRDS( paste0('intermediateData/HD/HD_SELEX_PBM_CV_motif.rsd'))
```

### Show base line accuracy
```{r}
evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
for(pos in motifPoses){
  ScreenPoses <- c(pos,pos)
  BaseLines.ddG <- getBaseLineAccuracy(all_motif, pos = ScreenPoses, randomSample = F)
  colnames(BaseLines.ddG) <- c('pred','true')
  BaseLines.PFM <- predTrue.ddG2frequency(BaseLines.ddG, PFM = T)
  plotPredTrue(BaseLines.ddG, xlab = 'Experiment 1', ylab = 'Experiment 2', main = paste0('Base Line -ΔΔG/RT at ',pos))
  boxplot(groupedR2(BaseLines.ddG$pred, BaseLines.ddG$true, member = length(ScreenPoses)*4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(BaseLines.ddG$pred, BaseLines.ddG$true, 
                                                          member = length(ScreenPoses)*4, mean= T),4)))
  plotPredTrue(BaseLines.PFM, xlab = 'Experiment 1', ylab = 'Experiment 2', main = paste0('Base Line PFM at ',pos))
  hit <- sum(round(groupedR2(BaseLines.PFM$pred, BaseLines.PFM$true, member = 4, mean= F, throughZero = F)^(1/2),4) > 0.5)/(nrow(BaseLines.PFM)/4)
  boxplot(groupedR2(BaseLines.PFM$pred, BaseLines.PFM$true, member = length(ScreenPoses)*4, mean= F, throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(BaseLines.PFM$pred, BaseLines.PFM$true, 
                                                     member = length(ScreenPoses)*4, mean= T)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(BaseLines.ddG$pred, 
        BaseLines.ddG$true, nrow(BaseLines.ddG), throughZero = T), 
        4), round(RMSD(BaseLines.ddG$pred, BaseLines.ddG$true), 
        4), round(groupedR2(BaseLines.PFM$pred, 
        BaseLines.PFM$true, nrow(BaseLines.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
}
evaMatrix_BL <- evaMatrix
```

### Make p-val table
```{r}
#genreate p-val tablle
pvalTable <- getPvalTable(CV_motif, Alignment = CV_alignment, pos_index = motifPoses)
#apply -log(10)
pvalTable <- -log(pvalTable,10)
#set NAs to 0
pvalTable[is.na(pvalTable)] <- 0
#plot heatmap
gplots::heatmap.2(as.matrix(as.data.frame(lapply(pvalTable, as.numeric))),dendrogram='none',
                  Rowv=FALSE, Colv=FALSE,trace='none',col = rev(heat.colors(12)), key.title = '-logP-val')


AAPosition <- 49
motifPosition <- 'Y6'
AAbp_Matrix <- AAbpCombination(CV_motif, CV_alignment, AAPosition, motifPosition)
plot_tetrahedron(AAbp_Matrix, color = T, vertex = F, size = 10, label = T)
pos <- AAPosition
u <- 1
svd <- matrixSVD(gene2pos(CV_motif, pos = motifPosition))
x <- svd$u[,u]
y <- substr(CV_alignment$alignment, pos,pos)
dt <- data.frame(AA = y, u = x)
dt$col <- AAcolor(dt$AA)
dt <- dplyr::arrange(dt,u)
colScheme <- unique(dt[,c(1,3)])
plot(x = dt$u, y = 1:length(dt$u)/length(dt$u), col = dt$col, pch = 19, main = paste0('eCDF for PC',u,' at position ', pos), xlab = paste0('PC',u), ylab = 'Frequency', cex = 1, cex.lab = 1.2, cex.axis = 1.2)
legend(x = "topleft", legend = colScheme$AA, col =colScheme$col, pch = 19, cex = 1, bty = 'n')
```

### save motif models
```{r}
library(R.utils)
mkdirs('HDlogos')
for(i in 1:length(CV_motif)){
  mono <- CV_motif[[i]]$matrix
  plot <- mononucleotide_logo(frequency2ddG(mono), axes = F, labels = F)
  png(paste0('HDlogos/',CV_alignment$name[i],'_',substr(CV_alignment$alignment[i], 5,15), '-',substr(CV_alignment$alignment[i], 44,54),'.png'),
       width = 600, height = 200,)
  plot(plot)
  dev.off()
}

```

### plot by residue logo
```{r}
AAPosition <- 46
motifPosition <- 'D4'
AAbp_Matrix <- AAbpCombination(CV_motif, CV_alignment, AAPosition, motifPosition)
plot_tetrahedron(AAbp_Matrix, color = T, vertex = F, size = 10)
mononucleotide_logo(frequency2ddG(AAbp_Matrix[,colnames(AAbp_Matrix) == 'N'][,c(21:50)]))
```



### Test on FamilyCode prediction with leave-one-out CV
Compare with paper: Wetzel, J. L., Zhang, K., & Singh, M. (2022). Learning probabilistic protein–DNA recognition codes from DNA-binding specificities using structural mappings. Genome Research, 32(9), 1776-1786.
Accuracy measurement in Figure 2
"PWM F, where entry fbj is the frequency with which nucleotide b is observed in column j in a set of aligned binding sites of that TF. "
```{r}
FCpredTrueList <- list()
for(pos in motifPoses){
  predTrue <- SVDregression.Iterative.CV(CV_motif,CV_alignment, pos = pos,Ftest_pVal = 0.001, useSimilarAA = T)
  #delete unsuccessful predictions
  predTrue <- predTrue[!is.na(predTrue$pred),]
  FCpredTrueList[[pos]] <- predTrue
  print(pos)
  
}
evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
PFMR2 <- matrix(ncol = 8, nrow = nrow(FCpredTrueList[[1]])/4, data = NA)
i = 1
for(pos in motifPoses){
  predTrue <- FCpredTrueList[[pos]]
  #delete unsuccessful predictions
  predTrue <- predTrue[!is.na(predTrue$pred),]
  
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('FamilyCode at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  #plot confidance
  attr(predTrue, 'confidence') <- attr(predTrue, 'confidence')[attr(predTrue, 'confidence') <= 1]
  plot(attr(predTrue, 'confidence'),groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F), ylab = 'Predict-True R^2', xlab = 'Confidence score', main = paste0('Confidence at ', pos))
  legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
                                                     attr(predTrue, 'confidence'), ),4)), bty = 'n')
  dt <- data.frame(conf = attr(predTrue, 'confidence'), R2 = groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F))
  lm <- lm(R2~conf, data= dt)
  abline(lm, col= 'red')
  
  plotPredTrue(predTrue.PFM, main = paste0('FamilyCode PFM at ',pos), xlab = 'Expermental', ylab = 'Predicted')
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F, throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F, throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T, throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  addData <- groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F, throughZero = F)^(1/2)
  PFMR2[1:length(addData),i] <- addData
  i <- i + 1
}
PFMR2 <- as.data.frame(PFMR2)
colnames(PFMR2) <- motifPoses
library(tidyr)
data_long <- PFMR2 %>%
  gather(key = "SampleGroup", value = "Value")

data_long$SampleGroup <- factor(data_long$SampleGroup, levels = motifPoses)
library(ggplot2)
library(lvplot)
ggplot(data_long, aes(x = SampleGroup, y = Value, fill = SampleGroup)) +
  geom_lv() +
  stat_summary(
    fun = median, 
    geom = "crossbar", 
    width = 1, 
    color = "black",      # Set the median bar color here
    size = 0.2          # Set the thickness of the median bar
  ) +
  scale_fill_manual(values = c(rep("grey80",8))) +
  theme_minimal() +
  theme(
    legend.position = "none",           # Removes the legend
    panel.background = element_rect(fill = "white"),  # Changes the plot background to white
    plot.background = element_rect(fill = "white"),
    panel.grid.major = element_blank(),               # Removes major grid lines
    panel.grid.minor = element_blank() # Changes the entire plot's background to white
  ) +
  labs(title = "",
       x = "Motif positions",
       y = "PCC between predicted and experimental PFM")

evaMatrix_FC <- evaMatrix
```

### Comparing prediction result with Closest paralog
from paper:
Weirauch, M. T., Yang, A., Albu, M., Cote, A. G., Montenegro-Montero, A., Drewe, P., ... & Hughes, T. R. (2014). Determination and inference of eukaryotic transcription factor sequence specificity. Cell, 158(6), 1431-1443.
Figure S6/S7
```{r}
evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
for(pos in motifPoses){
  predTrue <- closestSeqPred(CV_motif,CV_alignment, pos = pos)
  #delete unsuccessful predictions
  predTrue <- predTrue[!is.na(predTrue$pred),]
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  
  #plot.ecdf(predTrue$similarity)
  plotPredTrue(predTrue, main = paste0('Closest sequence prediction -ΔΔG/RT at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  
  plotPredTrue(predTrue.PFM, main = paste0('Closest sequence prediction PFM at ',pos))
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F, throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F, throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, 'Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T, throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
      predTrue$true, nrow(predTrue), throughZero = T), 
      4), round(RMSD(predTrue$pred, predTrue$true), 
      4), round(groupedR2(predTrue.PFM$pred, 
      predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
      4)^(1/2), round(hit,4))
}
evaMatrix_CS <- evaMatrix
```

### Comparing prediction result with Similarity regression
from paper:
Lambert, S. A., Yang, A. W., Sasse, A., Cowley, G., Albu, M., Caddick, M. X., ... & Hughes, T. R. (2019). Similarity regression predicts evolution of transcription factor sequence specificity. Nature genetics, 51(6), 981-989.
Figure S6/S7

```{r}
evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
for(pos in motifPoses){
  predTrue <- SRpred(CV_motif,CV_alignment, pos = pos, weightfile = 'cisBP_HD/F223_1.97d.json')
  #delete unsuccessful predictions
  predTrue <- predTrue[!is.na(predTrue$pred),]
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  
  plotPredTrue(predTrue, main = paste0('Similarity regression prediction -ΔΔG/RT at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  
  plotPredTrue(predTrue.PFM, main = paste0('Similarity regression prediction PFM at ',pos))
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F, throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F, throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, 'Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T, throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
    predTrue$true, nrow(predTrue), throughZero = T), 
    4), round(RMSD(predTrue$pred, predTrue$true), 
    4), round(groupedR2(predTrue.PFM$pred, 
    predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
    4)^(1/2), round(hit,4))
}
evaMatrix_SR <- evaMatrix
```
# Supplemental data S6
```{r SD1}
library(R.utils)
mkdirs('S6')
for(i in 1:length(CV_motif)){
  plot <- mononucleotide_logo(frequency2ddG(CV_motif[[i]]$matrix))
  plot <- plot + ggplot2::ggtitle(paste0(gsub('_','',CV_motif[[i]]$name)))
  png(paste0('S6/HD_',i,'.png'))
  plot(plot)
  dev.off()
}

```

# Supplemental data S7
```{r SD2}
library(plotly)
mkdirs('S7')
pos_index <- motifPoses
sigPos <- which(apply(pvalTable, 1, max) > -log(0.001/57/8))
for(i in 1:8){
  for(j in c(5,6,46,49,53,55)){
    AAPosition <- j
    motifPosition <- pos_index[i]
    AAbp_Matrix <- AAbpCombination(CV_motif, CV_alignment, AAPosition, motifPosition)
    plot <- plot_tetrahedron(AAbp_Matrix, color = T, vertex = F, size = 10)
    scene = list(camera = list(eye = list(x = 1.25, y = 1.25, z = 1.25)))
    plot <- plot%>%layout(scene=scene)
    htmlwidgets::saveWidget(as_widget(plot), paste0("S7/AA",j,"_",motifPosition,".html"))
  }
}
```

# HD WT cross-validation with hamming distance threshold

```{r}
CV_alignment <- readRDS( paste0('intermediateData/HD/HD_SELEX_PBM_CV_alignment.rsd'))
CV_motif <- readRDS( paste0('intermediateData/HD/HD_SELEX_PBM_CV_motif.rsd'))
motifPoses <- c('N1','N2','T3','D4','A5','Y6','N7','N8') #c('P-3','P-2','P-1')
library(Biostrings)

# Convert to DNAStringSet
aa_set <- AAStringSet(CV_alignment$alignment)

# Compute pairwise Hamming distances
dist_matrix <- stringDist(aa_set, method = "hamming")
dist.matrix <- as.matrix(dist_matrix)
step <- floor(min(apply(dist.matrix,2,max))/5)-1
```


```{r}
# predicting for each distance treshold with FC
FCR2 <- data.frame(NULL)
for(t in 0:step){
  R2ClusterFC <- data.frame(NULL)
  all_pred_motifs <- list()
  for(i in 1:nrow(dist.matrix)){
    train_motifs <- CV_motif[dist.matrix[i,] > 5*t]
    train_alignment <- CV_alignment[dist.matrix[i,] > 5*t,]
    if(length(train_motifs) < 1){
      next()
    }
    test_motifs <- CV_motif[i]
    test_alignment <- CV_alignment[i,]
    svdModel <- makeSVDModel(train_motifs, train_alignment, Positions = motifPoses)
    #predict mutant
    pred_motifs <- list()
    for(j in 1:length(test_motifs)){
      pred_matrix <- SVDmodelPredPSAM(svdModel,test_alignment$alignment[j])
      pred_motifs[[length(pred_motifs)+1]] <- list(name = test_alignment$name[j], matrix = pred_matrix)
    }
    all_pred_motifs[length(all_pred_motifs)+1] <- pred_motifs[1]
  }
  R2s <- c()
  for(pos in motifPoses){
    predMatrix <- frequency2ddG(gene2pos(all_pred_motifs, pos))
    testMatrix <- frequency2ddG(gene2pos(CV_motif, pos))
    predTrue <- data.frame(true = as.numeric(testMatrix), pred = as.numeric(predMatrix))
    #plotPredTrue(predTrue, main = paste0('FamilyCode prediction for cluster ',i,' at pos ', pos), xlab = 'Experimental -ΔΔG/RT', ylab = 'Predicted -ΔΔG/RT')
    R2s <- c(R2s, groupedR2(predTrue$true, predTrue$pred, member = nrow(predTrue)))
  }
  R2ClusterFC <- rbind.data.frame(R2ClusterFC, R2s)
  
  colnames(R2ClusterFC) <- motifPoses
  meanR2 <- apply(R2ClusterFC, 2, mean)
  FCR2 <- rbind.data.frame(FCR2, meanR2)
  print(t)
}
colnames(FCR2) <- motifPoses
```


```{r}
# predicting for each distance treshold with CS
CSR2 <- data.frame(NULL)
for(t in 0:step){
  R2Cluster <- data.frame(NULL)
  all_pred_motifs <- list()
  for(i in 1:nrow(dist.matrix)){
    train_motifs <- CV_motif[dist.matrix[i,] > 5*t]
    train_alignment <- CV_alignment[dist.matrix[i,] > 5*t,]
    if(length(train_motifs) < 1){
      next()
    }
    test_motifs <- CV_motif[i]
    test_alignment <- CV_alignment[i,]
    #predict mutant
    pred_motifs <- list()
    pred_motifs <- CSpredPSAM(test_alignment, train_motifs, train_alignment)
    all_pred_motifs[length(all_pred_motifs)+1] <- pred_motifs[1]
  }
  R2s <- c()
  for(pos in motifPoses){
    predMatrix <- frequency2ddG(gene2pos(all_pred_motifs, pos))
    testMatrix <- frequency2ddG(gene2pos(CV_motif, pos))
    predTrue <- data.frame(true = as.numeric(testMatrix), pred = as.numeric(predMatrix))
    #plotPredTrue(predTrue, main = paste0('FamilyCode prediction for cluster ',i,' at pos ', pos), xlab = 'Experimental -ΔΔG/RT', ylab = 'Predicted -ΔΔG/RT')
    R2s <- c(R2s, groupedR2(predTrue$true, predTrue$pred, member = nrow(predTrue)))
  }
  R2Cluster <- rbind.data.frame(R2Cluster, R2s)
  
  colnames(R2Cluster) <- motifPoses
  meanR2 <- apply(R2Cluster, 2, mean)
  CSR2 <- rbind.data.frame(CSR2, meanR2)
  print(t)
}
colnames(CSR2) <- motifPoses
```


```{r}
# predicting for each distance treshold with SR
SRR2 <- data.frame(NULL)
isbHLH <- sum('P-1' == motifPoses) + sum('P1' == motifPoses) > 0
if(isbHLH){
  mod_alignment <- CV_alignment
  mod_alignment$alignment <- paste0(substr(mod_alignment$alignment,1,34), '-','-',substr(mod_alignment$alignment,35,53))
  wf <- 'rawData/bHLH/SimilarityRegression/F082_1.97d.json'
}else{
  mod_alignment <- CV_alignment
  wf <- 'rawData/HD/SimilarityRegression/F223_1.97d.json'
}
for(t in 0:step){
  R2Cluster <- data.frame(NULL)
  all_pred_motifs <- list()
  for(i in 1:nrow(dist.matrix)){
    train_motifs <- CV_motif[dist.matrix[i,] > 5*t]
    train_alignment <- mod_alignment[dist.matrix[i,] > 5*t,]
    if(length(train_motifs) < 1){
      next()
    }
    test_motifs <- CV_motif[i]
    test_alignment <- mod_alignment[i,]
    #predict mutant
    pred_motifs <- list()
    pred_motifs <- suppressWarnings(SRpredPSAM(test_alignment, train_motifs, train_alignment, weightfile = wf))
    all_pred_motifs[length(all_pred_motifs)+1] <- pred_motifs[1]
  }
  R2s <- c()
  for(pos in motifPoses){
    predMatrix <- frequency2ddG(gene2pos(all_pred_motifs, pos))
    testMatrix <- frequency2ddG(gene2pos(CV_motif, pos))
    predTrue <- data.frame(true = as.numeric(testMatrix), pred = as.numeric(predMatrix))
    #plotPredTrue(predTrue, main = paste0('FamilyCode prediction for cluster ',i,' at pos ', pos), xlab = 'Experimental -ΔΔG/RT', ylab = 'Predicted -ΔΔG/RT')
    R2s <- c(R2s, groupedR2(predTrue$true, predTrue$pred, member = nrow(predTrue)))
  }
  R2Cluster <- rbind.data.frame(R2Cluster, R2s)
  
  colnames(R2Cluster) <- motifPoses
  meanR2 <- apply(R2Cluster, 2, mean)
  SRR2 <- rbind.data.frame(SRR2, meanR2)
  print(t)
}
colnames(SRR2) <- motifPoses
```

```{r}
HDR2s <- list(FCR2, CSR2, SRR2)
saveRDS(HDR2s, 'HD_threshold_R2s.rds')


pos <- 'N8'
step = 8
plot(0:step*5,HDR2s[[1]][,pos],  type = 'l', ylim = c(min(min(HDR2s[[1]]), min(HDR2s[[2]], min(HDR2s[[3]]))),max(max(HDR2s[[1]]), max(HDR2s[[2]], max(HDR2s[[3]])))), col = '#ff7f0e', ylab = 'mean R-square', xlab = 'hamming distance cutoff', main = paste0('predictions for HD motif position ',pos) , lwd = 3)
lines(0:step*5,HDR2s[[2]][,pos], col = 'skyblue3', lwd = 3)
lines(0:step*5,HDR2s[[3]][,pos], col = 'blue3', lwd = 3)
#legend("topright", legend = c("Family Code", "Sequence Homology", "Similarity Regression"),col = c('#ff7f0e', 'skyblue3', 'blue3'), lty = 1, cex = 0.8, bty = 'n', lwd = 3)
```
