---
title: "Family Code Procedure"
output: html_notebook
---

## Family-level Association Analysis 
### Instal family code package
```{r}
library(FamilyCode)
```

### Input data sets
1: Probound_runDir: Directory containing the ProBound results each in a folder.   
2: bHLH_info: CSV table containing experiment information.
3: bHLH_ali: Alignment file 
```{r}
proBound_runDIR <- 'E:\\Desktop\\HB_Rotation_1\\ProBound_run\\results_for_publication\\'
bHLH_info <- read.csv(file = "E:/Desktop/HB_Rotation_1/Base_Shape_rec/bHLH_info.csv", row.names = 1)
bHLH_ali <- read.table("E:/Desktop/HB_Rotation_1/Base_Shape_rec/bHLH_pb.sto.stockholm", quote="\"", fill = T)
```

### Preparing variables for subsequent processes
```{r}
proBound_files <- list.files(proBound_runDIR)
modelFile_Template <- paste0(proBound_runDIR, '$modelFile$/result/fit.models.consensus.json')
rows <- c("A", "C", "G", "T")
blank <- data.frame(P00 = c(1,1,1,1), P0 = c(1,1,1,1))

#getting expeirment identifier from bHLH_info, may vary according to different naming scheme
bHLH_index <- bHLH_info[,1:2]
```

### Entering quality measurements for each model
This step can be ignored if no quality measurements are performed, or when each sample protein is only appearing once.
```{r}
bHLH_index$quality <- 0
for(i in 1:nrow(bHLH_index)){
  tryCatch({
    dir <- paste0(proBound_runDIR,bHLH_index$gene_symbol[i],'_',bHLH_index$study[i])
    suppressWarnings(quality <- readChar(paste0(dir,'//fitEval.txt'),nchars = 100))
    quality <- strsplit(quality, ', ')[[1]][1]
    quality <- strsplit(quality, ' = ')[[1]][2]
    bHLH_index$quality[i] <- quality
  }, error = function(e){})
}
```

### Reading motif model data from ProBound results
Reads in all data of bHLH samples and compares the consensus sequence recognition score between the binding models output from ProBound. For bHLH, a logical value is given for symmetry. The motif with the highest consensus sequence recognition score is saved in the list of mono_motifs. The mono_motifs contains binding motifs in frequency format. 
```{r}
Motif_Table <- loadMono_motifs(bHLH_index, modelFile_Template, rec_seq = 'CANNTG', pos_index = c('P-3','P-2','P-1','P1','P2','P3'), checkSymmetry = T, withTable = T)

mono_motifs <- Motif_Table$motifs
bHLH_model_info <- Motif_Table$table
```

### Loading the Alignment
Input Alignment file generated with Clustal Omega and select the aligned DNA-binding Domain of the family. 
```{r}
bHLH_pbAlignment <- concatAli(bHLH_ali, 694,56)
```
For a different sample set, likely a different family, You can start from this point with properly structured list of mono_motifs and data frame of Alignment  

### Get baseline experimental accuracy 
Obtaining baseline experimental accuracy with pair-wise comparison between input models of the same sample. Only do this when there are multiple instances of multiple models from different studies performed on the same protein sample.

```{r}
BaseLines <- getBaseLineAccuracy(mono_motifs, posMotif = 'P1', randomSample = F)

plot(BaseLines$HT1,BaseLines$HT2, pch = 19, xlab = 'Experiment 1 ddG', ylab = 'Experiment 2 ddG', main = 'Pair-wise experiment comparison', col = c('green','blue','orange','red'))
legend( x = "bottomright", legend = c('A', 'C', 'G', 'T'), col =c('green','blue','orange','red'), pch = 19, bty='n')

lm <- lm(HT2~HT1+0, BaseLines)
print(paste0('R^2 = ', summary(lm)$r.squared))
print(paste0('RMSD = ', RMSD(BaseLines$HT2,BaseLines$HT1)))
```

### Filtering and reloading for saticfactory data samples
Family code uses the bHLH_index data frame to load models. Filter rows according to recorded parameters and reload mono_motifs. Here, the data frame: bHLH_model_info has contains parameters and the first two rows are in the same format as the bHLH_index data frame. 
```{r}
#filter for symmetry
bHLH_sym <- bHLH_model_info[bHLH_model_info$symmetry == 1,]
#filter for best model of each protein
bHLH_names <- unique(bHLH_sym$gene_symbol)
bHLH_index <- data.frame(NULL)
for(i in 1:length(bHLH_names)){
  qualities <- bHLH_sym[bHLH_sym$gene_symbol == bHLH_names[i],]
  add <- qualities[qualities$quality == max(qualities$quality),]
  bHLH_index <- rbind.data.frame(bHLH_index, add[1,])
}
#filter for quality > 0.15
bHLH_index <- bHLH_index[bHLH_index$quality > 0.15,]

#reload mono_motifs
mono_motifs <- loadMono_motifs(bHLH_index, modelFile_Template, rec_seq = 'CANNTG', pos_index = c('P-3','P-2','P-1','P1','P2','P3'), checkSymmetry = F, withTable = F)
```

### Data presentation in tetrahedron system
Plot a given position (pos = 'position of your choice') of all samples on the tetrahedron representation system.  
```{r}
posMatix <- gene2pos(mono_motifs, pos = 'P3')
colnames(posMatix) <- unlist(lapply(strsplit(colnames(posMatix), '_'), function(x) x[1]))
plot_tetrahedron(posMatix, label = T)
```

### Generate MANOVA p-value table
Generate the p-value table of MANOVA between each amino acid residue-motif position combination. The amino acid residue type is the independent variable and the tetrahedron coordinates representing the binding motif at a given position is the dependent variable. 
```{r}
pvalTable <- getPvalTable(mono_motifs, Alignment = bHLH_pbAlignment, pos_index = c('P-3','P-2','P-1','P1','P2','P3'))
#apply -log(10)
pvalTable <- -log(pvalTable,10)
#set NAs to 0
pvalTable[is.na(pvalTable)] <- 0
#plot heatmap
gplots::heatmap.2(as.matrix(as.data.frame(lapply(pvalTable, as.numeric))),dendrogram='none',
                  Rowv=FALSE, Colv=FALSE,trace='none',col = rev(heat.colors(12)), key.title = 'a')
```

### Visualize associated positions in tetrahedron
To visualize the association between residue type and base preference, you can select a position along the protein alignment and a position in the binding motif to examine the distribution.
```{r}
AAPosition <- 53
motifPosition <- 'P-1'
bHLH_pbAlignment <- matchAliMotif(mono_motifs, bHLH_pbAlignment)
AAbp_Matrix <- AAbpCombination(mono_motifs, bHLH_pbAlignment, AAPosition, motifPosition)
plot_tetrahedron(AAbp_Matrix, color = T)
```

## Single/Double Mutant Prediction
