---
title: "Family Code Procedure"
output: html_notebook
---

## Family-level Association Analysis

### Instal family code package

```{r}
devtools::install_github('ShaoxunLiu/FCpackage')
library(FamilyCode)
```

### Input data sets

1: Probound_runDir: Directory containing the ProBound results each in a folder.\
2: bHLH_info: CSV table containing experiment information. 3: bHLH_ali: Alignment file

```{r}
proBound_runDIR <- r'(HTdata/)'
bHLH_info <- read.csv(file = r"(bHLH_info.csv)", row.names = 1)
bHLH_ali <- read.table(r"(bHLH_pb.sto.stockholm)", quote="\"", fill = T)
```

### Preparing variables for subsequent processes

```{r}
proBound_files <- list.files(proBound_runDIR)
modelFile_Template <- paste0(proBound_runDIR, '$modelFile$/result/fit.models.consensus.json')
rows <- c("A", "C", "G", "T")
blank <- data.frame(P00 = c(1,1,1,1), P0 = c(1,1,1,1))

#getting expeirment identifier from bHLH_info, may vary according to different naming scheme
bHLH_index <- bHLH_info[,1:2]
```

### Entering quality measurements for each model

This step can be ignored if no quality measurements are performed, or when each sample protein is only appearing once.

```{r}
bHLH_index$quality <- 0
for(i in 1:nrow(bHLH_index)){
  tryCatch({
    dir <- paste0(proBound_runDIR,bHLH_index$gene_symbol[i],'_',bHLH_index$study[i])
    suppressWarnings(quality <- readChar(paste0(dir,'//fitEval.txt'),nchars = 100))
    quality <- strsplit(quality, ', ')[[1]][1]
    quality <- strsplit(quality, ' = ')[[1]][2]
    bHLH_index$quality[i] <- quality
  }, error = function(e){})
}
```

### Reading motif model data from ProBound results

Reads in all data of bHLH samples and compares the consensus sequence recognition score between the binding models output from ProBound. For bHLH, a logical value is given for symmetry. The motif with the highest consensus sequence recognition score is saved in the list of mono_motifs. The mono_motifs contains binding motifs in frequency format.

```{r}
Motif_Table <- loadMono_motifs(bHLH_index, modelFile_Template, rec_seq = 'CANNTG', pos_index = c('P-3','P-2','P-1','P1','P2','P3'), checkSymmetry = T, withTable = T)

all_mono_motifs <- Motif_Table$motifs
bHLH_model_info <- Motif_Table$table
```

### Loading the Alignment

Input Alignment file generated with Clustal Omega and select the aligned DNA-binding Domain of the family.

```{r}
bHLH_pbAlignment <- concatAli(bHLH_ali, 694,56)
```

For a different sample set, likely a different family, You can start from this point with properly structured list of mono_motifs and data frame of Alignment

### Get baseline experimental accuracy

Obtaining baseline experimental accuracy with pair-wise comparison between input models of the same sample. Only do this when there are multiple instances of multiple models from different studies performed on the same protein sample.

The accuracy of prediction models are measured in grouped R^2 and RMSD. Grouped R^2 reports the mean of R^2 values of regressions between a true and a predicted motif in ddG measurement. RMSD reports the pair-wise Root Mean Squared Difference
```{r}
BaseLines <- getBaseLineAccuracy(all_mono_motifs, pos = 'P-1', randomSample = F)

plot(BaseLines$HT1,BaseLines$HT2, pch = 19, xlab = 'Experiment 1 ddG', ylab = 'Experiment 2 ddG', main = 'Pair-wise experiment comparison', col = c('green','blue','orange','red'))
legend( x = "bottomright", legend = c('A', 'C', 'G', 'T'), col =c('green','blue','orange','red'), pch = 19, bty='n')

print(paste0('Grouped R^2 = ', groupedR2(BaseLines$HT1, BaseLines$HT2)))
print(paste0('RMSD = ', RMSD(BaseLines$HT2,BaseLines$HT1)))
```

### Filtering and reloading for saticfactory data samples

Family code uses the bHLH_index data frame to load models. Filter rows according to recorded parameters and reload mono_motifs. Here, the data frame: bHLH_model_info has contains parameters and the first two rows are in the same format as the bHLH_index data frame.

```{r}
#filter for symmetry
bHLH_sym <- bHLH_model_info[bHLH_model_info$symmetry == 1,]
#filter for best model of each protein
bHLH_names <- unique(bHLH_sym$gene_symbol)
bHLH_index <- data.frame(NULL)
for(i in 1:length(bHLH_names)){
  qualities <- bHLH_sym[bHLH_sym$gene_symbol == bHLH_names[i],]
  add <- qualities[qualities$quality == max(qualities$quality),]
  bHLH_index <- rbind.data.frame(bHLH_index, add[1,])
}
#filter for quality > 0.15
bHLH_index <- bHLH_index[bHLH_index$quality > 0.15,]


#reload mono_motifs
mono_motifs <- loadMono_motifs(bHLH_index, modelFile_Template, rec_seq = 'CANNTG', pos_index = c('P-3','P-2','P-1','P1','P2','P3'), checkSymmetry = F, withTable = F)
```

### Data presentation in tetrahedron system

Plot a given position (pos = 'position of your choice') of all samples on the tetrahedron representation system.

```{r}
posMatix <- gene2pos(mono_motifs, pos = 'P3')
colnames(posMatix) <- unlist(lapply(strsplit(colnames(posMatix), '_'), function(x) x[1]))
plot_tetrahedron(posMatix, label = F, color = F)
```

### Generate MANOVA p-value table

Generate the p-value table of MANOVA between each amino acid residue-motif position combination. The amino acid residue type is the independent variable and the tetrahedron coordinates representing the binding motif at a given position is the dependent variable.

```{r}
pvalTable <- getPvalTable(mono_motifs, Alignment = bHLH_pbAlignment, pos_index = c('P-3','P-2','P-1','P1','P2','P3'))
#apply -log(10)
pvalTable <- -log(pvalTable,10)
#set NAs to 0
pvalTable[is.na(pvalTable)] <- 0
#plot heatmap
gplots::heatmap.2(as.matrix(as.data.frame(lapply(pvalTable, as.numeric))),dendrogram='none',
                  Rowv=FALSE, Colv=FALSE,trace='none',col = rev(heat.colors(12)), key.title = '-logP-val')
```

### Visualize associated positions in tetrahedron

To visualize the association between residue type and base preference, you can select a position along the protein alignment and a position in the binding motif to examine the distribution.

```{r}
AAPosition <- 13
motifPosition <- 'P-1'
bHLH_pbAlignment <- matchAliMotif(mono_motifs, bHLH_pbAlignment)
AAbp_Matrix <- AAbpCombination(mono_motifs, bHLH_pbAlignment, AAPosition, motifPosition)
plot_tetrahedron(AAbp_Matrix, color = T)
```

## Single/Double Mutant Prediction

Predicting the binding motif of ASCL2 R5K, V13R, R5K+V13R, and HES2 K5R, R13V, K5R+R13V at the P-1 position.

### Load binding matrix of HES2 and ASCL2 from SELEX experiments

```{r}
groups <- c('HWT', 'H5', 'H13','HD', 'AWT', 'A5', 'A13', 'AD')
SELEXFile_Template <- r'(SELEXdata/$modelFile$/fit.models.consensus.json)'
SELEX_info <- data.frame(gene_symbol = groups, study = 'pb')
SELEX_motifs <- loadMono_motifs(SELEX_info, SELEXFile_Template, rec_seq = 'NCANNTGN', pos_index = c('P-4','P-3','P-2','P-1','P1','P2','P3', 'P4'), checkSymmetry = T, withTable = F, useMode = c(1,2,1,1,1,1,1,1))
#Create posMatrix at P-1
posMatix <- gene2pos(SELEX_motifs, pos = 'P-1')
colnames(posMatix) <- unlist(lapply(strsplit(colnames(posMatix), '_'), function(x) x[1]))
#visualize results in tetrahedron for position P-1
plot_tetrahedron(posMatix, label = T, color = F)
```

### Examine base-line accuracy
Examine base-line accuracy between SELEX experiment of this study and HT-SELEX experiment in database
```{r}
WTdf <- data.frame(HT = c(frequency2ddG(getMatrix(mono_motifs, 'HES2_Yin2017')$matrix)[,'P-1'], 
                          frequency2ddG(getMatrix(mono_motifs, 'ASCL2_Yin2017')$matrix)[,'P-1']), 
                   Exp = c(frequency2ddG(getMatrix(SELEX_motifs, 'HWT_pb')$matrix)[,'P-1'], 
                          frequency2ddG(getMatrix(SELEX_motifs, 'AWT_pb')$matrix)[,'P-1']))
#plotting HT-SELEX vs. experiment
plot(WTdf$Exp, WTdf$HT, col = c(1,1,1,1,2,2,2,2,1,1,1,1,2,2,2,2), pch = 19, xlab = 'Experiment ddG', ylab = 'High-troughput SELEX ddG', main = 'WT vs. WT')
legend( x = "bottomright", legend = c('HES2', 'ASCL2'), col =c(1:2), pch = 19, bty='n')
print(paste0('Grouped R^2 = ', groupedR2(WTdf$Exp, WTdf$HT)))
print(paste0('RMSD = ', RMSD(WTdf$HT,WTdf$Exp)))
```

### Check descrepency between WT and mutant sampels
Examine how much the WT results explains the mutants. This regression R^2 shoudl be low. 
```{r}
#Create Mutdf
Mutdf <- data.frame(Exp = c(frequency2ddG(getMatrix(SELEX_motifs, 'H5_pb')$matrix)[,'P-1'], 
                          frequency2ddG(getMatrix(SELEX_motifs, 'H13_pb')$matrix)[,'P-1'],
                          frequency2ddG(getMatrix(SELEX_motifs, 'HD_pb')$matrix)[,'P-1'],
                          frequency2ddG(getMatrix(SELEX_motifs, 'A5_pb')$matrix)[,'P-1'],
                          frequency2ddG(getMatrix(SELEX_motifs, 'A13_pb')$matrix)[,'P-1'],
                          frequency2ddG(getMatrix(SELEX_motifs, 'AD_pb')$matrix)[,'P-1']), 
                   Pred = c(frequency2ddG(getMatrix(SELEX_motifs, 'HWT_pb')$matrix)[,'P-1'],
                            frequency2ddG(getMatrix(SELEX_motifs, 'HWT_pb')$matrix)[,'P-1'],
                            frequency2ddG(getMatrix(SELEX_motifs, 'HWT_pb')$matrix)[,'P-1'],
                          frequency2ddG(getMatrix(SELEX_motifs, 'AWT_pb')$matrix)[,'P-1'],
                          frequency2ddG(getMatrix(SELEX_motifs, 'AWT_pb')$matrix)[,'P-1'],
                          frequency2ddG(getMatrix(SELEX_motifs, 'AWT_pb')$matrix)[,'P-1']))
#plot Predicted vs. Experimental
plot(Mutdf$Exp, Mutdf$Pred,  col = c(rep(1, nrow(Mutdf)/2), rep(2,nrow(Mutdf)/2)), pch = 19, xlab = 'Experiment ddG', ylab = 'Predicted ddG', main = 'Mutants vs. Mutants')
legend( x = "bottomright", legend = c('HES2', 'ASCL2'), col =c(1:2), pch = 19, bty='n')
print(paste0('Grouped R^2 = ', groupedR2(Mutdf$Exp, Mutdf$Pred)))
print(paste0('RMSD = ', RMSD(Mutdf$Pred,Mutdf$Exp)))
```

### Predict binding matrix at P-1
```{r}
#create dddGList with database
dddGList <- form.dddGList(bHLH_pbAlignment, mono_motifs, keyPos = c(5,13), posMotif = 'P-1')
#make prediction for HES2 and ASCL2 mutants
H5pred <- predict(dddGList, referenceMatrix = log(posMatix[,'HWT']), referenceSequence = '----K-------R', targetSequence = '----R-------R', keyPos = c(5,13), posMotif = 'P-1')
H13pred <- predict(dddGList, referenceMatrix = log(posMatix[,'HWT']), referenceSequence = '----K-------R', targetSequence = '----K-------V', keyPos = c(5,13), posMotif = 'P-1')
HDpred <- predict(dddGList, referenceMatrix = log(posMatix[,'HWT']), referenceSequence = '----K-------R', targetSequence = '----R-------V', keyPos = c(5,13), posMotif = 'P-1')
A5pred <- predict(dddGList, referenceMatrix = log(posMatix[,'AWT']), referenceSequence = '----R-------V', targetSequence = '----K-------V', keyPos = c(5,13), posMotif = 'P-1')
A13pred <- predict(dddGList, referenceMatrix = log(posMatix[,'AWT']), referenceSequence = '----R-------V', targetSequence = '----R-------R', keyPos = c(5,13), posMotif = 'P-1')
ADpred <- predict(dddGList, referenceMatrix = log(posMatix[,'AWT']), referenceSequence = '----R-------V', targetSequence = '----K-------R', keyPos = c(5,13), posMotif = 'P-1')
PredDt <- data.frame(H5pred, H13pred, HDpred, A5pred, A13pred, ADpred)
colnames(PredDt) <- c('H5', 'H13', 'HD', 'A5', 'A13', 'AD')
```

### Checking prediction accuracy with experimental values of HES2 and ASCL2 mutants
Examine how well the predicted ddG motif matrices explain the Experimental motifs.
```{r}
Mutdf <- data.frame(Exp = c(frequency2ddG(getMatrix(SELEX_motifs, 'H5_pb')$matrix)[,'P-1'], 
                          frequency2ddG(getMatrix(SELEX_motifs, 'H13_pb')$matrix)[,'P-1'],
                          frequency2ddG(getMatrix(SELEX_motifs, 'HD_pb')$matrix)[,'P-1'],
                          frequency2ddG(getMatrix(SELEX_motifs, 'A5_pb')$matrix)[,'P-1'],
                          frequency2ddG(getMatrix(SELEX_motifs, 'A13_pb')$matrix)[,'P-1'],
                          frequency2ddG(getMatrix(SELEX_motifs, 'AD_pb')$matrix)[,'P-1']), 
                   Pred = unlist(PredDt))
#plot Predicted vs. Experimental
plot(Mutdf$Exp, Mutdf$Pred,  col = c(rep(1, nrow(Mutdf)/2), rep(2,nrow(Mutdf)/2)), pch = 19, xlab = 'Experiment ddG', ylab = 'Predicted ddG', main = 'Mutants vs. Mutants')
legend( x = "bottomright", legend = c('HES2', 'ASCL2'), col =c(1:2), pch = 19, bty='n')
print(paste0('Grouped R^2 = ', groupedR2(Mutdf$Exp, Mutdf$Pred)))
print(paste0('RMSD = ', RMSD(Mutdf$Pred,Mutdf$Exp)))
```


## Full sequence prediction

### Generate SVD matrices
Screen SVD results with the whole training set
```{r}
posMatrix <- gene2pos(mono_motifs, pos = 'P-1')
bHLH_pbAlignment <- matchAliMotif(mono_motifs, bHLH_pbAlignment)
svd <- matrixSVD(posMatrix)
#ANOVA test on the level of association between SVD Principal components and residue type
svd.pval <- svdANOVA(svd, bHLH_pbAlignment)
#Plot p-vals 
plot(svd.pval$X1, pch = 1, col = "red", type = 'l', lwd = 1.5, xlab = 'AA Position', ylab = '-log p-val', main = 'Principle Componant-AA position Association')
lines(svd.pval$X2, pch = 1, col = "green", lwd = 1.5)
lines(svd.pval$X3, pch = 1, col = "blue", lwd = 1.5)
legend(l, x = "topright", legend = c('PC1', 'PC2', 'PC3'), lwd = 1.5, col =c('red','green','blue'), lty=1, cex = 0.8, bty = 'n')

```

###Build SVD-regression model
```{r}
#separate train and test
t <- c(20)
train_motifs <- mono_motifs[-t]
test_motifs <- mono_motifs[t]
train_alignment <- matchAliMotif(train_motifs, bHLH_pbAlignment)
test_alignment <- matchAliMotif(test_motifs, bHLH_pbAlignment)
#Train svd-regression model
svd <- matrixSVD(gene2pos(train_motifs, pos = 'P-1'))
svdModel <- trainSVD(svd, train_alignment)
#Predict binding motifs for test set
pred_motifs <- predict(svdModel, test_alignment, zero = 0.001)
#Comparing between true and predicted testing set motifs 
true <- frequency2ddG(gene2pos(test_motifs, pos = 'P-1'))
pred <- frequency2ddG(pred_motifs)
predTrue <- data.frame(true = unlist(as.numeric(true)), pred = unlist(as.numeric(pred)))

#Plot results
plot(true,pred, pch = 19, xlab = 'Experimental ddG', ylab = 'Predicted ddG', main = 'SVD prediction', col = c('green','blue','orange','red'))
legend( x = "bottomright", legend = c('A', 'C', 'G', 'T'), col =c('green','blue','orange','red'), pch = 19, bty='n')
print(paste0('Grouped R^2 = ', groupedR2(predTrue$pred, predTrue$true)))
print(paste0('RMSD = ', RMSD(predTrue$pred,predTrue$true)))
```
## leave-one-out test for all availiable samples
Perform leave-one-out testing for accuracy with all available training samples.
```{r}
#perform training and testing
predTrue <- data.frame(NULL)
for(t in 1:length(mono_motifs)){
  train_motifs <- mono_motifs[-t]
  test_motifs <- mono_motifs[t]
  train_alignment <- matchAliMotif(train_motifs, bHLH_pbAlignment)
  test_alignment <- matchAliMotif(test_motifs, bHLH_pbAlignment)
  #Train svd-regression model
  svd <- matrixSVD(gene2pos(train_motifs, pos = 'P-1'))
  svdModel <- trainSVD(svd, train_alignment)
  #Predict binding motifs for test set
  pred_motifs <- predict(svdModel, test_alignment, zero = 0.01)
  #Comparing between true and predicted testing set motifs 
  true <- frequency2ddG(gene2pos(test_motifs, pos = 'P-1'))
  pred <- frequency2ddG(pred_motifs)
  add <- data.frame(true = unlist(as.numeric(true)), pred = unlist(as.numeric(pred)))
  predTrue <- rbind.data.frame(predTrue, add)
}

#plot results
plot(predTrue$true,predTrue$pred, pch = 19, xlab = 'Experimental ddG', ylab = 'Predicted ddG', main = 'SVD prediction', col = c('green','blue','orange','red'))
legend( x = "bottomright", legend = c('A', 'C', 'G', 'T'), col =c('green','blue','orange','red'), pch = 19, bty='n')
print(paste0('Grouped R^2 = ', groupedR2(predTrue$pred, predTrue$true)))
print(paste0('RMSD = ', RMSD(predTrue$pred,predTrue$true)))
```

## Compair with closest-sequence prediction
Comparing the result of SVD-regression with the prediction accuracy with the closes-sequence prediction methods (Weirauch, 2014).
```{r}
CSpred <- closestSeqPred(mono_motifs, bHLH_pbAlignment, pos = 'P-1')
#plot results
plot(CSpred$true,CSpred$pred, pch = 19, xlab = 'Experimental ddG', ylab = 'Predicted ddG', main = 'SVD prediction', col = c('green','blue','orange','red'))
legend( x = "bottomright", legend = c('A', 'C', 'G', 'T'), col =c('green','blue','orange','red'), pch = 19, bty='n')
print(paste0('Grouped R^2 = ', groupedR2(CSpred$pred, CSpred$true)))
print(paste0('RMSD = ', RMSD(CSpred$pred,CSpred$true)))
```
## SVD prediction for single and double mutants
Applying the SVD-regression model to predict the binding motifs of the single and double mutants of HES2 and ASCL2. The key Position used are the mutated positions, and the number of positions to use for each PC is infered from the line plot of Principle Componant-AA position Association.
```{r}
train_motifs <- mono_motifs
test_motifs <- SELEX_motifs[-c(1,5)]
train_alignment <- matchAliMotif(train_motifs, bHLH_pbAlignment)
test_alignment <- data.frame(name = paste0(groups, '_pb')[-c(1,5)], alignment =c(
                             '-KSLRPLLEKRRRARINQSLSQLKGLI-L------PLLGRENS--NCSKLEKADVL',
                             '-KSLKPLLEKRRVARINQSLSQLKGLI-L------PLLGRENS--NCSKLEKADVL',
                             '-KSLRPLLEKRRVARINQSLSQLKGLI-L------PLLGRENS--NCSKLEKADVL',
                             '-VARKNERERNRVKLVNLGFQALRQHV-P------HGG-------ASKKLSKVETL',
                             '-VARRNERERNRRKLVNLGFQALRQHV-P------HGG-------ASKKLSKVETL',
                             '-VARKNERERNRRKLVNLGFQALRQHV-P------HGG-------ASKKLSKVETL')
                             )
#Train svd-regression model
svd <- matrixSVD(gene2pos(train_motifs, pos = 'P-1'))
svdModel <- trainSVD(svd, train_alignment, no.keyPos = c(2,1,0), keyPos = list(c(13,5), c(13,13), c(13,13)))
#Predict binding motifs for test set
pred_motifs <- predict(svdModel, test_alignment, zero = 0.001)
#Comparing between true and predicted testing set motifs 
true <- frequency2ddG(gene2pos(test_motifs, pos = 'P-1'))
pred <- frequency2ddG(pred_motifs)
predTrue <- data.frame(true = unlist(as.numeric(true)), pred = unlist(as.numeric(pred)))

#Plot results
plot(true,pred, pch = 19, xlab = 'Experimental ddG', ylab = 'Predicted ddG', main = 'SVD prediction on mutants', col = c('green','blue','orange','red'))
legend( x = "bottomright", legend = c('A', 'C', 'G', 'T'), col =c('green','blue','orange','red'), pch = 19, bty='n')
print(paste0('Grouped R^2 = ', groupedR2(predTrue$pred, predTrue$true)))
print(paste0('RMSD = ', RMSD(predTrue$pred,predTrue$true)))
```



## AA property association with PC values
Screening the association between the biophysical properties of amino acids and PC values from the SVD tranformation. 
```{r}
map <- aaPCmap(mono_motifs, bHLH_pbAlignment, pos = 'P-1', AApos = 13, PC = 2, property = 2)
print(map)
```

