---
title: "Process Kock et al., PBM data from intensity"
output:
  html_document:
    df_print: paged
---
CV_alignment and CV_motif from FamilyCodeOnSELEX+PBM_HD.Rmd after line 239

### Import Intensity matrix
```{r}
suppressMessages(library(readxl))
suppressMessages(library(FamilyCode))
options(warn = -1)
workDir <- 'KockHDPBM/'
KockMatrix <- readRDS(paste0(workDir,'IntensityMatrix.rds'))
col <- gsub('NKX2-','NKX2.',colnames(KockMatrix[,3:ncol(KockMatrix)]))
col <- gsub('NKX3-','NKX3.',col)
CV_alignment <- readRDS('HD_SELEX_PBM_CV_alignment.rsd')
CV_motif <- readRDS('HD_SELEX_PBM_CV_motif.rsd')
```

### loading missing sequence and motif place holders
```{r}
options(warn = -1)
ali <- read.table(paste0(workDir,"add_hmm.sto"), quote="\"", fill = T)
ali <- ali[-nrow(ali),]
ali$V1 <- apply(ali, 1, function(x) strsplit(x[1],'/')[[1]][1])
ali$V1 <- apply(ali, 1, function(x) strsplit(x[1],'\\|')[[1]][2])
colnames(ali) <- c('name','alignment')
ali$name <- c('HOXB6', 'NKX2-6', 'NKX2-4')
CV_alignment <- rbind.data.frame(CV_alignment, ali)

CV_motif[[415]] <- list(name = 'HOXB6', matrix = CV_motif[[414]]$matrix)
CV_motif[[416]] <- list(name = 'NKX2-6', matrix = CV_motif[[414]]$matrix)
CV_motif[[417]] <- list(name = 'NKX2-4', matrix = CV_motif[[414]]$matrix)
```

### get mutation data 
```{r}
SampleInfo <- read_excel("KockHDPBM/41467_2024_47396_MOESM5_ESM.xls")
ExpIDs <- unique(unlist(lapply(strsplit(col,'_'), function(x) x[1])))
mutationInfo <- data.frame(NULL)
for(i in 1:length(ExpIDs)){
  id <- ExpIDs[i]
  info <- strsplit(id,'-')[[1]]
  if(info[2] == 'REF'){
    addLine <- data.frame(TF = info[1], mutation = 'WT', position = NA, mutateFrom = 'WT')
  }else{
    id <- gsub('\\.','-',id)
    pos <- SampleInfo[SampleInfo$allele == id,]$canonical
    addLine <- data.frame(TF = info[1], mutation = substr(info[2],nchar(info[2]),nchar(info[2])), position = pos, mutateFrom = substr(info[2],1,1))
  }
  mutationInfo <- rbind.data.frame(mutationInfo, addLine)
}

#validate with CV_Alignment
valid <- c()
mutationInfo$TF <- gsub('\\.','-',mutationInfo$TF)
for(i in 1:nrow(mutationInfo)){
  seq <- CV_alignment[CV_alignment$name == mutationInfo$TF[i],]$alignment
  if(mutationInfo$mutation[i] == 'WT' ){
    valid <- c(valid, 'WT')
  }else if(length(seq) == 0){
    valid <- c(valid, 'notInAlignment')
  }else{
    match <- substr(seq[1], mutationInfo$position[i]-1,mutationInfo$position[i]-1) == mutationInfo$mutateFrom[i]
    valid <- c(valid, match)
  }
}
mutationInfo$valid <- valid
```

### name Kock HD alignment
```{r}
alignment <- c()
for(i in 1:nrow(mutationInfo)){
  seq <- CV_alignment[CV_alignment$name == mutationInfo$TF[i],]$alignment
  if(length(seq) == 0){
    alignment <- c(alignment, NA)
  }else if(mutationInfo$mutation[i] == 'WT' ){
    alignment <- c(alignment, seq[1])
  }else{
    #match <- substr(seq[1], mutationInfo$position[i]-1,mutationInfo$position[i]-1) == mutationInfo$mutateFrom[i]
    #if(match){
    substr(seq[1], mutationInfo$position[i]-1,  mutationInfo$position[i]-1) <-  mutationInfo$mutation[i]
    alignment <- c(alignment, seq[1])
    #}else{
    #  alignment <- c(alignment, NA)
    #}
  }
}
mutationInfo$alignment <- alignment
mutationInfo$name <- gsub('\\.','-',ExpIDs)
```


### Write count tables
```{r}
suppressMessages(library(R.utils))
mkdirs(paste0(workDir,'CountTables_Intensity'))
for(i in 1:(ncol(KockMatrix)-2)){
  CT <- data.frame(name = KockMatrix$seq, score = KockMatrix[,i+2])
  rownames(CT) <- KockMatrix$seq
  colnames(CT) <- c('','0')
  CT <- CT[!is.na(CT$`0`),]
  CT <- CT[CT$`0` > 0,]
  fname <- colnames(KockMatrix)[i+2]
  
  write.table(CT, file =paste0(workDir, 'CountTables_Intensity/', fname, '_CT.tsv'), quote = F, dec = '.', row.names = F, col.names = T, sep = '\t')
}
```
### Loading pwms from pyPB
```{r}
PWMsID <- colnames(KockMatrix)[3:ncol(KockMatrix)]
Kock_motifs <- list()
options(warn=-1)
for(i in 1:length(PWMsID)){
  motif <- PWMsID[i]
  motif <- read.table(paste0(workDir, 'pwmsI/', motif, '.txt'), header = F)
  motif <- ddG2frequency(motif)
  motif[motif < 0.001] <- 0.01#add psuedo count
  rownames(motif) <- DNA()
  add <- list()
  add$gene_symbol <- PWMsID[i]
  add$study <- i
  add$mode <- 1
  add$matrix <- as.matrix(motif)
  Kock_motifs[[length(Kock_motifs)+1]] <- add
  add <- list()
  #add reverse compliment motif
  add$gene_symbol <- PWMsID[i]
  add$study <- i
  add$mode <- -1
  if(ncol(motif) == 0){
    rev <- motif
  }else{
    rev <- as.matrix(motif)[nrow(motif):1, ncol(motif):1]
  }
  rownames(rev) <- DNA()
  add$matrix <- rev
  Kock_motifs[[length(Kock_motifs)+1]] <- add
}


```
### Load Homeo Domain 8mer seed of NNTRAYNN and score motifs with the seed
```{r}
#align to NNTDAYNN motif
HDseed <- matrix(nrow = 4,ncol = 8,data = c(0,0,0,0,
                                            0,0,0,0,
                                            0,0,0,1,
                                            1,0,1,1,
                                            1,0,0,0,
                                            0,1,0,1,
                                            0,0,0,0,
                                            0,0,0,0))
rownames(HDseed) <- DNA()

motifScores <- scoreMotifList(Kock_motifs, HDseed, weight = c(0,0,1,1,1,1,0,0))
```

### Filter and reorder motif information according to motif matching score.
The motif (forward or reverse compliment) with the lowest motif matching score of each entry will be kept. For the same protein, entries are arranged from lower score to higher score
```{r}
idstds <- unique(motifScores[,1:2])
all_motifs <- data.frame(NULL)
for(i in 1:nrow(idstds)){
  fDT <- motifScores[motifScores$gene_symbol == idstds[i,1],]
  fDT <- fDT[fDT$study == idstds[i,2],]
  all_motifs <- rbind.data.frame(all_motifs, fDT[which.min(fDT$score),])
}

all_motifs$study <- unlist(lapply(strsplit(all_motifs$gene_symbol,'_'), function(x) x[2]))
all_motifs$gene_symbol <- unlist(lapply(strsplit(all_motifs$gene_symbol,'_'), function(x) x[1]))
motifs_arranged <- data.frame(NULL)
motifScore <- c()
for(i in 1:length(unique(all_motifs$gene_symbol))){
  add <- all_motifs[all_motifs$gene_symbol == unique(all_motifs$gene_symbol)[i],]
  add <- dplyr::arrange(add, score)
  motifScore <- c(motifScore,add$score[1])
  motifs_arranged <- rbind.data.frame(motifs_arranged, add)
}
#Plot distribution of lowest motif matching score
plot.ecdf(as.numeric(motifScore))
```

### Construct motif list and alignment table for Base Line (Motif similarity between the two best PBM experiments)
```{r}
motifPoses <- c('N1','N2','T3','D4','A5','Y6','N7','N8')

Kock_motifs_set <- filterMotifList(motifs_arranged, Kock_motifs, 8, motifPoses)
Kock_alignment <- data.frame(name = mutationInfo$name, alignment = mutationInfo$alignment)
Kock_alignment <- Kock_alignment[match(motifs_arranged$gene_symbol,Kock_alignment$name),]

#check if motif List has same length as alignment
length(Kock_motifs_set) == nrow(Kock_alignment)
unlist(lapply(Kock_motifs_set, function(x) x$name)) == Kock_alignment$name

```

### Find baseline Accuracy

```{r}
evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
for(pos in motifPoses){
  ScreenPoses <- c(pos,pos)
  BaseLines.ddG <- getBaseLineAccuracy(Kock_motifs_set, pos = ScreenPoses, randomSample = F)
  colnames(BaseLines.ddG) <- c('pred','true')
  BaseLines.PFM <- predTrue.ddG2frequency(BaseLines.ddG, PFM = T)
  plotPredTrue(BaseLines.ddG, xlab = 'Experiment 1', ylab = 'Experiment 2', main = paste0('Base Line -ΔΔG/RT at ',pos))
  boxplot(groupedR2(BaseLines.ddG$pred, BaseLines.ddG$true, member = length(ScreenPoses)*4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(BaseLines.ddG$pred, BaseLines.ddG$true, 
                                                          member = length(ScreenPoses)*4, mean= T),4)))
  plotPredTrue(BaseLines.PFM, xlab = 'Experiment 1', ylab = 'Experiment 2', main = paste0('Base Line PFM at ',pos))
  hit <- sum(round(groupedR2(BaseLines.PFM$pred, BaseLines.PFM$true, member = 4, mean= F, throughZero = F)^(1/2),4) > 0.5)/(nrow(BaseLines.PFM)/4)
  boxplot(groupedR2(BaseLines.PFM$pred, BaseLines.PFM$true, member = length(ScreenPoses)*4, mean= F, throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(BaseLines.PFM$pred, BaseLines.PFM$true, 
                                                     member = length(ScreenPoses)*4, mean= T, throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(BaseLines.ddG$pred, 
        BaseLines.ddG$true, nrow(BaseLines.ddG), throughZero = T), 
        4), round(RMSD(BaseLines.ddG$pred, BaseLines.ddG$true), 
        4), round(groupedR2(BaseLines.PFM$pred, 
        BaseLines.PFM$true, nrow(BaseLines.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
}
evaMatrix_BL <- evaMatrix
```

```{r}
mutantsExp2 <- mutationInfo$name
Exp2TrueMotifs <- list()
for(i in 1:length(mutantsExp2)){
  key <- which(Kock_alignment$name == mutantsExp2[i])[2]
  Exp2TrueMotifs[[i]] <- Kock_motifs_set[[key]]
}
```

### adding WT into training set
```{r}
del <- which(duplicated(Kock_alignment$name))
Kock_alignment <- Kock_alignment[-del,]
Kock_motifs_set <- Kock_motifs_set[-del]
#del <- which(is.na(mutationInfo$alignment))
#Kock_alignment <- Kock_alignment[-del,]
#mutationInfo <- mutationInfo[-del,]
#Kock_motifs_set <- Kock_motifs_set[-del]
mutationInfo$name == Kock_alignment$name
```


argument for NOT adding WT to training set would be that it induces bias towards nearest neighbor prediction

!!!adjustment made!!!
```{r}
CV_alignment_Kock <- CV_alignment[1:414,]
CV_motif_Kock <- CV_motif[1:414]

#for(i in 1:nrow(mutationInfo)){
#  if(mutationInfo$mutateFrom[i] == 'WT'){
#    id <- which(CV_alignment$name == mutationInfo$TF[i])
#    CV_motif_Kock[[id[1]]]$matrix <- Kock_motifs_set[[i]]$matrix
#  }
#}
```

### predict with FamilyCode overall model 
```{r}
svdModel <- makeSVDModel(CV_motif_Kock, CV_alignment_Kock, Positions = motifPoses)
predMotifs <- lapply(Kock_alignment$alignment, function(x) SVDmodelPredPSAM(svdModel,x))

mutantID <- which(mutationInfo$mutation != 'WT')
predMotifs <- predMotifs[mutantID]
trueMotifs <- Kock_motifs_set[mutantID]

evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
i = 1
for(pos in motifPoses){
  pred <- unlist(lapply(predMotifs, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(trueMotifs, function(x) frequency2ddG(x$matrix)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('FamilyCode at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  
  plotPredTrue(predTrue.PFM, main = paste0('FamilyCode PFM at ',pos), xlab = 'Expermental', ylab = 'Predicted')
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T,throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  i <- i + 1
}
evaMatrix_FC <- evaMatrix
```


### case by case model training
```{r}
mutantID <- which(mutationInfo$mutate != 'WT')

mutationPredList <- list()
mutationPredConfidence <- c()
for(i in mutantID){
  svdModel <- makeSVDModel.singleMutation(CV_motif_Kock, CV_alignment_Kock, Positions = motifPoses, mutation = as.numeric(mutationInfo$position[i])-1)
  predPSAM <- data.frame(pred = c(0,0,0,0))
  rownames(predPSAM) <- DNA()
  for(pos in motifPoses){
    predColumn <- predict(svdModel[[pos]], Kock_alignment[i,], useSimilarAA = T)
    predPSAM <- cbind.data.frame(predPSAM, predColumn)
    predConfidence <- attr(predColumn, 'confidence')
  }
  predPSAM <- predPSAM[,-1]
  colnames(predPSAM) <- motifPoses
  mutationPredList[[length(mutationPredList)+1]] <- predPSAM
  mutationPredConfidence <- c(mutationPredConfidence, predConfidence)
}

trueMotifs <- Kock_motifs_set[mutantID]
predMotifs <- mutationPredList

evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
i = 1
for(pos in motifPoses){
  pred <- unlist(lapply(predMotifs, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(trueMotifs, function(x) frequency2ddG(x$matrix)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('FamilyCode at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  
  plot(mutationPredConfidence,groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F), ylab = 'Predict-True R^2', xlab = 'Confidence score', main = paste0('Confidence at ', pos))
  legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
                                                     mutationPredConfidence, ),4)), bty = 'n')
  dt <- data.frame(conf = mutationPredConfidence, R2 = groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F))
  lm <- lm(R2~conf, data= dt)
  abline(lm, col= 'red')
  
  plotPredTrue(predTrue.PFM, main = paste0('FamilyCode PFM at ',pos), xlab = 'Expermental', ylab = 'Predicted')
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T,throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  i <- i + 1
}

evaMatrix_FC_byPosition <- evaMatrix
```

### Nearest neighbour / similarity regression 
```{r}
predMotifs <- list()
for(i in mutantID){
  predMotifs[[length(predMotifs)+1]] <- SRpredPSAM(mutationInfo[i,], CV_motif_Kock, CV_alignment_Kock,  weightfile = 'cisBP_HD/F223_1.97d.json')[[1]]$matrix
}


trueMotifs <- Kock_motifs_set[mutantID]

evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
i = 1
for(pos in motifPoses){
  pred <- unlist(lapply(predMotifs, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(trueMotifs, function(x) frequency2ddG(x$matrix)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('Closest sequence at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  
  plotPredTrue(predTrue.PFM, main = paste0('Closest sequence PFM at ',pos), xlab = 'Expermental', ylab = 'Predicted')
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T,throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  i <- i + 1
}
evaMatrix_CS <- evaMatrix
```

### predicting dddGs with FC
```{r}
mutantID <- which(mutationInfo$mutate != 'WT')
WTID <- which(mutationInfo$mutate == 'WT')

standard_Alignment <- list()
for(i in 1:nrow(Kock_alignment)){
  WTid <- which(CV_alignment$name == mutationInfo$TF[i])
  add <- CV_alignment[WTid[1],]
  standard_Alignment <- rbind.data.frame(standard_Alignment, add)
}

#WTmodel <- makeSVDModel(CV_motif_Kock, CV_alignment_Kock, Positions = motifPoses, Ftest_pVal = 0.01)

mutationPredList <- list()
mutationPredConfidence <- c()
StandardPredList <- list()
inTraining <- c()
for(i in mutantID){
  mutpos <- as.numeric(mutationInfo$position[i])-1
  svdModel <- makeSVDModel.singleMutation(CV_motif_Kock, CV_alignment_Kock, Positions = motifPoses, mutation = mutpos)
  inTraining <- c(sum(substr(Kock_alignment$alignment[i],mutpos,mutpos) == substr(CV_alignment_Kock$alignment,mutpos,mutpos)), inTraining)
  #predict mutant
  predPSAM <- data.frame(pred = c(0,0,0,0))
  rownames(predPSAM) <- DNA()
  for(pos in motifPoses){
    predColumn <- predict(svdModel[[pos]], Kock_alignment[i,], useSimilarAA = T)
    #if(sum(mutpos == unique(unlist(WTmodel[[pos]]$keyPos))) == 0){
    #  predColumn[,1] <- Kock_motifs_set[[max(WTID[WTID<i])]]$matrix[,pos]
    #}
    predPSAM <- cbind.data.frame(predPSAM, predColumn)
    predConfidence <- attr(predColumn, 'confidence')
  }
  predPSAM <- predPSAM[,-1]
  colnames(predPSAM) <- motifPoses
  #predict standard
  WTPSAM <- data.frame(pred = c(0,0,0,0))
  rownames(WTPSAM) <- DNA()
  for(pos in motifPoses){
    predColumn <- predict(svdModel[[pos]], standard_Alignment[i,], useSimilarAA = T)
    WTPSAM <- cbind.data.frame(WTPSAM, predColumn)
  }
  WTPSAM <- WTPSAM[,-1]
  colnames(WTPSAM) <- motifPoses
  
  mutationPredList[[length(mutationPredList)+1]] <- predPSAM
  StandardPredList[[length(StandardPredList)+1]] <- WTPSAM
  mutationPredConfidence <- c(mutationPredConfidence, predConfidence)
}

trueMotifs <- Kock_motifs_set[mutantID]
predMotifs <- mutationPredList

trueStandard <- list()
for(i in mutantID){
  setA <- which(mutationInfo$TF == mutationInfo$TF[i])
  setB <- which(mutationInfo$mutation == 'WT')
  WTid <- intersect(setA, setB)
  add <- Kock_motifs_set[[WTid[1]]]
  trueStandard[[length(trueStandard)+1]] <- add$matrix
}

predStandard <- trueStandard
#predStandard <- StandardPredList

#for(i in mutantID){
#  WTid <- which(CV_alignment_Kock$name == mutationInfo$TF[i])
#  add <- CV_motif_Kock[[WTid[1]]]
#  predStandard[[length(predStandard)+1]] <- add$matrix
#}
```

```{r}
dddGTrue <- list()
dddGpred <- list()
R2List <- c()
for(i in 1:length(trueMotifs)){
  predDiffTetra <- matrix2tetrahedron(predMotifs[[i]]) - matrix2tetrahedron(predStandard[[i]])
  if(sum(predDiffTetra) == 0){
    predDiffTetra[1] <- 0.01
    mod <- TRUE
  }else{
    mod <- FALSE
  }
  trueDiffTetra <- matrix2tetrahedron(trueMotifs[[i]]$matrix) - matrix2tetrahedron(trueStandard[[i]])
  predDiffMatrix <- tetrahedron2matrix(predDiffTetra)
  trueDiffMatrix <- tetrahedron2matrix(trueDiffTetra)
  predDiffMatrix[predDiffMatrix < 0] <- 0.01
  trueDiffMatrix[trueDiffMatrix < 0] <- 0.01
  rownames(trueDiffMatrix) <- DNA()
  rownames(predDiffMatrix) <- DNA()
  if(mod){
    predDiffMatrix[1,] <- 1
  }
  pred <- as.numeric(frequency2ddG(predDiffMatrix))
  true <- as.numeric(frequency2ddG(trueDiffMatrix))
  predTrue <- data.frame(true = true, pred = pred)
  dddGTrue[[length(dddGTrue) + 1]] <- trueDiffMatrix
  dddGpred[[length(dddGpred) + 1]] <- predDiffMatrix
  plotPredTrue(predTrue, main = paste0('ΔΔΔG comparison for ',unlist(lapply(trueMotifs, function(x) x$name))[i], ' with confidence: ', round(mutationPredConfidence[i],4)))
  R2List <- c(R2List, groupedR2(predTrue$true, predTrue$pred, member = nrow(predTrue)))
}

sum(R2List>0)
sum(R2List>0.43)
plot(mutationPredConfidence,R2List)
dt <- data.frame(conf = mutationPredConfidence, R2 = R2List)
lm <- lm(R2~conf, data= dt)
abline(lm, col= 'red')
legend('topright', paste0('R^2 = ', round(summary(lm)$r.squared,4)), bty = 'n')
```

```{r}
plot.ecdf(R2List)
plot(mutationInfo$position[mutantID],R2List)
R2Position <- data.frame(position = mutationInfo$position[mutantID],R2List)
byPositionR2 <- data.frame(NULL)
for(i in unique(R2Position$position)){
  R2 <- mean(R2Position$R2List[R2Position$position == i])
  addLine <- data.frame(position = i, R2 = R2)
  byPositionR2 <- rbind.data.frame(byPositionR2, addLine)
}
dplyr::arrange(byPositionR2, desc(R2))
```


```{r}

evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
i = 1
for(pos in motifPoses){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('FamilyCode at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  
  #plot(mutationPredConfidence,groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F), ylab = 'Predict-True R^2', xlab = 'Confidence score', main = paste0('Confidence at ', pos))
  #legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
  #                                                   mutationPredConfidence, ),4)), bty = 'n')
  #dt <- data.frame(conf = mutationPredConfidence, R2 = groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F))
  #lm <- lm(R2~conf, data= dt)
  #abline(lm, col= 'red')
  
  plotPredTrue(predTrue.PFM, main = paste0('FamilyCode PFM at ',pos), xlab = 'Expermental', ylab = 'Predicted')
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T,throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  i <- i + 1
}
evaMatrix_FC_dddG <- evaMatrix
```

```{r}
mean(evaMatrix_FC_dddG[1,])
```
F-test pval   0.001   0.01    0.1    0.5    0.8    1      +AA51
mean R2       0.0381  0.0662  0.0431 0.0645 0.0867 0.1456 0.1598

### prediction R2 for large efects (experimental dddG > +-1)

```{r}

evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
i = 1
for(pos in motifPoses){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
  trueMagnitude <- c()
  predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
  predMagnitude <- c()
  for(set in 1:ncol(trueMatrix)){
    trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
    predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
  }
  
  plot(trueMagnitude, groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T),
       main = 'Experimental Maginitude vs Prediction accuracy', xlab = 'magnitude', ylab = 'R2')
  abline(lm(y~x, data = data.frame(x = trueMagnitude, y = groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T))), col = 'red')
  legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
                                                    trueMagnitude ),4)), bty = 'n')
  plot(predMagnitude, groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T),
       main = paste0('Prediction Maginitude vs accuracy at ',pos), xlab = 'magnitude', ylab = 'R2')
  abline(lm(y~x, data = data.frame(x = predMagnitude, y = groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T))), col = 'red')
  legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
                                                    predMagnitude ),4)), bty = 'n')
  
  predTrue <- predTrue[rep(trueMagnitude > 1, each = 4),]
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('FamilyCode at ',pos))
  plotPredTrue(predTrue.PFM, main = paste0('FamilyCode PFM at ',pos), xlab = 'Expermental', ylab = 'Predicted')
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T,throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  i <- i + 1
}
evaMatrix_FC_dddG_largeChange <- evaMatrix
```


### inspect R2 with different magnitudes 
500/500
```{r}
library(pROC)

blues <- colorRampPalette(c("lightblue", "blue", "darkblue"))
blue_spectrum <- blues(18)

R2TrueCutOff <- c()
R2PredCutOff <- c()
AUCs <- c()
for(t in 0:18){
  threshold <- t/4
  predTruePredT <- data.frame(NULL)
  predTrueTrueT <- data.frame(NULL)
  trueMagnitudeAll <- c()
  predMagnitudeAll <- c()
  for(pos in motifPoses){
    pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
    true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
    predTrue <- data.frame(true = true, pred = pred)
    predTrue <- predTrue[!is.na(predTrue$pred),]
    trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
    trueMagnitude <- c()
    predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
    predMagnitude <- c()
    for(set in 1:ncol(trueMatrix)){
      trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
      predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
    }
    trueMagnitudeAll <- c(trueMagnitudeAll, trueMagnitude)
    predMagnitudeAll <- c(predMagnitudeAll, predMagnitude)
    predTrueP <- predTrue[rep(predMagnitude > threshold, each = 4),]
    predTrueT <- predTrue[rep(trueMagnitude > threshold, each = 4),]
    predTrueTrueT <- rbind.data.frame(predTrueTrueT, predTrueT)
    predTruePredT <- rbind.data.frame(predTruePredT, predTrueP)
  }
  R2TrueCutOff <- c(R2TrueCutOff, round(groupedR2(predTrueTrueT$pred, 
        predTrueTrueT$true, nrow(predTrueTrueT), throughZero = T), 
        4))
  R2PredCutOff <- c(R2PredCutOff, round(groupedR2(predTruePredT$pred, 
        predTruePredT$true, nrow(predTruePredT), throughZero = T), 
        4))
  if(t == 0){
    suppressMessages(roc_obj1 <- roc(trueMagnitudeAll > threshold, predMagnitudeAll))
    plot(roc_obj1, col = blue_spectrum[t+1], lwd = 2, main = "FamilyCode ROC curve", xlim = c(1,0),  ylim = c(0,1))
  }else{
    suppressMessages(roc_obj1 <- roc(trueMagnitudeAll > threshold, predMagnitudeAll))
    plot(roc_obj1, col = blue_spectrum[t+1], lwd = 2, xlim = c(1,0),  ylim = c(0,1), add = T)
  }
  AUCs <- c(AUCs, roc_obj1$auc)
}
legend("bottomright", legend = c('threshold = 0', 'threshold = 6.75',''), col = c(blue_spectrum[c(1,18)],'white'), lwd = 2, bty = 'n')
legend("topleft", legend = paste0('\nMax AUC: ', round(max(AUCs),4), '\nMean AUC: ', round(mean(AUCs),4)), bty = 'n')
plot((0:18)/4, R2PredCutOff, type = 'l', ylim = c(0,1))
lines((0:18)/4, R2TrueCutOff, type = 'l', ylim = c(0,1), col = 'red')
```

### screening over different thresholds sync. between true and pred sets.
500/500
```{r}
bins <- 50
trueMagnitudeAll <- c()
predMagnitudeAll <- c()
for(pos in motifPoses){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
  trueMagnitude <- c()
  predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
  predMagnitude <- c()
  for(set in 1:ncol(trueMatrix)){
    trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
    predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
  }
  trueMagnitudeAll <- c(trueMagnitudeAll, trueMagnitude)
  predMagnitudeAll <- c(predMagnitudeAll, predMagnitude)
}

max <- max(trueMagnitudeAll)

FPR <- c()
TPR <- c()
for(t in 0:bins){
  threshold <- max/bins*t*1.001
  classifiedPredTrue <- data.frame(true = paste0('Emperical.',trueMagnitudeAll > threshold), pred = paste0('predicted.',predMagnitudeAll > threshold))
  classifiedPredTrue$true <- factor(classifiedPredTrue$true, levels = c('Emperical.TRUE', 'Emperical.FALSE'))
  classifiedPredTrue$pred <- factor(classifiedPredTrue$pred, levels = c('predicted.TRUE', 'predicted.FALSE'))
  conf_matrix <- table(classifiedPredTrue$true, classifiedPredTrue$pred)
  FPR <- c(FPR, conf_matrix['Emperical.FALSE', 'predicted.TRUE'] / (conf_matrix['Emperical.FALSE', 'predicted.TRUE'] + conf_matrix['Emperical.FALSE', 'predicted.FALSE']))
  TPR <- c(TPR, conf_matrix['Emperical.TRUE', 'predicted.TRUE'] / (conf_matrix['Emperical.TRUE', 'predicted.TRUE'] + conf_matrix['Emperical.TRUE', 'predicted.FALSE']))
  FPR[is.na(FPR)] <- 0
  TPR[is.na(TPR)] <- 0
}
ROC_FC_MS <- data.frame(FPR = FPR, TPR = TPR)

```

### predicting dddGs with FC general model
```{r}
#WTmodel <- makeSVDModel(CV_motif_Kock, CV_alignment_Kock, Positions = motifPoses, Ftest_pVal = 0.01)

mutationPredList <- list()
mutationPredConfidence <- c()
StandardPredList <- list()
inTraining <- c()
svdModel <- makeSVDModel(CV_motif_Kock, CV_alignment_Kock, Positions = motifPoses)
for(i in mutantID){
  mutpos <- as.numeric(mutationInfo$position[i])-1
  inTraining <- c(sum(substr(Kock_alignment$alignment[i],mutpos,mutpos) == substr(CV_alignment_Kock$alignment,mutpos,mutpos)), inTraining)
  #predict mutant
  predPSAM <- data.frame(pred = c(0,0,0,0))
  rownames(predPSAM) <- DNA()
  for(pos in motifPoses){
    predColumn <- predict(svdModel[[pos]], Kock_alignment[i,], useSimilarAA = T)
    #if(sum(mutpos == unique(unlist(WTmodel[[pos]]$keyPos))) == 0){
    #  predColumn[,1] <- Kock_motifs_set[[max(WTID[WTID<i])]]$matrix[,pos]
    #}
    predPSAM <- cbind.data.frame(predPSAM, predColumn)
    predConfidence <- attr(predColumn, 'confidence')
  }
  predPSAM <- predPSAM[,-1]
  colnames(predPSAM) <- motifPoses
  #predict standard
  WTPSAM <- data.frame(pred = c(0,0,0,0))
  rownames(WTPSAM) <- DNA()
  for(pos in motifPoses){
    predColumn <- predict(svdModel[[pos]], standard_Alignment[i,], useSimilarAA = T)
    WTPSAM <- cbind.data.frame(WTPSAM, predColumn)
  }
  WTPSAM <- WTPSAM[,-1]
  colnames(WTPSAM) <- motifPoses
  
  mutationPredList[[length(mutationPredList)+1]] <- predPSAM
  StandardPredList[[length(StandardPredList)+1]] <- WTPSAM
  mutationPredConfidence <- c(mutationPredConfidence, predConfidence)
}

trueMotifs <- Kock_motifs_set[mutantID]
predMotifs <- mutationPredList

trueStandard <- list()
for(i in mutantID){
  setA <- which(mutationInfo$TF == mutationInfo$TF[i])
  setB <- which(mutationInfo$mutation == 'WT')
  WTid <- intersect(setA, setB)
  add <- Kock_motifs_set[[WTid[1]]]
  trueStandard[[length(trueStandard)+1]] <- add$matrix
}


#predStandard <- StandardPredList
predStandard <- trueStandard
#for(i in mutantID){
#  WTid <- which(CV_alignment_Kock$name == mutationInfo$TF[i])
#  add <- CV_motif_Kock[[WTid[1]]]
#  predStandard[[length(predStandard)+1]] <- add$matrix
#}
```

```{r}
dddGTrue <- list()
dddGpred <- list()
R2ListGeneric <- c()
for(i in 1:length(trueMotifs)){
  predDiffTetra <- matrix2tetrahedron(predMotifs[[i]]) - matrix2tetrahedron(predStandard[[i]])
  if(sum(predDiffTetra) == 0){
    predDiffTetra[1] <- 0.01
    mod <- TRUE
  }else{
    mod <- FALSE
  }
  trueDiffTetra <- matrix2tetrahedron(trueMotifs[[i]]$matrix) - matrix2tetrahedron(trueStandard[[i]])
  predDiffMatrix <- tetrahedron2matrix(predDiffTetra)
  trueDiffMatrix <- tetrahedron2matrix(trueDiffTetra)
  predDiffMatrix[predDiffMatrix < 0] <- 0.01
  trueDiffMatrix[trueDiffMatrix < 0] <- 0.01
  rownames(trueDiffMatrix) <- DNA()
  rownames(predDiffMatrix) <- DNA()
  if(mod){
    predDiffMatrix[1,] <- 1
  }
  pred <- as.numeric(frequency2ddG(predDiffMatrix))
  true <- as.numeric(frequency2ddG(trueDiffMatrix))
  predTrue <- data.frame(true = true, pred = pred)
  dddGTrue[[length(dddGTrue) + 1]] <- trueDiffMatrix
  dddGpred[[length(dddGpred) + 1]] <- predDiffMatrix
  plotPredTrue(predTrue, main = paste0('ΔΔΔG comparison for ',unlist(lapply(trueMotifs, function(x) x$name))[i], ' with confidence: ', round(mutationPredConfidence[i],4)))
  R2ListGeneric <- c(R2ListGeneric, groupedR2(predTrue$true, predTrue$pred, member = nrow(predTrue)))
}

sum(R2ListGeneric>0)
sum(R2ListGeneric>0.41)
plot(mutationPredConfidence,R2ListGeneric)
dt <- data.frame(conf = mutationPredConfidence, R2 = R2ListGeneric)
lm <- lm(R2~conf, data= dt)
abline(lm, col= 'red')
legend('topright', paste0('R^2 = ', round(summary(lm)$r.squared,4)), bty = 'n')
```

```{r}
plot.ecdf(R2ListGeneric)
plot(mutationInfo$position[mutantID],R2ListGeneric)
R2Position <- data.frame(position = mutationInfo$position[mutantID],R2ListGeneric)
byPositionR2 <- data.frame(NULL)
for(i in unique(R2Position$position)){
  R2 <- mean(R2Position$R2List[R2Position$position == i])
  addLine <- data.frame(position = i, R2 = R2)
  byPositionR2 <- rbind.data.frame(byPositionR2, addLine)
}
dplyr::arrange(byPositionR2, desc(R2))
```


```{r}

evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
i = 1
for(pos in motifPoses){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('FamilyCode at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  
  #plot(mutationPredConfidence,groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F), ylab = 'Predict-True R^2', xlab = 'Confidence score', main = paste0('Confidence at ', pos))
  #legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
  #                                                   mutationPredConfidence, ),4)), bty = 'n')
  #dt <- data.frame(conf = mutationPredConfidence, R2 = groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F))
  #lm <- lm(R2~conf, data= dt)
  #abline(lm, col= 'red')
  
  plotPredTrue(predTrue.PFM, main = paste0('FamilyCode PFM at ',pos), xlab = 'Expermental', ylab = 'Predicted')
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T,throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  i <- i + 1
}
evaMatrix_FCGeneric_dddG <- evaMatrix
```

### prediction R2 for large efects (experimental dddG > +-1)

```{r}

evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
i = 1
for(pos in motifPoses){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
  trueMagnitude <- c()
  predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
  predMagnitude <- c()
  for(set in 1:ncol(trueMatrix)){
    trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
    predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
  }
  
  plot(trueMagnitude, groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T),
       main = 'Experimental Maginitude vs Prediction accuracy', xlab = 'magnitude', ylab = 'R2')
  abline(lm(y~x, data = data.frame(x = trueMagnitude, y = groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T))), col = 'red')
  legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
                                                    trueMagnitude ),4)), bty = 'n')
  plot(predMagnitude, groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T),
       main = paste0('Prediction Maginitude vs accuracy at ',pos), xlab = 'magnitude', ylab = 'R2')
  abline(lm(y~x, data = data.frame(x = predMagnitude, y = groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T))), col = 'red')
  legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
                                                    predMagnitude ),4)), bty = 'n')
  
  predTrue <- predTrue[rep(trueMagnitude > 1.5, each = 4),]
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('FamilyCode Generic at ',pos))
  plotPredTrue(predTrue.PFM, main = paste0('FamilyCode Generic at ',pos), xlab = 'Expermental', ylab = 'Predicted')
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T,throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  i <- i + 1
}
evaMatrix_FCGeneric_dddG_largeChange <- evaMatrix
```

### inspect R2 with different magnitudes 
500/500
```{r}
library(pROC)

blues <- colorRampPalette(c("lightblue", "blue", "darkblue"))
blue_spectrum <- blues(18)

R2TrueCutOff <- c()
R2PredCutOff <- c()
AUCs <- c()
for(t in 0:15){
  threshold <- t/4
  predTruePredT <- data.frame(NULL)
  predTrueTrueT <- data.frame(NULL)
  trueMagnitudeAll <- c()
  predMagnitudeAll <- c()
  for(pos in motifPoses){
    pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
    true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
    predTrue <- data.frame(true = true, pred = pred)
    predTrue <- predTrue[!is.na(predTrue$pred),]
    trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
    trueMagnitude <- c()
    predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
    predMagnitude <- c()
    for(set in 1:ncol(trueMatrix)){
      trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
      predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
    }
    trueMagnitudeAll <- c(trueMagnitudeAll, trueMagnitude)
    predMagnitudeAll <- c(predMagnitudeAll, predMagnitude)
    predTrueP <- predTrue[rep(predMagnitude > threshold, each = 4),]
    predTrueT <- predTrue[rep(trueMagnitude > threshold, each = 4),]
    predTrueTrueT <- rbind.data.frame(predTrueTrueT, predTrueT)
    predTruePredT <- rbind.data.frame(predTruePredT, predTrueP)
  }
  R2TrueCutOff <- c(R2TrueCutOff, round(groupedR2(predTrueTrueT$pred, 
        predTrueTrueT$true, nrow(predTrueTrueT), throughZero = T), 
        4))
  R2PredCutOff <- c(R2PredCutOff, round(groupedR2(predTruePredT$pred, 
        predTruePredT$true, nrow(predTruePredT), throughZero = T), 
        4))
  if(t == 0){
    suppressMessages(roc_obj1 <- roc(trueMagnitudeAll > threshold, predMagnitudeAll))
    plot(roc_obj1, col = blue_spectrum[t+1], lwd = 2, main = "FamilyCode generic ROC curve", xlim = c(1,0),  ylim = c(0,1))
  }else{
    suppressMessages(roc_obj1 <- roc(trueMagnitudeAll > threshold, predMagnitudeAll))
    plot(roc_obj1, col = blue_spectrum[t+1], lwd = 2, xlim = c(1,0),  ylim = c(0,1), add = T)
  }
  AUCs <- c(AUCs, roc_obj1$auc)
}
legend("bottomright", legend = c('threshold = 0', 'threshold = 6.75',''), col = c(blue_spectrum[c(1,18)],'white'), lwd = 2, bty = 'n')
legend("topleft", legend = paste0('\nMax AUC: ', round(max(AUCs),4), '\nMean AUC: ', round(mean(AUCs),4)), bty = 'n')
plot((0:(length(R2PredCutOff)-1))/4, R2PredCutOff, type = 'l', ylim = c(0,1))
lines((0:(length(R2PredCutOff)-1))/4, R2TrueCutOff, type = 'l', ylim = c(0,1), col = 'red')
```

### screening over different thresholds sync. between true and pred sets.
500/500
```{r}
bins <- 50
trueMagnitudeAll <- c()
predMagnitudeAll <- c()
for(pos in motifPoses){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
  trueMagnitude <- c()
  predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
  predMagnitude <- c()
  for(set in 1:ncol(trueMatrix)){
    trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
    predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
  }
  trueMagnitudeAll <- c(trueMagnitudeAll, trueMagnitude)
  predMagnitudeAll <- c(predMagnitudeAll, predMagnitude)
}

max <- max(trueMagnitudeAll)

FPR <- c()
TPR <- c()
for(t in 0:bins){
  threshold <- max/bins*t*1.001
  classifiedPredTrue <- data.frame(true = paste0('Emperical.',trueMagnitudeAll > threshold), pred = paste0('predicted.',predMagnitudeAll > threshold))
  classifiedPredTrue$true <- factor(classifiedPredTrue$true, levels = c('Emperical.TRUE', 'Emperical.FALSE'))
  classifiedPredTrue$pred <- factor(classifiedPredTrue$pred, levels = c('predicted.TRUE', 'predicted.FALSE'))
  conf_matrix <- table(classifiedPredTrue$true, classifiedPredTrue$pred)
  FPR <- c(FPR, conf_matrix['Emperical.FALSE', 'predicted.TRUE'] / (conf_matrix['Emperical.FALSE', 'predicted.TRUE'] + conf_matrix['Emperical.FALSE', 'predicted.FALSE']))
  TPR <- c(TPR, conf_matrix['Emperical.TRUE', 'predicted.TRUE'] / (conf_matrix['Emperical.TRUE', 'predicted.TRUE'] + conf_matrix['Emperical.TRUE', 'predicted.FALSE']))
  FPR[is.na(FPR)] <- 0
  TPR[is.na(TPR)] <- 0
}
ROC_FC_GM <- data.frame(FPR = FPR, TPR = TPR)

```


### predicting dddGs with FC_rCLAMPS features
```{r}
mutantID <- which(mutationInfo$mutate != 'WT')
WTID <- which(mutationInfo$mutate == 'WT')

CV_alignment_rCF <- readRDS('CV_alignment_rCLAMPSFeatures.rds')
CV_motif_rCLAMPS <- readRDS('CV_motif_rCLAMPS.rds')

standard_Alignment <- list()
for(i in 1:nrow(Kock_alignment)){
  WTid <- which(CV_alignment$name == mutationInfo$TF[i])
  add <- CV_alignment[WTid[1],]
  standard_Alignment <- rbind.data.frame(standard_Alignment, add)
}

#WTmodel <- makeSVDModel(CV_motif_Kock, CV_alignment_Kock, Positions = motifPoses, Ftest_pVal = 0.01)

mutationPredList <- list()
mutationPredConfidence <- c()
StandardPredList <- list()
inTraining <- c()
Kock_alignment_rC <- Kock_alignment
positions <- c(1,2,3,4,46,49,50,53,54)
Kock_alignment_rC$alignment <- sapply(Kock_alignment_rC$alignment, function(s) {
  paste(substring(s, positions, positions), collapse = "")
}, USE.NAMES = FALSE)

WTID <- which(mutationInfo$mutation == 'WT')
for(i in mutantID){
  mutpos <- which(strsplit(Kock_alignment_rC$alignment[i], '')[[1]][1] != strsplit(Kock_alignment_rC$alignment[max(WTID[WTID < i])], '')[[1]][1])
  if(length(mutpos) == 0){
    predPSAM <- Kock_motifs_set[[max(WTID[WTID < i])]]$matrix[,3:8]
    WTPSAM <- Kock_motifs_set[[max(WTID[WTID < i])]]$matrix[,3:8]
    predConfidence <- 0
  }else{
    svdModel <- makeSVDModel.singleMutation(CV_motif_rCLAMPS, CV_alignment_rCF, Positions = motifPoses[3:8], mutation = mutpos)
    inTraining <- c(sum(substr(Kock_alignment$alignment[i],mutpos,mutpos) == substr(CV_alignment_Kock$alignment,mutpos,mutpos)), inTraining)
    #predict mutant
    predPSAM <- data.frame(pred = c(0,0,0,0))
    rownames(predPSAM) <- DNA()
    for(pos in motifPoses[3:8]){
      predColumn <- predict(svdModel[[pos]], Kock_alignment[i,], useSimilarAA = T)
      #if(sum(mutpos == unique(unlist(WTmodel[[pos]]$keyPos))) == 0){
      #  predColumn[,1] <- Kock_motifs_set[[max(WTID[WTID<i])]]$matrix[,pos]
      #}
      predPSAM <- cbind.data.frame(predPSAM, predColumn)
      predConfidence <- attr(predColumn, 'confidence')
    }
    predPSAM <- predPSAM[,-1]
    colnames(predPSAM) <- motifPoses[3:8]
    #predict standard
    WTPSAM <- Kock_motifs_set[[max(WTID[WTID < i])]]$matrix[,3:8]
    colnames(WTPSAM) <- motifPoses[3:8]
  }
  mutationPredList[[length(mutationPredList)+1]] <- predPSAM
  StandardPredList[[length(StandardPredList)+1]] <- WTPSAM
  mutationPredConfidence <- c(mutationPredConfidence, predConfidence)
}

trueMotifs <- Kock_motifs_set[mutantID]
predMotifs <- mutationPredList

trueStandard <- list()
for(i in mutantID){
  setA <- which(mutationInfo$TF == mutationInfo$TF[i])
  setB <- which(mutationInfo$mutation == 'WT')
  WTid <- intersect(setA, setB)
  add <- Kock_motifs_set[[WTid[1]]]
  trueStandard[[length(trueStandard)+1]] <- add$matrix
}

predStandard <- trueStandard
#predStandard <- StandardPredList

#for(i in mutantID){
#  WTid <- which(CV_alignment_Kock$name == mutationInfo$TF[i])
#  add <- CV_motif_Kock[[WTid[1]]]
#  predStandard[[length(predStandard)+1]] <- add$matrix
#}
```

```{r}
dddGTrue <- list()
dddGpred <- list()
R2ListRCSM <- c()
for(i in 1:length(trueMotifs)){
  predDiffTetra <- matrix2tetrahedron(predMotifs[[i]]) - matrix2tetrahedron(predStandard[[i]][,3:8])
  if(sum(predDiffTetra) == 0){
    predDiffTetra[1] <- 0.01
    mod <- TRUE
  }else{
    mod <- FALSE
  }
  trueDiffTetra <- matrix2tetrahedron(trueMotifs[[i]]$matrix[,3:8]) - matrix2tetrahedron(trueStandard[[i]][,3:8])
  predDiffMatrix <- tetrahedron2matrix(predDiffTetra)
  trueDiffMatrix <- tetrahedron2matrix(trueDiffTetra)
  predDiffMatrix[predDiffMatrix < 0] <- 0.01
  trueDiffMatrix[trueDiffMatrix < 0] <- 0.01
  rownames(trueDiffMatrix) <- DNA()
  rownames(predDiffMatrix) <- DNA()
  if(mod){
    predDiffMatrix[1,] <- 1
  }
  pred <- as.numeric(frequency2ddG(predDiffMatrix))
  true <- as.numeric(frequency2ddG(trueDiffMatrix))
  predTrue <- data.frame(true = true, pred = pred)
  dddGTrue[[length(dddGTrue) + 1]] <- trueDiffMatrix
  dddGpred[[length(dddGpred) + 1]] <- predDiffMatrix
  plotPredTrue(predTrue, main = paste0('ΔΔΔG comparison for ',unlist(lapply(trueMotifs, function(x) x$name))[i], ' with confidence: ', round(mutationPredConfidence[i],4)))
  R2ListRCSM <- c(R2ListRCSM, groupedR2(predTrue$true, predTrue$pred, member = nrow(predTrue)))
}

sum(R2ListRCSM>0)
sum(R2ListRCSM>0.41)
plot(mutationPredConfidence,R2ListRCSM)
dt <- data.frame(conf = mutationPredConfidence, R2 = R2ListRCSM)
lm <- lm(R2~conf, data= dt)
abline(lm, col= 'red')
legend('topright', paste0('R^2 = ', round(summary(lm)$r.squared,4)), bty = 'n')
```

```{r}
plot.ecdf(R2ListRCSM)
plot(mutationInfo$position[mutantID],R2ListRCSM)
R2Position <- data.frame(position = mutationInfo$position[mutantID],R2ListRCSM)
byPositionR2 <- data.frame(NULL)
for(i in unique(R2Position$position)){
  R2 <- mean(R2Position$R2List[R2Position$position == i])
  addLine <- data.frame(position = i, R2 = R2)
  byPositionR2 <- rbind.data.frame(byPositionR2, addLine)
}
dplyr::arrange(byPositionR2, desc(R2))
```


```{r}

evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
i = 1
for(pos in motifPoses[3:8]){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('FamilyCode with rCLAMPS at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  
  #plot(mutationPredConfidence,groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F), ylab = 'Predict-True R^2', xlab = 'Confidence score', main = paste0('Confidence at ', pos))
  #legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
  #                                                   mutationPredConfidence, ),4)), bty = 'n')
  #dt <- data.frame(conf = mutationPredConfidence, R2 = groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F))
  #lm <- lm(R2~conf, data= dt)
  #abline(lm, col= 'red')
  
  plotPredTrue(predTrue.PFM, main = paste0('FamilyCode with rCLAMPS PFM at ',pos), xlab = 'Expermental', ylab = 'Predicted')
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T,throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  i <- i + 1
}
evaMatrix_FC_RC_dddG <- evaMatrix
```

### inspect R2 with different magnitudes 
500/500
```{r}
library(pROC)

blues <- colorRampPalette(c("lightblue", "blue", "darkblue"))
blue_spectrum <- blues(18)

R2TrueCutOff <- c()
R2PredCutOff <- c()
AUCs <- c()
for(t in 0:15){
  threshold <- t/4
  predTruePredT <- data.frame(NULL)
  predTrueTrueT <- data.frame(NULL)
  trueMagnitudeAll <- c()
  predMagnitudeAll <- c()
  for(pos in motifPoses[3:8]){
    pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
    true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
    predTrue <- data.frame(true = true, pred = pred)
    predTrue <- predTrue[!is.na(predTrue$pred),]
    trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
    trueMagnitude <- c()
    predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
    predMagnitude <- c()
    for(set in 1:ncol(trueMatrix)){
      trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
      predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
    }
    trueMagnitudeAll <- c(trueMagnitudeAll, trueMagnitude)
    predMagnitudeAll <- c(predMagnitudeAll, predMagnitude)
    predTrueP <- predTrue[rep(predMagnitude > threshold, each = 4),]
    predTrueT <- predTrue[rep(trueMagnitude > threshold, each = 4),]
    predTrueTrueT <- rbind.data.frame(predTrueTrueT, predTrueT)
    predTruePredT <- rbind.data.frame(predTruePredT, predTrueP)
  }
  R2TrueCutOff <- c(R2TrueCutOff, round(groupedR2(predTrueTrueT$pred, 
        predTrueTrueT$true, nrow(predTrueTrueT), throughZero = T), 
        4))
  R2PredCutOff <- c(R2PredCutOff, round(groupedR2(predTruePredT$pred, 
        predTruePredT$true, nrow(predTruePredT), throughZero = T), 
        4))
  if(t == 0){
    suppressMessages(roc_obj1 <- roc(trueMagnitudeAll > threshold, predMagnitudeAll))
    plot(roc_obj1, col = blue_spectrum[t+1], lwd = 2, main = "FamilyCode generic with rCLAMPS ROC curve", xlim = c(1,0),  ylim = c(0,1))
  }else{
    suppressMessages(roc_obj1 <- roc(trueMagnitudeAll > threshold, predMagnitudeAll))
    plot(roc_obj1, col = blue_spectrum[t+1], lwd = 2, xlim = c(1,0),  ylim = c(0,1), add = T)
  }
  AUCs <- c(AUCs, roc_obj1$auc)
}
legend("bottomright", legend = c('threshold = 0', 'threshold = 6.75',''), col = c(blue_spectrum[c(1,18)],'white'), lwd = 2, bty = 'n')
legend("topleft", legend = paste0('\nMax AUC: ', round(max(AUCs),4), '\nMean AUC: ', round(mean(AUCs),4)), bty = 'n')
plot((0:(length(R2PredCutOff)-1))/4, R2PredCutOff, type = 'l', ylim = c(0,1))
lines((0:(length(R2PredCutOff)-1))/4, R2TrueCutOff, type = 'l', ylim = c(0,1), col = 'red')
```

### screening over different thresholds sync. between true and pred sets.
500/500
```{r}
bins <- 50
trueMagnitudeAll <- c()
predMagnitudeAll <- c()
for(pos in motifPoses[3:8]){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
  trueMagnitude <- c()
  predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
  predMagnitude <- c()
  for(set in 1:ncol(trueMatrix)){
    trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
    predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
  }
  trueMagnitudeAll <- c(trueMagnitudeAll, trueMagnitude)
  predMagnitudeAll <- c(predMagnitudeAll, predMagnitude)
}

max <- max(trueMagnitudeAll)

FPR <- c()
TPR <- c()
for(t in 0:bins){
  threshold <- max/bins*t*1.001
  classifiedPredTrue <- data.frame(true = paste0('Emperical.',trueMagnitudeAll > threshold), pred = paste0('predicted.',predMagnitudeAll > threshold))
  classifiedPredTrue$true <- factor(classifiedPredTrue$true, levels = c('Emperical.TRUE', 'Emperical.FALSE'))
  classifiedPredTrue$pred <- factor(classifiedPredTrue$pred, levels = c('predicted.TRUE', 'predicted.FALSE'))
  conf_matrix <- table(classifiedPredTrue$true, classifiedPredTrue$pred)
  FPR <- c(FPR, conf_matrix['Emperical.FALSE', 'predicted.TRUE'] / (conf_matrix['Emperical.FALSE', 'predicted.TRUE'] + conf_matrix['Emperical.FALSE', 'predicted.FALSE']))
  TPR <- c(TPR, conf_matrix['Emperical.TRUE', 'predicted.TRUE'] / (conf_matrix['Emperical.TRUE', 'predicted.TRUE'] + conf_matrix['Emperical.TRUE', 'predicted.FALSE']))
  FPR[is.na(FPR)] <- 0
  TPR[is.na(TPR)] <- 0
}
ROC_FC_rCSM <- data.frame(FPR = FPR, TPR = TPR)
```


### predicting dddGs with FC general model with rCLAMPS features 
```{r}
#WTmodel <- makeSVDModel(CV_motif_Kock, CV_alignment_Kock, Positions = motifPoses, Ftest_pVal = 0.01)
CV_alignment_rCF <- readRDS('CV_alignment_rCLAMPSFeatures.rds')
CV_motif_rCLAMPS <- readRDS('CV_motif_rCLAMPS.rds')

mutationPredList <- list()
mutationPredConfidence <- c()
StandardPredList <- list()
inTraining <- c()
svdModel <- makeSVDModel(CV_motif_rCLAMPS, CV_alignment_rCF, Positions = motifPoses[3:8])
for(i in mutantID){
  mutpos <- as.numeric(mutationInfo$position[i])-1
  inTraining <- c(sum(substr(Kock_alignment$alignment[i],mutpos,mutpos) == substr(CV_alignment_Kock$alignment,mutpos,mutpos)), inTraining)
  #predict mutant
  predPSAM <- data.frame(pred = c(0,0,0,0))
  rownames(predPSAM) <- DNA()
  for(pos in motifPoses[3:8]){
    predColumn <- predict(svdModel[[pos]], Kock_alignment_rC[i,], useSimilarAA = T)
    #if(sum(mutpos == unique(unlist(WTmodel[[pos]]$keyPos))) == 0){
    #  predColumn[,1] <- Kock_motifs_set[[max(WTID[WTID<i])]]$matrix[,pos]
    #}
    predPSAM <- cbind.data.frame(predPSAM, predColumn)
    predConfidence <- attr(predColumn, 'confidence')
  }
  predPSAM <- predPSAM[,-1]
  colnames(predPSAM) <- motifPoses[3:8]
  #predict standard
  WTPSAM <- data.frame(pred = c(0,0,0,0))
  rownames(WTPSAM) <- DNA()
  for(pos in motifPoses[3:8]){
    predColumn <- predict(svdModel[[pos]], standard_Alignment[i,], useSimilarAA = T)
    WTPSAM <- cbind.data.frame(WTPSAM, predColumn)
  }
  WTPSAM <- WTPSAM[,-1]
  colnames(WTPSAM) <- motifPoses[3:8]
  
  mutationPredList[[length(mutationPredList)+1]] <- predPSAM
  StandardPredList[[length(StandardPredList)+1]] <- WTPSAM
  mutationPredConfidence <- c(mutationPredConfidence, predConfidence)
}

trueMotifs <- Kock_motifs_set[mutantID]
predMotifs <- mutationPredList

trueStandard <- list()
for(i in mutantID){
  setA <- which(mutationInfo$TF == mutationInfo$TF[i])
  setB <- which(mutationInfo$mutation == 'WT')
  WTid <- intersect(setA, setB)
  add <- Kock_motifs_set[[WTid[1]]]
  trueStandard[[length(trueStandard)+1]] <- add$matrix
}


#predStandard <- StandardPredList
predStandard <- trueStandard
#for(i in mutantID){
#  WTid <- which(CV_alignment_Kock$name == mutationInfo$TF[i])
#  add <- CV_motif_Kock[[WTid[1]]]
#  predStandard[[length(predStandard)+1]] <- add$matrix
#}
```

```{r}
dddGTrue <- list()
dddGpred <- list()
R2ListRCGM <- c()
for(i in 1:length(trueMotifs)){
  predDiffTetra <- matrix2tetrahedron(predMotifs[[i]]) - matrix2tetrahedron(predStandard[[i]][,3:8])
  if(sum(predDiffTetra) == 0){
    predDiffTetra[1] <- 0.01
    mod <- TRUE
  }else{
    mod <- FALSE
  }
  trueDiffTetra <- matrix2tetrahedron(trueMotifs[[i]]$matrix[,3:8]) - matrix2tetrahedron(trueStandard[[i]][,3:8])
  predDiffMatrix <- tetrahedron2matrix(predDiffTetra)
  trueDiffMatrix <- tetrahedron2matrix(trueDiffTetra)
  predDiffMatrix[predDiffMatrix < 0] <- 0.01
  trueDiffMatrix[trueDiffMatrix < 0] <- 0.01
  rownames(trueDiffMatrix) <- DNA()
  rownames(predDiffMatrix) <- DNA()
  if(mod){
    predDiffMatrix[1,] <- 1
  }
  pred <- as.numeric(frequency2ddG(predDiffMatrix))
  true <- as.numeric(frequency2ddG(trueDiffMatrix))
  predTrue <- data.frame(true = true, pred = pred)
  dddGTrue[[length(dddGTrue) + 1]] <- trueDiffMatrix
  dddGpred[[length(dddGpred) + 1]] <- predDiffMatrix
  plotPredTrue(predTrue, main = paste0('ΔΔΔG comparison for ',unlist(lapply(trueMotifs, function(x) x$name))[i], ' with confidence: ', round(mutationPredConfidence[i],4)))
  R2ListRCGM <- c(R2ListRCGM, groupedR2(predTrue$true, predTrue$pred, member = nrow(predTrue)))
}

sum(R2ListRCGM>0)
sum(R2ListRCGM>0.41)
plot(mutationPredConfidence,R2ListRCGM)
dt <- data.frame(conf = mutationPredConfidence, R2 = R2ListRCGM)
lm <- lm(R2~conf, data= dt)
abline(lm, col= 'red')
legend('topright', paste0('R^2 = ', round(summary(lm)$r.squared,4)), bty = 'n')
```

```{r}
plot.ecdf(R2ListRCGM)
plot(mutationInfo$position[mutantID],R2ListRCGM)
R2Position <- data.frame(position = mutationInfo$position[mutantID],R2ListRCGM)
byPositionR2 <- data.frame(NULL)
for(i in unique(R2Position$position)){
  R2 <- mean(R2Position$R2List[R2Position$position == i])
  addLine <- data.frame(position = i, R2 = R2)
  byPositionR2 <- rbind.data.frame(byPositionR2, addLine)
}
dplyr::arrange(byPositionR2, desc(R2))
```


```{r}

evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
i = 1
for(pos in motifPoses[3:8]){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('FamilyCode with rCLAMPS at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  
  #plot(mutationPredConfidence,groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F), ylab = 'Predict-True R^2', xlab = 'Confidence score', main = paste0('Confidence at ', pos))
  #legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
  #                                                   mutationPredConfidence, ),4)), bty = 'n')
  #dt <- data.frame(conf = mutationPredConfidence, R2 = groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F))
  #lm <- lm(R2~conf, data= dt)
  #abline(lm, col= 'red')
  
  plotPredTrue(predTrue.PFM, main = paste0('FamilyCode  with rCLAMPS PFM at ',pos), xlab = 'Expermental', ylab = 'Predicted')
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T,throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  i <- i + 1
}
evaMatrix_FCGeneric_RC_dddG <- evaMatrix
```

### inspect R2 with different magnitudes 
500/500
```{r}
library(pROC)

blues <- colorRampPalette(c("lightblue", "blue", "darkblue"))
blue_spectrum <- blues(18)

R2TrueCutOff <- c()
R2PredCutOff <- c()
AUCs <- c()
for(t in 0:15){
  threshold <- t/4
  predTruePredT <- data.frame(NULL)
  predTrueTrueT <- data.frame(NULL)
  trueMagnitudeAll <- c()
  predMagnitudeAll <- c()
  for(pos in motifPoses[3:8]){
    pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
    true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
    predTrue <- data.frame(true = true, pred = pred)
    predTrue <- predTrue[!is.na(predTrue$pred),]
    trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
    trueMagnitude <- c()
    predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
    predMagnitude <- c()
    for(set in 1:ncol(trueMatrix)){
      trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
      predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
    }
    trueMagnitudeAll <- c(trueMagnitudeAll, trueMagnitude)
    predMagnitudeAll <- c(predMagnitudeAll, predMagnitude)
    predTrueP <- predTrue[rep(predMagnitude > threshold, each = 4),]
    predTrueT <- predTrue[rep(trueMagnitude > threshold, each = 4),]
    predTrueTrueT <- rbind.data.frame(predTrueTrueT, predTrueT)
    predTruePredT <- rbind.data.frame(predTruePredT, predTrueP)
  }
  R2TrueCutOff <- c(R2TrueCutOff, round(groupedR2(predTrueTrueT$pred, 
        predTrueTrueT$true, nrow(predTrueTrueT), throughZero = T), 
        4))
  R2PredCutOff <- c(R2PredCutOff, round(groupedR2(predTruePredT$pred, 
        predTruePredT$true, nrow(predTruePredT), throughZero = T), 
        4))
  if(t == 0){
    suppressMessages(roc_obj1 <- roc(trueMagnitudeAll > threshold, predMagnitudeAll))
    plot(roc_obj1, col = blue_spectrum[t+1], lwd = 2, main = "FamilyCode generic with rCLAMPS ROC curve", xlim = c(1,0),  ylim = c(0,1))
  }else{
    suppressMessages(roc_obj1 <- roc(trueMagnitudeAll > threshold, predMagnitudeAll))
    plot(roc_obj1, col = blue_spectrum[t+1], lwd = 2, xlim = c(1,0),  ylim = c(0,1), add = T)
  }
  AUCs <- c(AUCs, roc_obj1$auc)
}
legend("bottomright", legend = c('threshold = 0', 'threshold = 6.75',''), col = c(blue_spectrum[c(1,18)],'white'), lwd = 2, bty = 'n')
legend("topleft", legend = paste0('\nMax AUC: ', round(max(AUCs),4), '\nMean AUC: ', round(mean(AUCs),4)), bty = 'n')
plot((0:(length(R2PredCutOff)-1))/4, R2PredCutOff, type = 'l', ylim = c(0,1))
lines((0:(length(R2PredCutOff)-1))/4, R2TrueCutOff, type = 'l', ylim = c(0,1), col = 'red')
```

### screening over different thresholds sync. between true and pred sets.
500/500
```{r}
bins <- 50
trueMagnitudeAll <- c()
predMagnitudeAll <- c()
for(pos in motifPoses[3:8]){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
  trueMagnitude <- c()
  predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
  predMagnitude <- c()
  for(set in 1:ncol(trueMatrix)){
    trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
    predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
  }
  trueMagnitudeAll <- c(trueMagnitudeAll, trueMagnitude)
  predMagnitudeAll <- c(predMagnitudeAll, predMagnitude)
}

max <- max(trueMagnitudeAll)

FPR <- c()
TPR <- c()
for(t in 0:bins){
  threshold <- max/bins*t*1.001
  classifiedPredTrue <- data.frame(true = paste0('Emperical.',trueMagnitudeAll > threshold), pred = paste0('predicted.',predMagnitudeAll > threshold))
  classifiedPredTrue$true <- factor(classifiedPredTrue$true, levels = c('Emperical.TRUE', 'Emperical.FALSE'))
  classifiedPredTrue$pred <- factor(classifiedPredTrue$pred, levels = c('predicted.TRUE', 'predicted.FALSE'))
  conf_matrix <- table(classifiedPredTrue$true, classifiedPredTrue$pred)
  FPR <- c(FPR, conf_matrix['Emperical.FALSE', 'predicted.TRUE'] / (conf_matrix['Emperical.FALSE', 'predicted.TRUE'] + conf_matrix['Emperical.FALSE', 'predicted.FALSE']))
  TPR <- c(TPR, conf_matrix['Emperical.TRUE', 'predicted.TRUE'] / (conf_matrix['Emperical.TRUE', 'predicted.TRUE'] + conf_matrix['Emperical.TRUE', 'predicted.FALSE']))
  FPR[is.na(FPR)] <- 0
  TPR[is.na(TPR)] <- 0
}
ROC_FC_rCGM <- data.frame(FPR = FPR, TPR = TPR)
```


### test for baseline dddG
```{r}
dddGTrue <- list()
dddGpred <- list()
R2ListbaseLine <- c()
for(i in 1:length(trueMotifs)){
  mutid <- mutantID[i]
  if(length(Exp2TrueMotifs[[mutid]]) == 0){
    R2ListbaseLine <- c(R2ListbaseLine, 0)
    next
  }
  predDiffTetra <- matrix2tetrahedron(Exp2TrueMotifs[[mutid]]$matrix) - matrix2tetrahedron(trueStandard[[i]])
  if(sum(predDiffTetra) == 0){
    R2ListbaseLine <- c(R2ListbaseLine, 0)
    next
  }
  trueDiffTetra <- matrix2tetrahedron(trueMotifs[[i]]$matrix) - matrix2tetrahedron(trueStandard[[i]])
  predDiffMatrix <- tetrahedron2matrix(predDiffTetra)
  trueDiffMatrix <- tetrahedron2matrix(trueDiffTetra)
  predDiffMatrix[predDiffMatrix < 0] <- 0.01
  trueDiffMatrix[trueDiffMatrix < 0] <- 0.01
  rownames(trueDiffMatrix) <- DNA()
  rownames(predDiffMatrix) <- DNA()
  pred <- as.numeric(frequency2ddG(predDiffMatrix))
  true <- as.numeric(frequency2ddG(trueDiffMatrix))
  predTrue <- data.frame(true = true, pred = pred)
  dddGTrue[[length(dddGTrue) + 1]] <- trueDiffMatrix
  dddGpred[[length(dddGpred) + 1]] <- predDiffMatrix
  plotPredTrue(predTrue, main = paste0('ΔΔΔG comparison for ',unlist(lapply(trueMotifs, function(x) x$name))[i], ' with confidence: ', round(mutationPredConfidence[i],4)))
  R2ListbaseLine <- c(R2ListbaseLine, groupedR2(predTrue$true, predTrue$pred, member = nrow(predTrue)))
}

sum(R2ListbaseLine>0)/length(R2ListbaseLine)
mean(R2ListbaseLine)
sum(R2ListbaseLine>0.9)/length(R2ListbaseLine)
```

### corrlation between experimental and prediction R2s 
```{r}
plot(R2ListbaseLine, R2List)
abline(lm(y~x, data = data.frame(x = R2ListbaseLine, y = R2List)), col = 'red')
legend('top', paste0('R^2 = ', round(groupedR2(R2ListbaseLine,
                                                    R2List),4)), bty = 'n')

sum((R2ListbaseLine == 0) + (R2List == 0) == 2)
sum(R2ListbaseLine[R2List == 0] < 0.1)
R2ListbaseLine
```



```{r}

evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
i = 1
for(pos in motifPoses){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('FamilyCode at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  
  #plot(mutationPredConfidence,groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F), ylab = 'Predict-True R^2', xlab = 'Confidence score', main = paste0('Confidence at ', pos))
  #legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
  #                                                   mutationPredConfidence, ),4)), bty = 'n')
  #dt <- data.frame(conf = mutationPredConfidence, R2 = groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F))
  #lm <- lm(R2~conf, data= dt)
  #abline(lm, col= 'red')
  
  plotPredTrue(predTrue.PFM, main = paste0('FamilyCode PFM at ',pos), xlab = 'Expermental', ylab = 'Predicted')
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T,throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  i <- i + 1
}
evaMatrix_BaseLine_dddG <- evaMatrix
```
### prediction R2 for large efects (experimental dddG > +-1)

```{r}

evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
i = 1
for(pos in motifPoses){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
  trueMagnitude <- c()
  predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
  predMagnitude <- c()
  for(set in 1:ncol(trueMatrix)){
    trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
    predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
  }
  
  plot(trueMagnitude, groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T),
       main = 'Experimental Maginitude vs Prediction accuracy', xlab = 'magnitude', ylab = 'R2')
  abline(lm(y~x, data = data.frame(x = trueMagnitude, y = groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T))), col = 'red')
  legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
                                                    trueMagnitude ),4)), bty = 'n')
  plot(predMagnitude, groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T),
       main = 'Prediction Maginitude vs Prediction accuracy', xlab = 'magnitude', ylab = 'R2')
  abline(lm(y~x, data = data.frame(x = predMagnitude, y = groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T))), col = 'red')
  legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
                                                    predMagnitude ),4)), bty = 'n')
  
  predTrue <- predTrue[rep(trueMagnitude > 1.5, each = 4),]
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('FamilyCode at ',pos))
  
  plotPredTrue(predTrue.PFM, main = paste0('FamilyCode PFM at ',pos), xlab = 'Expermental', ylab = 'Predicted')
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T,throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  i <- i + 1
}
evaMatrix_BaseLine_dddG_largeChange <- evaMatrix
```
### inspect R2 with different magnitudes 
500/500
```{r}
library(pROC)

blues <- colorRampPalette(c("lightblue", "blue", "darkblue"))
blue_spectrum <- blues(18)

R2TrueCutOff <- c()
R2PredCutOff <- c()
AUCs <- c()
for(t in 0:18){
  threshold <- t/4
  predTruePredT <- data.frame(NULL)
  predTrueTrueT <- data.frame(NULL)
  trueMagnitudeAll <- c()
  predMagnitudeAll <- c()
  for(pos in motifPoses){
    pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
    true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
    predTrue <- data.frame(true = true, pred = pred)
    predTrue <- predTrue[!is.na(predTrue$pred),]
    trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
    trueMagnitude <- c()
    predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
    predMagnitude <- c()
    for(set in 1:ncol(trueMatrix)){
      trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
      predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
    }
    trueMagnitudeAll <- c(trueMagnitudeAll, trueMagnitude)
    predMagnitudeAll <- c(predMagnitudeAll, predMagnitude)
    predTrueP <- predTrue[rep(predMagnitude > threshold, each = 4),]
    predTrueT <- predTrue[rep(trueMagnitude > threshold, each = 4),]
    predTrueTrueT <- rbind.data.frame(predTrueTrueT, predTrueT)
    predTruePredT <- rbind.data.frame(predTruePredT, predTrueP)
  }
  R2TrueCutOff <- c(R2TrueCutOff, round(groupedR2(predTrueTrueT$pred, 
        predTrueTrueT$true, nrow(predTrueTrueT), throughZero = T), 
        4))
  R2PredCutOff <- c(R2PredCutOff, round(groupedR2(predTruePredT$pred, 
        predTruePredT$true, nrow(predTruePredT), throughZero = T), 
        4))
  if(t == 0){
    suppressMessages(roc_obj1 <- roc(trueMagnitudeAll > threshold, predMagnitudeAll))
    plot(roc_obj1, col = blue_spectrum[t+1], lwd = 2, main = "BaseLine ROC curve", xlim = c(1,0),  ylim = c(0,1))
  }else{
    suppressMessages(roc_obj1 <- roc(trueMagnitudeAll > threshold, predMagnitudeAll))
    plot(roc_obj1, col = blue_spectrum[t+1], lwd = 2, xlim = c(1,0),  ylim = c(0,1), add = T)
  }
  AUCs <- c(AUCs, roc_obj1$auc)
}
legend("bottomright", legend = c('threshold = 0', 'threshold = 6.75',''), col = c(blue_spectrum[c(1,18)],'white'), lwd = 2, bty = 'n')
legend("topleft", legend = paste0('\nMax AUC: ', round(max(AUCs),4), '\nMean AUC: ', round(mean(AUCs),4)), bty = 'n')
plot((0:18)/4, R2PredCutOff, type = 'l', ylim = c(0,1))
lines((0:18)/4, R2TrueCutOff, type = 'l', ylim = c(0,1), col = 'red')
```

### screening over different thresholds sync. between true and pred sets.
500/500
```{r}
bins <- 50
trueMagnitudeAll <- c()
predMagnitudeAll <- c()
for(pos in motifPoses){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
  trueMagnitude <- c()
  predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
  predMagnitude <- c()
  for(set in 1:ncol(trueMatrix)){
    trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
    predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
  }
  trueMagnitudeAll <- c(trueMagnitudeAll, trueMagnitude)
  predMagnitudeAll <- c(predMagnitudeAll, predMagnitude)
}

max <- max(trueMagnitudeAll)

FPR <- c()
TPR <- c()
for(t in 0:bins){
  threshold <- max/bins*t*1.001
  classifiedPredTrue <- data.frame(true = paste0('Emperical.',trueMagnitudeAll > threshold), pred = paste0('predicted.',predMagnitudeAll > threshold))
  classifiedPredTrue$true <- factor(classifiedPredTrue$true, levels = c('Emperical.TRUE', 'Emperical.FALSE'))
  classifiedPredTrue$pred <- factor(classifiedPredTrue$pred, levels = c('predicted.TRUE', 'predicted.FALSE'))
  conf_matrix <- table(classifiedPredTrue$true, classifiedPredTrue$pred)
  FPR <- c(FPR, conf_matrix['Emperical.FALSE', 'predicted.TRUE'] / (conf_matrix['Emperical.FALSE', 'predicted.TRUE'] + conf_matrix['Emperical.FALSE', 'predicted.FALSE']))
  TPR <- c(TPR, conf_matrix['Emperical.TRUE', 'predicted.TRUE'] / (conf_matrix['Emperical.TRUE', 'predicted.TRUE'] + conf_matrix['Emperical.TRUE', 'predicted.FALSE']))
  FPR[is.na(FPR)] <- 0
  TPR[is.na(TPR)] <- 0
}
ROC_Baseline <- data.frame(FPR = FPR, TPR = TPR)

```

### output mutated TF fasta for rCLAMPS

```{r}
gene_ids <- unique(mutationInfo$TF)
searchUnitTemplate <- '(gene:"GENE" AND "Homo_sapiens")'
searchQ <- gsub('GENE', gene_ids[1], searchUnitTemplate)
for(i in 1:length(gene_ids)){#nrow(gene_ids)
  addQ <- gsub('GENE', gene_ids[i], searchUnitTemplate)
  searchQ <- paste0(searchQ, ' OR ', addQ)
}
cat(searchQ)
```

```{r}
geneInfoUniprot <- read.delim(paste0(workDir,"UniprotProteinSequence.tsv"))
geneInfoUniprot$Gene.Names <- paste0(geneInfoUniprot$Gene.Names, ' ')

geneInfo <- data.frame()
for(i in 1:length(gene_ids)){
  add <- geneInfoUniprot[grep(paste0(gene_ids[i],' '), geneInfoUniprot$Gene.Names),]
  if(nrow(add) != 0){
    add <- add[nchar(add$Sequence) == max(nchar(add$Sequence)),][1,]
    addSeq <- add$Sequence
  }else{
    addSeq <- NA 
  }
  addLine <- data.frame(name = gene_ids[i], seq = addSeq)
  geneInfo <- rbind.data.frame(geneInfo, addLine)
}

geneInfo <- geneInfo[!is.na(geneInfo$seq),]


outFasta <- data.frame(NULL)
for(i in mutantID){
  wtSeq <- mutationInfo$alignment[i]
  substr(wtSeq, mutationInfo$position[i]-1, mutationInfo$position[i]-1) <- mutationInfo$mutateFrom[i]
  wtSeq <- gsub('-','',wtSeq)
  mutSeq <- gsub(wtSeq, gsub('-','',mutationInfo$alignment[i]), geneInfo[geneInfo$name == mutationInfo$TF[i],'seq'])
  addLine <- data.frame(name = mutationInfo$name[i], seq = mutSeq)
  outFasta <- rbind.data.frame(outFasta, addLine)
}


writeFasta(outFasta, 'KockMutantHD.fasta')
```
perform rCLAMPS prediction

### inspect rCLAMPS predicted motifs 
```{r}
rCLAMPS_predicted_pwms <- read.delim("E:/Desktop/FamilyCode/KockHDPBM/rCLAMPS_predicted_pwms.txt")
rCLAMPSTFs <- mutationInfo$name[mutationInfo$mutation != 'WT']
predMotifs <- list()
for(i in 1:length(rCLAMPSTFs)){
  motifInfo <- rCLAMPS_predicted_pwms[rCLAMPS_predicted_pwms$prot == rCLAMPSTFs[i],]
  matrix <- matrix(ncol = 6, nrow = 4, data = motifInfo$prob, byrow = F)
  rownames(matrix) <- DNA()
  colnames(matrix) <- motifPoses[3:8]
  predMotifs[[length(predMotifs)+1]] <- list(name = motifInfo$prot[1], matrix = matrix)
}

key <- match(rCLAMPSTFs, Kock_alignment$name)
trueMotifs <- Kock_motifs_set[key]

length(predMotifs) == length(trueMotifs)

```

```{r}
evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
i = 1
for(pos in motifPoses[3:8]){
  pred <- unlist(lapply(predMotifs, function(x) frequency2ddG(x$matrix)[,pos]))
  true <- unlist(lapply(trueMotifs, function(x) frequency2ddG(x$matrix)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('Closest sequence at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  
  plotPredTrue(predTrue.PFM, main = paste0('Closest sequence PFM at ',pos), xlab = 'Expermental', ylab = 'Predicted')
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T,throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  i <- i + 1
}
evaMatrix_rCLAMPS <- evaMatrix
```

### test rCLAMPS dddG prediction
```{r}
key <- match(rCLAMPSTFs, Kock_alignment$name[mutantID])
trueStandard2 <- trueStandard[key]


dddGTrue <- list()
dddGpred <- list()
R2List_rCLAMPS <- c()
for(i in 1:length(trueMotifs)){
  predDiffTetra <- matrix2tetrahedron(predMotifs[[i]]$matrix) - matrix2tetrahedron(trueStandard2[[i]][,3:8])
  if(sum(predDiffTetra) == 0){
    next
  }
  trueDiffTetra <- matrix2tetrahedron(trueMotifs[[i]]$matrix[,3:8]) - matrix2tetrahedron(trueStandard2[[i]][,3:8])
  predDiffMatrix <- tetrahedron2matrix(predDiffTetra)
  trueDiffMatrix <- tetrahedron2matrix(trueDiffTetra)
  predDiffMatrix[predDiffMatrix < 0] <- 0.01
  trueDiffMatrix[trueDiffMatrix < 0] <- 0.01
  rownames(trueDiffMatrix) <- DNA()
  rownames(predDiffMatrix) <- DNA()
  pred <- as.numeric(frequency2ddG(predDiffMatrix))
  true <- as.numeric(frequency2ddG(trueDiffMatrix))
  predTrue <- data.frame(true = true, pred = pred)
  dddGTrue[[length(dddGTrue) + 1]] <- trueDiffMatrix
  dddGpred[[length(dddGpred) + 1]] <- predDiffMatrix
  plotPredTrue(predTrue, main = paste0('ΔΔΔG comparison for ',unlist(lapply(trueMotifs, function(x) x$name))[i]))
  R2List_rCLAMPS <- c(R2List_rCLAMPS, groupedR2(predTrue$true, predTrue$pred, member = nrow(predTrue)))
}
sum(R2List_rCLAMPS > 0) 
sum(R2List_rCLAMPS > 0.42) 
```

```{r}
plot.ecdf(R2List_rCLAMPS)

plot.ecdf(R2List, add = T, col = 'red')
plot(mutationInfo$position[mutantID][key],R2List_rCLAMPS)
R2Position <- data.frame(position = mutationInfo$position[mutantID][key],R2List_rCLAMPS)
byPositionR2 <- data.frame(NULL)
for(i in unique(R2Position$position)){
  R2 <- mean(R2Position$R2List[R2Position$position == i])
  addLine <- data.frame(position = i, R2 = R2)
  byPositionR2 <- rbind.data.frame(byPositionR2, addLine)
}
dplyr::arrange(byPositionR2, desc(R2))
```

```{r}

evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
i = 1
for(pos in motifPoses[3:8]){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('FamilyCode at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  
  plotPredTrue(predTrue.PFM, main = paste0('FamilyCode PFM at ',pos), xlab = 'Expermental', ylab = 'Predicted')
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T,throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  i <- i + 1
}
evaMatrix_rCLAMPS_dddG <- evaMatrix
```


### prediction R2 for large efects (experimental dddG > +-1)

```{r}

evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
i = 1
for(pos in motifPoses[3:8]){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
  trueMagnitude <- c()
  predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
  predMagnitude <- c()
  for(set in 1:ncol(trueMatrix)){
    trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
    predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
  }
  
  plot(trueMagnitude, groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T),
       main = 'Experimental Maginitude vs Prediction accuracy', xlab = 'magnitude', ylab = 'R2')
  abline(lm(y~x, data = data.frame(x = trueMagnitude, y = groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T))), col = 'red')
  legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
                                                    trueMagnitude ),4)), bty = 'n')
  plot(predMagnitude, groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T),
       main = 'Prediction Maginitude vs Prediction accuracy', xlab = 'magnitude', ylab = 'R2')
  abline(lm(y~x, data = data.frame(x = predMagnitude, y = groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T))), col = 'red')
  legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
                                                    predMagnitude ),4)), bty = 'n')
  
  predTrue <- predTrue[rep(trueMagnitude > 1.5, each = 4),]
  if(nrow(predTrue) == 0){
    evaMatrix[,pos] <- c(0,0,0,0)
    next
  }
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('rCLAMPS at ',pos))
  
  plotPredTrue(predTrue.PFM, main = paste0('rCLAMPS PFM at ',pos), xlab = 'Expermental', ylab = 'Predicted')
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T,throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  i <- i + 1
}
evaMatrix_rCLAMPS_dddG_largeChange <- evaMatrix
```

### inspect R2 with different magnitudes 

```{r}
blues <- colorRampPalette(c("lightblue", "blue", "darkblue"))
blue_spectrum <- blues(18)

R2TrueCutOff <- c()
R2PredCutOff <- c()
AUCs <- c()
for(t in 0:18){
  threshold <- t/4
  predTruePredT <- data.frame(NULL)
  predTrueTrueT <- data.frame(NULL)
  trueMagnitudeAll <- c()
  predMagnitudeAll <- c()
  for(pos in motifPoses[3:8]){
    pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
    true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
    predTrue <- data.frame(true = true, pred = pred)
    predTrue <- predTrue[!is.na(predTrue$pred),]
    trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
    trueMagnitude <- c()
    predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
    predMagnitude <- c()
    for(set in 1:ncol(trueMatrix)){
      trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
      predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
    }
    trueMagnitudeAll <- c(trueMagnitudeAll, trueMagnitude)
    predMagnitudeAll <- c(predMagnitudeAll, predMagnitude)
    predTrueP <- predTrue[rep(predMagnitude > threshold, each = 4),]
    predTrueT <- predTrue[rep(trueMagnitude > threshold, each = 4),]
    predTrueTrueT <- rbind.data.frame(predTrueTrueT, predTrueT)
    predTruePredT <- rbind.data.frame(predTruePredT, predTrueP)
  }
  R2TrueCutOff <- c(R2TrueCutOff, round(groupedR2(predTrueTrueT$pred, 
        predTrueTrueT$true, nrow(predTrueTrueT), throughZero = T), 
        4))
  R2PredCutOff <- c(R2PredCutOff, round(groupedR2(predTruePredT$pred, 
        predTruePredT$true, nrow(predTruePredT), throughZero = T), 
        4))
  if(t == 0){
    suppressMessages(roc_obj1 <- roc(trueMagnitudeAll > threshold, predMagnitudeAll))
    plot(roc_obj1, col = blue_spectrum[t+1], lwd = 2, main = "rCLAMPS ROC curve", xlim = c(1,0),  ylim = c(0,1))
  }else{
    suppressMessages(roc_obj1 <- roc(trueMagnitudeAll > threshold, predMagnitudeAll))
    plot(roc_obj1, col = blue_spectrum[t+1], lwd = 2, xlim = c(1,0),  ylim = c(0,1), add = T)
  }
  AUCs <- c(AUCs, roc_obj1$auc)
}
legend("bottomright", legend = c('threshold = 0', 'threshold = 6.75',''), col = c(blue_spectrum[c(1,18)],'white'), lwd = 2, bty = 'n')
legend("topleft", legend = paste0('\nMax AUC: ', round(max(AUCs),4), '\nMean AUC: ', round(mean(AUCs),4)), bty = 'n')
plot((0:18)/4, R2PredCutOff, type = 'l', ylim = c(0,1))
lines((0:18)/4, R2TrueCutOff, type = 'l', ylim = c(0,1), col = 'red')
```

### screening over different thresholds sync. between true and pred sets.
500/500
```{r}
bins <- 50
trueMagnitudeAll <- c()
predMagnitudeAll <- c()
for(pos in motifPoses[3:8]){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
  trueMagnitude <- c()
  predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
  predMagnitude <- c()
  for(set in 1:ncol(trueMatrix)){
    trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
    predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
  }
  trueMagnitudeAll <- c(trueMagnitudeAll, trueMagnitude)
  predMagnitudeAll <- c(predMagnitudeAll, predMagnitude)
}

max <- max(trueMagnitudeAll)

FPR <- c()
TPR <- c()
for(t in 0:bins){
  threshold <- max/bins*t*1.001
  classifiedPredTrue <- data.frame(true = paste0('Emperical.',trueMagnitudeAll > threshold), pred = paste0('predicted.',predMagnitudeAll > threshold))
  classifiedPredTrue$true <- factor(classifiedPredTrue$true, levels = c('Emperical.TRUE', 'Emperical.FALSE'))
  classifiedPredTrue$pred <- factor(classifiedPredTrue$pred, levels = c('predicted.TRUE', 'predicted.FALSE'))
  conf_matrix <- table(classifiedPredTrue$true, classifiedPredTrue$pred)
  FPR <- c(FPR, conf_matrix['Emperical.FALSE', 'predicted.TRUE'] / (conf_matrix['Emperical.FALSE', 'predicted.TRUE'] + conf_matrix['Emperical.FALSE', 'predicted.FALSE']))
  TPR <- c(TPR, conf_matrix['Emperical.TRUE', 'predicted.TRUE'] / (conf_matrix['Emperical.TRUE', 'predicted.TRUE'] + conf_matrix['Emperical.TRUE', 'predicted.FALSE']))
  FPR[is.na(FPR)] <- 0
  TPR[is.na(TPR)] <- 0
}
ROC_rCLAMPS <- data.frame(FPR = FPR, TPR = TPR)

```

### inspect deepPBS predicted motifs 
### load deepPBS motifs
```{r}
deepPBSdir <- 'deepPBS/HDKock/results/'
deepPBSmotifFiles <- list.files(deepPBSdir, pattern = '.tsv')

deepPBSTFs <- gsub('.tsv','',deepPBSmotifFiles)
deepPBSTFs <- gsub('_','-',deepPBSTFs)
deepPBSdMotifs <- list()
for(i in 1:length(deepPBSmotifFiles)){
  PBSmodels <- read.delim(paste0(deepPBSdir, deepPBSmotifFiles[i]), header=FALSE)
  PBSmotif <- t(PBSmodels)
  motif <- apply(PBSmotif, 2, function(x) x/max(x))
  rownames(motif) <- DNA()
  add <- list()
  add$gene_symbol <- deepPBSTFs[i]
  add$study <- i
  add$mode <- 1
  add$matrix <- as.matrix(motif)
  deepPBSdMotifs[[length(deepPBSdMotifs)+1]] <- add
  add <- list()
  #add reverse compliment motif
  add$gene_symbol <- deepPBSTFs[i]
  add$study <- i
  add$mode <- -1
  if(ncol(motif) == 0){
    rev <- motif
  }else{
    rev <- as.matrix(motif)[nrow(motif):1, ncol(motif):1]
  }
  rownames(rev) <- DNA()
  add$matrix <- rev
  deepPBSdMotifs[[length(deepPBSdMotifs)+1]] <- add
}


```


### Load Homeo Domain 8mer seed of NNTRAYNN and score motifs with the seed
```{r}
#align to NNTDAYNN motif
HDseed <- matrix(nrow = 4,ncol = 8,data = c(0,0,0,0,
                                            0,0,0,0,
                                            0,0,0,1,
                                            1,0,1,1,
                                            1,0,0,0,
                                            0,1,0,1,
                                            0,0,0,0,
                                            0,0,0,0))
rownames(HDseed) <- DNA()

motifScores <- scoreMotifList(deepPBSdMotifs, HDseed, weight = c(0,0,1,1,1,1,0,0))
```

### Filter and reorder motif information according to motif matching score.
The motif (forward or reverse compliment) with the lowest motif matching score of each entry will be kept. For the same protein, entries are arranged from lower score to higher score
```{r}
idstds <- unique(motifScores[,1:2])
all_motifs <- data.frame(NULL)
for(i in 1:nrow(idstds)){
  fDT <- motifScores[motifScores$gene_symbol == idstds[i,1],]
  fDT <- fDT[fDT$study == idstds[i,2],]
  all_motifs <- rbind.data.frame(all_motifs, fDT[which.min(fDT$score),])
}

all_motifs$study <- unlist(lapply(strsplit(all_motifs$gene_symbol,'_'), function(x) x[2]))
all_motifs$gene_symbol <- unlist(lapply(strsplit(all_motifs$gene_symbol,'_'), function(x) x[1]))
motifs_arranged <- data.frame(NULL)
motifScore <- c()
for(i in 1:length(unique(all_motifs$gene_symbol))){
  add <- all_motifs[all_motifs$gene_symbol == unique(all_motifs$gene_symbol)[i],]
  add <- dplyr::arrange(add, score)
  motifScore <- c(motifScore,add$score[1])
  motifs_arranged <- rbind.data.frame(motifs_arranged, add)
}
#Plot distribution of lowest motif matching score
plot.ecdf(as.numeric(motifScore))
```

```{r}
deepPBS_motifs_set <- filterMotifList(motifs_arranged, deepPBSdMotifs, 8, motifPoses)
predMotifs <- deepPBS_motifs_set
key <- match(deepPBSTFs, tolower(Kock_alignment$name))
trueMotifs <- Kock_motifs_set[key]
```


```{r}
evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
i = 1
for(pos in motifPoses){
  pred <- unlist(lapply(predMotifs, function(x) frequency2ddG(x$matrix)[,pos]))
  true <- unlist(lapply(trueMotifs, function(x) frequency2ddG(x$matrix)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('Closest sequence at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  
  plotPredTrue(predTrue.PFM, main = paste0('Closest sequence PFM at ',pos), xlab = 'Expermental', ylab = 'Predicted')
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T,throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  i <- i + 1
}
evaMatrix_deepPBS <- evaMatrix
```

### test deepPBS dddG prediction
```{r}
key <- match(deepPBSTFs, tolower(mutationInfo$name[mutantID]))
trueStandard2 <- trueStandard[key]


dddGTrue <- list()
dddGpred <- list()
R2List_deepPBS <- c()
for(i in 1:length(trueMotifs)){
  predDiffTetra <- matrix2tetrahedron(predMotifs[[i]]$matrix) - matrix2tetrahedron(trueStandard2[[i]])
  if(sum(predDiffTetra) == 0){
    next
  }
  trueDiffTetra <- matrix2tetrahedron(trueMotifs[[i]]$matrix) - matrix2tetrahedron(trueStandard2[[i]])
  predDiffMatrix <- tetrahedron2matrix(predDiffTetra)
  trueDiffMatrix <- tetrahedron2matrix(trueDiffTetra)
  predDiffMatrix[predDiffMatrix < 0] <- 0.01
  trueDiffMatrix[trueDiffMatrix < 0] <- 0.01
  rownames(trueDiffMatrix) <- DNA()
  rownames(predDiffMatrix) <- DNA()
  pred <- as.numeric(frequency2ddG(predDiffMatrix))
  true <- as.numeric(frequency2ddG(trueDiffMatrix))
  predTrue <- data.frame(true = true, pred = pred)
  dddGTrue[[length(dddGTrue) + 1]] <- trueDiffMatrix
  dddGpred[[length(dddGpred) + 1]] <- predDiffMatrix
  plotPredTrue(predTrue, main = paste0('ΔΔΔG comparison for ',unlist(lapply(trueMotifs, function(x) x$name))[i]))
  R2List_deepPBS <- c(R2List_deepPBS, groupedR2(predTrue$true, predTrue$pred, member = nrow(predTrue)))
}
sum(R2List_deepPBS>0)
mean(R2List_deepPBS)
sum(R2List_deepPBS>0.41)
```

```{r}
plot.ecdf(R2ListbaseLine)

plot.ecdf(R2List, add = T, col = 'red')
plot.ecdf(R2List_rCLAMPS, add = T, col = 'blue')
plot.ecdf(R2List_deepPBS, add = T, col = 'green')

```

```{r}

evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
i = 1
for(pos in motifPoses){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('FamilyCode at ',pos))
  boxplot(groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F), ylim = c(0,1),
        main = paste0(pos,' Mean R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, 
                                                          member = 4, mean= T),4)))
  
  plotPredTrue(predTrue.PFM, main = paste0('FamilyCode PFM at ',pos), xlab = 'Expermental', ylab = 'Predicted')
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T,throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  i <- i + 1
}
evaMatrix_deepPBS_dddG <- evaMatrix
```
### prediction R2 for large efects (experimental dddG > +-1)

```{r}

evaMatrix <- matrix(nrow = 4, ncol = length(motifPoses), data = 0)
colnames(evaMatrix) <- motifPoses
i = 1
for(pos in motifPoses){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
  trueMagnitude <- c()
  predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
  predMagnitude <- c()
  for(set in 1:ncol(trueMatrix)){
    trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
    predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
  }
  
  plot(trueMagnitude, groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T),
       main = 'Experimental Maginitude vs Prediction accuracy', xlab = 'magnitude', ylab = 'R2')
  abline(lm(y~x, data = data.frame(x = trueMagnitude, y = groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T))), col = 'red')
  legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
                                                    trueMagnitude ),4)), bty = 'n')
  plot(predMagnitude, groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T),
       main = 'Prediction Maginitude vs Prediction accuracy', xlab = 'magnitude', ylab = 'R2')
  abline(lm(y~x, data = data.frame(x = predMagnitude, y = groupedR2(predTrue$pred, predTrue$true, member = 4, mean= F,throughZero = T))), col = 'red')
  legend('bottomright', paste0('R^2 = ', round(groupedR2(groupedR2(predTrue$pred, predTrue$true,member = 4, mean = F),
                                                    predMagnitude ),4)), bty = 'n')
  
  predTrue <- predTrue[rep(trueMagnitude > 1.5, each = 4),]
  if(nrow(predTrue) == 0){
    evaMatrix[,pos] <- c(0,0,0,0)
    next
  }
  predTrue.PFM <- predTrue.ddG2frequency(predTrue, PFM = T)
  plotPredTrue(predTrue, main = paste0('rCLAMPS at ',pos))
  
  plotPredTrue(predTrue.PFM, main = paste0('rCLAMPS PFM at ',pos), xlab = 'Expermental', ylab = 'Predicted')
  hit <- sum(round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2),4) > 0.5)/(nrow(predTrue.PFM)/4)
  boxplot(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, member = 4, mean= F,throughZero = F)^(1/2), ylim = c(0,1),
        main = paste0(pos, ' Mean PCC = ', round(groupedR2(predTrue.PFM$pred, predTrue.PFM$true, 
                                                     member = 4, mean= T,throughZero = F)^(1/2),4), ' Accuracy = ', round(hit,4)))
  evaMatrix[,pos] <- c(round(groupedR2(predTrue$pred, 
        predTrue$true, nrow(predTrue), throughZero = T), 
        4), round(RMSD(predTrue$pred, predTrue$true), 
        4), round(groupedR2(predTrue.PFM$pred, 
        predTrue.PFM$true, nrow(predTrue.PFM), throughZero = F), 
        4)^(1/2), round(hit,4))
  i <- i + 1
}
evaMatrix_deepPBS_dddG_largeChange <- evaMatrix
```


### inspect R2 with different magnitudes 

```{r}
blues <- colorRampPalette(c("lightblue", "blue", "darkblue"))
blue_spectrum <- blues(18)

R2TrueCutOff <- c()
R2PredCutOff <- c()
AUCs <- c()
for(t in 0:18){
  threshold <- t/4
  predTruePredT <- data.frame(NULL)
  predTrueTrueT <- data.frame(NULL)
  trueMagnitudeAll <- c()
  predMagnitudeAll <- c()
  for(pos in motifPoses){
    pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
    true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
    predTrue <- data.frame(true = true, pred = pred)
    predTrue <- predTrue[!is.na(predTrue$pred),]
    trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
    trueMagnitude <- c()
    predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
    predMagnitude <- c()
    for(set in 1:ncol(trueMatrix)){
      trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
      predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
    }
    trueMagnitudeAll <- c(trueMagnitudeAll, trueMagnitude)
    predMagnitudeAll <- c(predMagnitudeAll, predMagnitude)
    predTrueP <- predTrue[rep(predMagnitude > threshold, each = 4),]
    predTrueT <- predTrue[rep(trueMagnitude > threshold, each = 4),]
    predTrueTrueT <- rbind.data.frame(predTrueTrueT, predTrueT)
    predTruePredT <- rbind.data.frame(predTruePredT, predTrueP)
  }
  R2TrueCutOff <- c(R2TrueCutOff, round(groupedR2(predTrueTrueT$pred, 
        predTrueTrueT$true, nrow(predTrueTrueT), throughZero = T), 
        4))
  R2PredCutOff <- c(R2PredCutOff, round(groupedR2(predTruePredT$pred, 
        predTruePredT$true, nrow(predTruePredT), throughZero = T), 
        4))
  if(t == 0){
    suppressMessages(roc_obj1 <- roc(trueMagnitudeAll > threshold, predMagnitudeAll))
    plot(roc_obj1, col = blue_spectrum[t+1], lwd = 2, main = "deepPBS  ROC curve", xlim = c(1,0),  ylim = c(0,1))
  }else{
    suppressMessages(roc_obj1 <- roc(trueMagnitudeAll > threshold, predMagnitudeAll))
    plot(roc_obj1, col = blue_spectrum[t+1], lwd = 2, xlim = c(1,0),  ylim = c(0,1), add = T)
  }
  AUCs <- c(AUCs, roc_obj1$auc)
}
legend("bottomright", legend = c('threshold = 0', 'threshold = 6.75',''), col = c(blue_spectrum[c(1,18)],'white'), lwd = 2, bty = 'n')
legend("topleft", legend = paste0('\nMax AUC: ', round(max(AUCs),4), '\nMean AUC: ', round(mean(AUCs),4)), bty = 'n')
plot((0:18)/4, R2PredCutOff, type = 'l', ylim = c(0,1))
lines((0:18)/4, R2TrueCutOff, type = 'l', ylim = c(0,1), col = 'red')
```

### screening over different thresholds sync. between true and pred sets.
500/500
```{r}
bins <- 50
trueMagnitudeAll <- c()
predMagnitudeAll <- c()
for(pos in motifPoses){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
  trueMagnitude <- c()
  predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
  predMagnitude <- c()
  for(set in 1:ncol(trueMatrix)){
    trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
    predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
  }
  trueMagnitudeAll <- c(trueMagnitudeAll, trueMagnitude)
  predMagnitudeAll <- c(predMagnitudeAll, predMagnitude)
}

max <- max(trueMagnitudeAll)

FPR <- c()
TPR <- c()
for(t in 0:bins){
  threshold <- max/bins*t*1.001
  classifiedPredTrue <- data.frame(true = paste0('Emperical.',trueMagnitudeAll > threshold), pred = paste0('predicted.',predMagnitudeAll > threshold))
  classifiedPredTrue$true <- factor(classifiedPredTrue$true, levels = c('Emperical.TRUE', 'Emperical.FALSE'))
  classifiedPredTrue$pred <- factor(classifiedPredTrue$pred, levels = c('predicted.TRUE', 'predicted.FALSE'))
  conf_matrix <- table(classifiedPredTrue$true, classifiedPredTrue$pred)
  FPR <- c(FPR, conf_matrix['Emperical.FALSE', 'predicted.TRUE'] / (conf_matrix['Emperical.FALSE', 'predicted.TRUE'] + conf_matrix['Emperical.FALSE', 'predicted.FALSE']))
  TPR <- c(TPR, conf_matrix['Emperical.TRUE', 'predicted.TRUE'] / (conf_matrix['Emperical.TRUE', 'predicted.TRUE'] + conf_matrix['Emperical.TRUE', 'predicted.FALSE']))
  FPR[is.na(FPR)] <- 0
  TPR[is.na(TPR)] <- 0
}
ROC_deepPBS <- data.frame(FPR = FPR, TPR = TPR)

```

### Show R2 comparison 
```{r}
R2table <- data.frame(BaseLine = evaMatrix_BL[1,], FamilyCode_genericModel = evaMatrix_FC[1,], FamilyCode_mutationSpecificModel = evaMatrix_FC_byPosition[1,], NearestNeibor_SimilarityRegression = evaMatrix_CS[1,], BaseLine_dddG = evaMatrix_BaseLine_dddG[1,], BaseLine_dddG_large = evaMatrix_BaseLine_dddG_largeChange[1,], FamilyCode_dddG = evaMatrix_FC_dddG[1,],FamilyCode_dddG_large = evaMatrix_FC_dddG_largeChange[1,], rCLAMPS = evaMatrix_rCLAMPS[1,], rCLAMPS_dddG = evaMatrix_rCLAMPS_dddG[1,], rCLAMPS_dddG_large = evaMatrix_rCLAMPS_dddG_largeChange[1,], deepPBS = evaMatrix_deepPBS[1,], deepPBS_dddG = evaMatrix_deepPBS_dddG[1,], deepPBS_dddG_large = evaMatrix_deepPBS_dddG_largeChange[1,])

R2table <- t(R2table)
R2table <- R2table[c(1,2,3,4,9,12,5,7,10,13,6,8,11,14),]
R2table
```

### Show PCC comparison 
```{r}
Acuracytable <- data.frame(BaseLine = evaMatrix_BL[4,], FamilyCode_genericModel = evaMatrix_FC[4,], FamilyCode_mutationSpecificModel = evaMatrix_FC_byPosition[4,], NearestNeibor_SimilarityRegression = evaMatrix_CS[4,], BaseLine_dddG = evaMatrix_BaseLine_dddG[4,], BaseLine_dddG_large = evaMatrix_BaseLine_dddG_largeChange[4,], FamilyCode_dddG = evaMatrix_FC_dddG[4,],FamilyCode_dddG_large = evaMatrix_FC_dddG_largeChange[4,], rCLAMPS = evaMatrix_rCLAMPS[4,], rCLAMPS_dddG = evaMatrix_rCLAMPS_dddG[4,], rCLAMPS_dddG_large = evaMatrix_rCLAMPS_dddG_largeChange[4,], deepPBS = evaMatrix_deepPBS[4,], deepPBS_dddG = evaMatrix_deepPBS_dddG[4,], deepPBS_dddG_large = evaMatrix_deepPBS_dddG_largeChange[4,])
t(Acuracytable)
```

### box plot of individual R2s
```{r}
boxplot(R2ListbaseLine, R2List, R2List_rCLAMPS, R2List_deepPBS, main = 'ΔΔΔG/RT prediciton R^2 comparison', ylab = 'R^2s', names = c('Base Line','Family Code', 'rCLAMPS', 'deepPBS'))

t.test(R2List_deepPBS, R2List_rCLAMPS)
```

```{r}
library(ggplot2)
library(ggpubr)
# Custom order for the groups

#rCmutpos <- which(!is.na(match(mutationInfo[mutationInfo$mutation != 'WT',]$position, (positions + 1))))
rCmutpos <- 1:92
data <- data.frame(
  R2 = c(R2ListbaseLine[rCmutpos], R2List[rCmutpos], R2List_rCLAMPS[rCmutpos], R2List_deepPBS[rCmutpos]),
  Group = factor(c(rep("Base Line",length(rCmutpos)), rep("Family Code", length(rCmutpos)),rep("rCLAMPS", length(rCmutpos)), rep("deepPBS",length(rCmutpos))))
)

data$Group <- factor(data$Group, levels = c('Base Line', 'Family Code', 'rCLAMPS', 'deepPBS'))

# Custom colors for the groups
custom_colors <- c('Base Line' = '#1f77b4', 
                   'Family Code' = '#ff7f0e', 
                   'rCLAMPS' = '#2ca02c', 
                   'deepPBS' = '#d61828')

# Create boxplot with custom colors and order
p <- ggplot(data, aes(x = Group, y = R2, fill = Group)) +
  geom_boxplot() +
  labs(title = "ΔΔΔG/RT Prediction R^2 Comparison", y = "R^2s", x = "") +
  scale_fill_manual(values = custom_colors) + # Apply custom colors
  theme_minimal() +
  theme(
    legend.position = "bottom",        # Place legend at the bottom
    legend.title = element_blank(),    # Remove legend title
    panel.grid.major = element_blank(),# Remove major grid lines
    panel.grid.minor = element_blank(),
    panel.border = element_blank(),# Remove minor grid lines
    panel.background = element_rect(fill = "white"), # White panel background
    plot.background = element_rect(color = "white"),
    plot.title = element_text(face = "bold", size = 14, hjust = 0.5)
    # White plot background
  )

# Add significance bars
p <- p + stat_compare_means(comparisons = list(
                                               c("Family Code", "rCLAMPS"),
                                               c("Family Code", "deepPBS")),
                            label = "p.signif",
                            method = "t.test") +
  theme(legend.position = "none")

# Print the plot
plot(p)
```


```{r}
#rCmutpos <- which(!is.na(match(mutationInfo[mutationInfo$mutation != 'WT',]$position, (positions + 1))))
rCmutpos <- 1:92
combined_data <- data.frame(
  Value = c(R2ListbaseLine[rCmutpos], R2List[rCmutpos], R2List_rCLAMPS[rCmutpos], R2List_deepPBS[rCmutpos]),
  Group = factor(c(rep("Base Line",length(rCmutpos)), rep("Family Code", length(rCmutpos)),rep("rCLAMPS", length(rCmutpos)), rep("deepPBS",length(rCmutpos))))
)

combined_data$Group <- factor(combined_data$Group, levels = c('Base Line', 'Family Code', 'rCLAMPS', 'deepPBS'))
# ECDF plot
p <- ggplot(combined_data, aes(x = Value, color = Group)) +
  stat_ecdf(size = 1) +  # ECDF plot with line size
  scale_color_manual(
    values = c("Base Line" = '#1f77b4', "Family Code" = '#ff7f0e', "rCLAMPS" =  '#2ca02c', 'deepPBS' = '#d61828')
  ) +
  theme_minimal() +
  labs(
    title = "Experimental and Predicted ΔΔΔG/RT",
    x = "R^2",
    y = "ECDF",
    color = "Method"  # Legend title
  ) +
  theme(
    legend.position = "bottom",        # Place legend at the bottom
    legend.title = element_blank(),    # Remove legend title
    panel.grid.major = element_blank(),# Remove major grid lines
    panel.grid.minor = element_blank(),# Remove minor grid lines
    panel.background = element_rect(fill = "white"), # White panel background
    plot.background = element_rect(color = "white"),   # White plot background
    plot.title = element_text(face = "bold", size = 14, hjust = 0.5)
  )
plot(p)
```

### plot ROC-like 
```{r}
ROC_Baseline <- rbind(ROC_Baseline, 1)
ROC_FC_rCSM <- rbind(ROC_FC_rCSM, 1)
ROC_Baseline$Dataset <- "Base Line"
ROC_FC_GM$Dataset <- "FamilyCode GenericModel"
ROC_FC_MS$Dataset <- "FamilyCode"
ROC_rCLAMPS$Dataset <- "rCLAMPS"
ROC_deepPBS$Dataset <- "deepPBS"
ROC_FC_rCSM$Dataset <- "rCSM"
ROC_FC_rCGM$Dataset <- "rCGM"

# Combine all data frames into one
combined_df <- rbind(ROC_Baseline,ROC_FC_MS, ROC_rCLAMPS, ROC_deepPBS, ROC_FC_rCGM, ROC_FC_rCSM)

ggplot(combined_df, aes(x = FPR, y = TPR, color = Dataset)) +
  geom_line(size = 1) +
  scale_color_manual(values = c("#1f77b4", "#d61828", "#ff7f0e","#2ca02c", 'yellow3','orange3')) +  # Customize colors
  coord_fixed(xlim = c(0, 1), ylim = c(0, 1)) +  # Set axis limits from 0 to 1
  labs(title = "ROC Curves at different ΔΔG range",
       x = "False Positive Rate (FPR)",
       y = "True Positive Rate (TPR)") +
  theme_minimal() +
  theme(plot.title = element_text(hjust = 0.5),
        legend.position = "bottom",        # Place legend at the bottom
        legend.title = element_blank(),
        panel.grid.major = element_blank(),# Remove major grid lines
    panel.grid.minor = element_blank(),# Remove minor grid lines
    panel.background = element_rect(fill = "white"), # White panel background
    plot.background = element_rect(color = "white"))
```

### overall R2 with baseline threshold
```{r}
predTruetotal <- data.frame(NULL)
for(pos in motifPoses){
  pred <- unlist(lapply(dddGpred, function(x) frequency2ddG(x)[,pos]))
  true <- unlist(lapply(dddGTrue, function(x) frequency2ddG(x)[,pos]))
  predTrue <- data.frame(true = true, pred = pred)
  predTrue <- predTrue[!is.na(predTrue$pred),]
  trueMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$true)
  trueMagnitude <- c()
  predMatrix <- matrix(nrow = 4, ncol = nrow(predTrue)/4, byrow = F, data = predTrue$pred)
  predMagnitude <- c()
  for(set in 1:ncol(trueMatrix)){
    trueMagnitude <- c(trueMagnitude, max(trueMatrix[,set]) - min(trueMatrix[,set]))
    predMagnitude <- c(predMagnitude, max(predMatrix[,set]) - min(predMatrix[,set]))
  }
  predTrue <- predTrue[rep(trueMagnitude > 1, each = 4),]
  predTruetotal <- rbind.data.frame(predTruetotal, predTrue)
}
plotPredTrue(predTruetotal, main = paste0('FamilyCode for range > ',1))
```
