---
title: "FamilyCode_plots"
output:
  pdf_document: default
  html_notebook: default
---
## Load data
```{r}
library(FamilyCode)
proBound_runDIR <- r'(HTdata/)'
bHLH_info <- read.csv(file = r"(bHLH_info.csv)", row.names = 1)
bHLH_ali <- read.table(r"(bHLH_pb.sto.stockholm)", quote="\"", fill = T)
proBound_files <- list.files(proBound_runDIR)
modelFile_Template <- paste0(proBound_runDIR, '$modelFile$/result/fit.models.consensus.json')
rows <- c("A", "C", "G", "T")
blank <- data.frame(P00 = c(1,1,1,1), P0 = c(1,1,1,1))

#getting expeirment identifier from bHLH_info, may vary according to different naming scheme
bHLH_index <- bHLH_info[,1:2]
bHLH_index$quality <- 0
for(i in 1:nrow(bHLH_index)){
  tryCatch({
    dir <- paste0(proBound_runDIR,bHLH_index$gene_symbol[i],'_',bHLH_index$study[i])
    suppressWarnings(quality <- readChar(paste0(dir,'//fitEval.txt'),nchars = 100))
    quality <- strsplit(quality, ', ')[[1]][1]
    quality <- strsplit(quality, ' = ')[[1]][2]
    bHLH_index$quality[i] <- quality
  }, error = function(e){})
}
Motif_Table <- loadMono_motifs(bHLH_index, modelFile_Template, rec_seq = 'CANNTG', pos_index = c('P-3','P-2','P-1','P1','P2','P3'), checkSymmetry = T, withTable = T)

all_mono_motifs <- Motif_Table$motifs
bHLH_model_info <- Motif_Table$table
bHLH_pbAlignment <- concatAli(bHLH_ali, 694,56)

#filter for symmetry
bHLH_sym <- bHLH_model_info[bHLH_model_info$symmetry == 1,]
#filter for best model of each protein
bHLH_names <- unique(bHLH_sym$gene_symbol)
bHLH_index <- data.frame(NULL)
for(i in 1:length(bHLH_names)){
  qualities <- bHLH_sym[bHLH_sym$gene_symbol == bHLH_names[i],]
  add <- qualities[qualities$quality == max(qualities$quality),]
  bHLH_index <- rbind.data.frame(bHLH_index, add[1,])
}
#filter for quality > 0.15
bHLH_index <- bHLH_index[bHLH_index$quality > 0.15,]


#reload mono_motifs
mono_motifs <- loadMono_motifs(bHLH_index, modelFile_Template, rec_seq = 'CANNTG', pos_index = c('P-3','P-2','P-1','P1','P2','P3'), checkSymmetry = F, withTable = F)
groups <- c('HWT', 'H5', 'H13','HD', 'AWT', 'A5', 'A13', 'AD')
SELEXFile_Template <- r'(SELEXdata/$modelFile$/fit.models.consensus.json)'
SELEX_info <- data.frame(gene_symbol = groups, study = 'pb')

SELEX_motifs <- loadMono_motifs(SELEX_info, SELEXFile_Template, rec_seq = 'NCANNTGN', pos_index = c('P-4','P-3','P-2','P-1','P1','P2','P3', 'P4'), checkSymmetry = T, withTable = F, useMode = c(1,2,1,1,1,1,1,1))

#save(mono_motifs, bHLH_pbAlignment, file = 'bHLH_training.Rdata')
```


# Figure 2
## Panal A
```{r}
library(eulerr)
vennTable <- unique(bHLH_info[,1:2])
vennTable <- table(vennTable)
vennTable <- vennTable == 1
plot(euler(vennTable[, 1:3], shape = "ellipse"), quantities = TRUE)
```
## Panel D 
```{r}
library(readr)
library(R.utils)

mkdirs('logos')
for(i in 1:nrow(bHLH_index)){
  JSON_Lines <- readLines(paste0("HTdata/", bHLH_index$gene_symbol[i], '_', bHLH_index$study[i],'/result/fit.models.consensus.json'))
  mono <- JSON2Matrix(JSON_Lines, mode = bHLH_index$motif_model[i])
  plot <- mononucleotide_logo(mono, axes = F)
  png(paste0('logos/',bHLH_index$gene_symbol[i],'.png'))
  plot(plot)
  dev.off()
}
```

# Figure 2
## Panal A
```{r}
library(plotly)
AAPosition <- 13
motifPosition <- 'P-1'
bHLH_pbAlignment <- matchAliMotif(mono_motifs, bHLH_pbAlignment)
AAbp_Matrix <- AAbpCombination(mono_motifs, bHLH_pbAlignment, AAPosition, motifPosition)
plot <- plot_tetrahedron(AAbp_Matrix, color = T, vertex = F, size = 10)
scene = list(camera = list(eye = list(x = 1.25, y = 1.25, z = 1.25)))
plot%>%layout(scene=scene)
```
## Panel B
```{r}
AAPosition <- 5
motifPosition <- 'P-1'
bHLH_pbAlignment <- matchAliMotif(mono_motifs, bHLH_pbAlignment)
AAbp_Matrix <- AAbpCombination(mono_motifs, bHLH_pbAlignment, AAPosition, motifPosition)
plot <- plot_tetrahedron(AAbp_Matrix, color = T, vertex = F, size = 10)
scene = list(camera = list(eye = list(x = 1.25, y = 1.25, z = 1.25)))
plot%>%layout(scene=scene)
```

## Panel C
```{r}
AAPosition <- 13
motifPosition <- 'P-3'
bHLH_pbAlignment <- matchAliMotif(mono_motifs, bHLH_pbAlignment)
AAbp_Matrix <- AAbpCombination(mono_motifs, bHLH_pbAlignment, AAPosition, motifPosition)
plot <- plot_tetrahedron(AAbp_Matrix, color = T, vertex = F, size = 10)
scene = list(camera = list(eye = list(x = 1.25, y = 1.25, z = 1.25)))
plot%>%layout(scene=scene)
```

## Panel D
```{r}
pvalTable <- getPvalTable(mono_motifs, Alignment = bHLH_pbAlignment, pos_index = c('P-3','P-2','P-1','P1','P2','P3'))
#apply -log(10)
pvalTable <- -log(pvalTable,10)
#set NAs to 0
pvalTable[is.na(pvalTable)] <- 0
#plot heatmap
gplots::heatmap.2(as.matrix(as.data.frame(lapply(pvalTable, as.numeric))),dendrogram='none',
                  Rowv=FALSE, Colv=FALSE,trace='none',col = rev(heat.colors(12)), key.title = '-logP-val')
```

## Panel E
```{r}
colorCode <- pvalTable$`P-1`
colorCode[is.na(colorCode)] <- median(colorCode[!is.na(colorCode)])
max <- max(colorCode)
colorList <- colorCode/max*255


complexSample <- 'bHLH_processed.pse'
addColorTemplate <- 'color 0x$colorCode$, polymer.protein and resi $resiNum$'
script <- paste('load ', complexSample, sep = '')
for(i in 1:length(colorList)){
  colorCode <- rgb(colorList[i], 0, (255-colorList[i]), maxColorValue = 255)
  colorCode <- gsub('#','',colorCode)
  addScript <- gsub('\\$colorCode\\$',colorCode,addColorTemplate)
  addScript <- gsub('\\$resiNum\\$', i, addScript)
  script <- c(script, addScript)
}
run.pymol(pymol.dir = 'path/to/PyMOLWin.exe',script = script)
```

# Figure 3
## Panel A
```{r}
library(plotly)
AAPosition <- 13
motifPosition <- 'P-1'
bHLH_pbAlignment <- matchAliMotif(mono_motifs, bHLH_pbAlignment)
AAbp_Matrix <- AAbpCombination(mono_motifs, bHLH_pbAlignment, AAPosition, motifPosition)
Rmatirx <- AAbp_Matrix[,colnames(AAbp_Matrix) == 'R']
RTM <- matrix2tetrahedron(Rmatirx)
meanPointR <- apply((RTM), 2, function(x) mean(x))
Rmean <- tetrahedron2matrix(meanPointR)
svd <- FamilyCode::matrixSVD(AAbp_Matrix)
Allmean <- tetrahedron2matrix(svd$tetra_mean)
AAbp_Matrix <- AAbp_Matrix[,colnames(AAbp_Matrix) == 'R']
AAbp_Matrix <- cbind.data.frame(AAbp_Matrix, X = Allmean)
connect <- cbind.data.frame(Rmean = meanPointR, Allmean = svd$tetra_mean)
scene = list(camera = list(eye = list(x = 1.25, y = 1.25, z = 1.25)))

plotTest <- plot_tetrahedron(AAbp_Matrix, color = T, size = 10)
plotTest%>%
  add_trace(connect,x = as.numeric(connect[1,1:2]),
            y=as.numeric(connect[2,1:2]),z= as.numeric(connect[3,1:2]),color = I('black'),
            type = 'scatter3d',
            mode = 'lines',
            line = list(width = 5),
            opacity = 1,
            showlegend = F
  )%>%
  add_trace(
    type= "cone",
    x= connect[1,1], y= connect[2,1], z= connect[3,1],
    u= connect[1,1] - connect[1,2], v= connect[2,1] - connect[2,2], w= connect[3,1] - connect[3,2],
    sizeref= 0.5,
    colorscale = list(list(0, as.character('black')),
                      list(1, as.character('black'))),
    anchor = 'tip',
    showscale= F
  )%>%layout(scene=scene)

rownames(Rmean) <- c('A','C','G','T')
mononucleotide_logo(Rmean)

```

## Panel B
```{r}
library(plotly)
AAPosition <- 13
motifPosition <- 'P-1'
bHLH_pbAlignment <- matchAliMotif(mono_motifs, bHLH_pbAlignment)
AAbp_Matrix <- AAbpCombination(mono_motifs, bHLH_pbAlignment, AAPosition, motifPosition)
Rmatirx <- AAbp_Matrix[,colnames(AAbp_Matrix) == 'V']
RTM <- matrix2tetrahedron(Rmatirx)
meanPointR <- apply((RTM), 2, function(x) mean(x))
Rmean <- tetrahedron2matrix(meanPointR)
svd <- FamilyCode::matrixSVD(AAbp_Matrix)
Allmean <- tetrahedron2matrix(svd$tetra_mean)
AAbp_Matrix <- AAbp_Matrix[,colnames(AAbp_Matrix) == 'V']
AAbp_Matrix <- cbind.data.frame(AAbp_Matrix, X = Allmean)
connect <- cbind.data.frame(Rmean = meanPointR, Allmean = svd$tetra_mean)
scene = list(camera = list(eye = list(x = 1.25, y = 1.25, z = 1.25)))

plotTest <- plot_tetrahedron(AAbp_Matrix, color = T, size = 10)
plotTest%>%
  add_trace(connect,x = as.numeric(connect[1,1:2]),
            y=as.numeric(connect[2,1:2]),z= as.numeric(connect[3,1:2]),color = I('black'),
            type = 'scatter3d',
            mode = 'lines',
            line = list(width = 10),
            opacity = 1,
            showlegend = F
  )%>%
  add_trace(
    type= "cone",
    x= connect[1,1], y= connect[2,1], z= connect[3,1],
    u= connect[1,1] - connect[1,2], v= connect[2,1] - connect[2,2], w= connect[3,1] - connect[3,2],
    sizeref= 0.5,
    colorscale = list(list(0, as.character('black')),
                      list(1, as.character('black'))),
    anchor = 'tip',
    showscale= F
  )%>%layout(scene=scene)

rownames(Rmean) <- c('A','C','G','T')
mononucleotide_logo(log(Rmean))

```

# Figure 4
## Panal A
```{r}
AAPosition <- 13
motifPosition <- 'P-1'
bHLH_pbAlignment <- matchAliMotif(mono_motifs, bHLH_pbAlignment)
AAbp_Matrix <- AAbpCombination(mono_motifs, bHLH_pbAlignment, AAPosition, motifPosition)

svd <- FamilyCode::matrixSVD(AAbp_Matrix)
vCoord <- svd$v
meanPointAll <- svd$tetra_mean
connect <- matrix(nrow = 4, ncol = 3, data = 0)
connect[4,] <- meanPointAll
connect[1,] <- connect[4,] + vCoord[,1]
connect[2,] <- connect[4,] + vCoord[,2]
connect[3,] <- connect[4,] + vCoord[,3]


tetrahedron <-  matrix(data = c(1,1,1,1,-1,-1,-1,1,-1,-1,-1,1,1,1,1,-1,1,-1,1,-1,-1,-1,-1,1), nrow = 8, ncol = 3, byrow = TRUE)
tetrahedron <- as.data.frame(tetrahedron)

plotTest <- plot_tetrahedron(AAbp_Matrix, color = T, size = 10)


plotTest <- plotTest%>%
  add_trace(tetrahedron, x = connect[c(1,4),1],
            y=connect[c(1,4),2],z=connect[c(1,4),3],color = I('grey'),
            type = 'scatter3d',
            mode = 'lines',
            line = list(width = 5),
            opacity = 0.5,
            showlegend = F
  )%>%
  add_trace(
    type= "cone",
    x= connect[1,1], y= connect[1,2], z= connect[1,3],
    u= connect[1,1] - connect[4,1], v= connect[1,2] - connect[4,2], w= connect[1,3] - connect[4,3],
    sizeref= 0.2,
    colorscale = list(list(0, as.character('grey')),
                      list(1, as.character('grey'))),
    anchor = 'tip',
    showscale= F
  )%>%
  add_trace(tetrahedron, x = connect[c(2,4),1],
            y=connect[c(2,4),2],z=connect[c(2,4),3],color = I('grey'),
            type = 'scatter3d',
            mode = 'lines',
            line = list(width = 5),
            opacity = 0.5,
            showlegend = F
  )%>%
  add_trace(
    type= "cone",
    x= connect[2,1], y= connect[2,2], z= connect[2,3],
    u= connect[2,1] - connect[4,1], v= connect[2,2] - connect[4,2], w= connect[2,3] - connect[4,3],
    sizeref= 0.2,
    colorscale = list(list(0, as.character('grey')),
                      list(1, as.character('grey'))),
    anchor = 'tip',
    showscale= F
  )%>%
  add_trace(tetrahedron, x = connect[c(3,4),1],
            y=connect[c(3,4),2],z=connect[c(3,4),3],color = I('grey'),
            type = 'scatter3d',
            mode = 'lines',
            line = list(width = 5),
            opacity = 0.5,
            showlegend = F
  )%>%
  add_trace(
    type= "cone",
    x= connect[3,1], y= connect[3,2], z= connect[3,3],
    u= connect[3,1] - connect[4,1], v= connect[3,2] - connect[4,2], w= connect[3,3] - connect[4,3],
    sizeref= 0.2,
    colorscale = list(list(0, as.character('grey')),
                      list(1, as.character('grey'))),
    anchor = 'tip',
    showscale= F
  )%>%
  add_text(x = connect[c(1:3),1], y = connect[c(1:3),2], z = connect[c(1:3),3], text = c('PC1', 'PC2','PC3'),
           showlegend = F)

  axx <- list(
    title = "",
    zeroline = TRUE,
    showline = TRUE,
    showticklabels = TRUE,
    showgrid = TRUE
  )
scene = list(camera = list(eye = list(x = 0, y = 0, z = -2)))

plotTest%>%layout(scene=scene)
```

## Panel B
```{r}
#550/600
pos <- 13
u <- 2
svd <- matrixSVD(gene2pos(mono_motifs, pos = 'P-1'))
x <- svd$u[,u]
y <- substr(bHLH_pbAlignment$alignment, pos,pos)
dt <- data.frame(AA = y, u = x)
dt$col <- AAcolor(dt$AA)
dt <- arrange(dt,u)
colScheme <- unique(dt[,c(1,3)])
plot(x = dt$u, y = 1:52/52, col = dt$col, pch = 19, main = paste0('eCDF for PC',u,' at position ', pos), xlab = paste0('PC',u), ylab = 'Frequency', cex = 1, cex.lab = 1.2, cex.axis = 1.2)
legend(x = "topleft", legend = colScheme$AA, col =colScheme$col, pch = 19, cex = 1, bty = 'n')
```

## Panel C
```{r}
posMatrix <- gene2pos(mono_motifs, pos = 'P-1')
bHLH_pbAlignment <- matchAliMotif(mono_motifs, bHLH_pbAlignment)
svd <- matrixSVD(posMatrix)
#ANOVA test on the level of association between SVD Principal components and residue type
svd.pval <- svdANOVA(svd, bHLH_pbAlignment)
#Plot p-vals 
plot(svd.pval$X1, pch = 1, col = "red", type = 'l', lwd = 3, xlab = 'AA Position', ylab = '-log p-val', main = 'Statistical significance of ANOVA test for PCs at DNA position –1/1')
lines(svd.pval$X2, pch = 1, col = "green", lwd = 3)
lines(svd.pval$X3, pch = 1, col = "blue", lwd = 3)
legend(l, x = "topright", legend = c('PC1', 'PC2', 'PC3'), lwd = 2, col =c('red','green','blue'), lty=1, cex = 0.8, bty = 'n')
```

## Panel D
```{r}
#550/600
BaseLines <- getBaseLineAccuracy(all_mono_motifs, pos = 'P-1', randomSample = F)

plot(BaseLines$HT1,BaseLines$HT2, pch = 19, xlab = 'Experiment 1 ΔΔG', ylab = 'Experiment 2 ΔΔG', main = 'Pair-wise experiment comparison', col = c('green','blue','orange','red'), xlim = c(-3,3), ylim = c(-3,3), cex = 1, cex.lab = 1.2, cex.axis = 1.2)
legend( x = "bottomright", legend = c('A', 'C', 'G', 'T'), col =c('green','blue','orange','red'), pch = 19, bty='n')
legend('topleft', paste0('R^2 = ', round(groupedR2(BaseLines$HT1, BaseLines$HT2, nrow(BaseLines)),4), '\n',
                         'RMSD = ', round(RMSD(BaseLines$HT1,BaseLines$HT2),4)), bty = 'n')
```

## Panel E (iterative prediction)
```{r}
#perform training and testing
predTrue <- SVDregression.Iterative.CV(mono_motifs,bHLH_pbAlignment, pos = 'P-1',Ftest_pVal = 0.05)
#attr(predTrue, 'no.keyPos')
#plot results
plot(predTrue$true,predTrue$pred, pch = 19, xlab = 'Experimental ΔΔG', ylab = 'Predicted ΔΔG', main = 'SVD prediction', col = c('green','blue','orange','red'), xlim = c(-3,3), ylim = c(-3,3), cex = 1, cex.lab = 1.2, cex.axis = 1.2)
legend( x = "bottomright", legend = c('A', 'C', 'G', 'T'), col =c('green','blue','orange','red'), pch = 19, bty='n')
legend('topleft', paste0('R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, nrow(predTrue)),4), '\n',
                         'RMSD = ', round(RMSD(predTrue$pred,predTrue$true),4)), bty = 'n')
```

## Panel F
```{r}
CSpred <- closestSeqPred(mono_motifs, bHLH_pbAlignment, pos = 'P-1')
#plot results
plot(CSpred$true,CSpred$pred, pch = 19, xlab = 'Experimental ΔΔG', ylab = 'Predicted ΔΔG', main = 'Closest Sequence prediction', col = c('green','blue','orange','red'), xlim = c(-3,3), ylim = c(-3,3), cex = 1, cex.lab = 1.2, cex.axis = 1.2)
legend( x = "bottomright", legend = c('A', 'C', 'G', 'T'), col =c('green','blue','orange','red'), pch = 19, bty='n')
legend('topleft', paste0('R^2 = ', round(groupedR2(CSpred$pred, CSpred$true, nrow(CSpred)),4), '\n',
                         'RMSD = ', round(RMSD(CSpred$pred,CSpred$true),4)), bty = 'n')

```

## Panel G (non-Iterative)
```{r}
train_motifs <- mono_motifs
test_motifs <- SELEX_motifs[-c(1,5)]
train_alignment <- matchAliMotif(train_motifs, bHLH_pbAlignment)
test_alignment <- data.frame(name = paste0(groups, '_pb')[-c(1,5)], alignment =c(
                             '-KSLRPLLEKRRRARINQSLSQLKGLI-L------PLLGRENS--NCSKLEKADVL',
                             '-KSLKPLLEKRRVARINQSLSQLKGLI-L------PLLGRENS--NCSKLEKADVL',
                             '-KSLRPLLEKRRVARINQSLSQLKGLI-L------PLLGRENS--NCSKLEKADVL',
                             '-VARKNERERNRVKLVNLGFQALRQHV-P------HGG-------ASKKLSKVETL',
                             '-VARRNERERNRRKLVNLGFQALRQHV-P------HGG-------ASKKLSKVETL',
                             '-VARKNERERNRRKLVNLGFQALRQHV-P------HGG-------ASKKLSKVETL')
                             )
#Train svd-regression model
svd <- matrixSVD(gene2pos(train_motifs, pos = 'P-1'))
svdModel <- trainSVD(svd, train_alignment, no.keyPos = c(2,1,0), keyPos = list(c(13,5), c(13,13), c(13,13)))
#Predict binding motifs for test set
pred_motifs <- predict(svdModel, test_alignment, zero = 0.001)
#Comparing between true and predicted testing set motifs 
true <- frequency2ddG(gene2pos(test_motifs, pos = 'P-1'))
pred <- frequency2ddG(pred_motifs)
predTrue <- data.frame(true = unlist(as.numeric(true)), pred = unlist(as.numeric(pred)))

#plot results
plot(predTrue$true,predTrue$pred, pch = 19, xlab = 'Experimental ΔΔG', ylab = 'Predicted ΔΔG', main = 'SVD prediction', col = c('green','blue','orange','red'), xlim = c(-2,2), ylim = c(-2,2), cex = 1, cex.lab = 1.2, cex.axis = 1.2)
legend( x = "bottomright", legend = c('A', 'C', 'G', 'T'), col =c('green','blue','orange','red'), pch = 19, bty='n')
legend('topleft', paste0('R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, nrow(predTrue)),4), '\n',
                         'RMSD = ', round(RMSD(predTrue$pred,predTrue$true),4)), bty = 'n')
```

## Panel G (Iterative)
```{r}
train_motifs <- mono_motifs
test_motifs <- SELEX_motifs[-c(1,5)]
train_alignment <- matchAliMotif(train_motifs, bHLH_pbAlignment)
test_alignment <- data.frame(name = paste0(groups, '_pb')[-c(1,5)], alignment =c(
                             '-KSLRPLLEKRRRARINQSLSQLKGLI-L------PLLGRENS--NCSKLEKADVL',
                             '-KSLKPLLEKRRVARINQSLSQLKGLI-L------PLLGRENS--NCSKLEKADVL',
                             '-KSLRPLLEKRRVARINQSLSQLKGLI-L------PLLGRENS--NCSKLEKADVL',
                             '-VARKNERERNRVKLVNLGFQALRQHV-P------HGG-------ASKKLSKVETL',
                             '-VARRNERERNRRKLVNLGFQALRQHV-P------HGG-------ASKKLSKVETL',
                             '-VARKNERERNRRKLVNLGFQALRQHV-P------HGG-------ASKKLSKVETL')
                             )
#Train svd-regression model
svd <- matrixSVD(gene2pos(train_motifs, pos = 'P-1'))
svdModel <- trainSVD.Iterative(svd, train_alignment, Ftest_pVal = 0.05)
#Predict binding motifs for test set
pred_motifs <- predict(svdModel, test_alignment, zero = 0.001)
#Comparing between true and predicted testing set motifs 
true <- frequency2ddG(gene2pos(test_motifs, pos = 'P-1'))
pred <- frequency2ddG(pred_motifs)
predTrue <- data.frame(true = unlist(as.numeric(true)), pred = unlist(as.numeric(pred)))

#plot results
plot(predTrue$true,predTrue$pred, pch = 19, xlab = 'Experimental ddG', ylab = 'Predicted ddG', main = 'SVD prediction', col = c('green','blue','orange','red'), xlim = c(-2,2), ylim = c(-2,2))
legend( x = "bottomright", legend = c('A', 'C', 'G', 'T'), col =c('green','blue','orange','red'), pch = 19, bty='n')
legend('topleft', paste0('R^2 = ', round(groupedR2(predTrue$pred, predTrue$true, nrow(predTrue)),4), '\n',
                         'RMSD = ', round(RMSD(predTrue$pred,predTrue$true),4)), bty = 'n')
```

# Figure 5
## Panel B
```{r}
result <- readLines(paste0("HTdata/HES2_Yin2017/result/fit.models.consensus.json"))
mono <- JSON2Matrix(result, mode = 1)
mononucleotide_logo(frequency2ddG(mono))
result <- readLines(paste0("HTdata/ASCL2_Yin2017/result/fit.models.consensus.json"))
mono <- JSON2Matrix(result, mode = 1)
mononucleotide_logo(frequency2ddG(mono))
```

## Panel C
```{r}
result <- readLines(paste0("SELEXdata/HWT_pb/fit.models.consensus.json"))
mono <- JSON2Matrix(result, mode = 1)
mononucleotide_logo(frequency2ddG(mono))
result <- readLines(paste0("SELEXdata/AWT_pb/fit.models.consensus.json"))
mono <- JSON2Matrix(result, mode = 1)
mononucleotide_logo(frequency2ddG(mono))
```

## Panel D
```{r}
for(i in 1:length(SELEX_motifs)){
  mono <- SELEX_motifs[[i]]$matrix
  plot(mononucleotide_logo(frequency2ddG(mono[,c('P-1','P1')])))
}

train_motifs <- mono_motifs
test_motifs <- SELEX_motifs[-c(1,5)]
train_alignment <- matchAliMotif(train_motifs, bHLH_pbAlignment)
test_alignment <- data.frame(name = paste0(groups, '_pb')[-c(1,5)], alignment =c(
                             '-KSLRPLLEKRRRARINQSLSQLKGLI-L------PLLGRENS--NCSKLEKADVL',
                             '-KSLKPLLEKRRVARINQSLSQLKGLI-L------PLLGRENS--NCSKLEKADVL',
                             '-KSLRPLLEKRRVARINQSLSQLKGLI-L------PLLGRENS--NCSKLEKADVL',
                             '-VARKNERERNRVKLVNLGFQALRQHV-P------HGG-------ASKKLSKVETL',
                             '-VARRNERERNRRKLVNLGFQALRQHV-P------HGG-------ASKKLSKVETL',
                             '-VARKNERERNRRKLVNLGFQALRQHV-P------HGG-------ASKKLSKVETL')
                             )
#Train svd-regression model
svd <- matrixSVD(gene2pos(train_motifs, pos = 'P-1'))
svdModel <- trainSVD(svd, train_alignment, no.keyPos = c(2,1,0), keyPos = list(c(13,5), c(13,13), c(13,13)))
#Predict binding motifs for test set
pred_motifs_P.1 <- predict(svdModel, test_alignment, zero = 0.001)
svd <- matrixSVD(gene2pos(train_motifs, pos = 'P1'))
svdModel <- trainSVD(svd, train_alignment, no.keyPos = c(2,1,0), keyPos = list(c(13,5), c(13,13), c(13,13)))
#Predict binding motifs for test set
pred_motifs_P1 <- predict(svdModel, test_alignment, zero = 0.001)

for(i in 1:ncol(pred_motifs_P1)){
  mono <- cbind.data.frame(pred_motifs_P.1[,i], pred_motifs_P1[,i])
  plot(mononucleotide_logo(frequency2ddG(mono)))
}

```

## Panel F
```{r}
library(readxl)
library(ggplot2)
library(ggsignif)
bandIntensity <- read_excel("bandIntensity.xlsx")

proportions <- bandIntensity$...6[c(1:6,13:18)]
proportions <- c(proportions[1:3], exp(1.788), 8.09,
                 proportions[4:6],exp(1.956), 28.41,
                 proportions[7:9], exp(-3.948),0.023,
                 proportions[10:12], exp(-2.15), 0.135)
names <- c(rep('ASCL2_WT',5),rep('ASCL2_R5K',5),rep('ASCL2_V13R',5),rep('ASCL2_Double',5))
names <- as.factor(names)
cols <- c(rep('#6cab42',1),rep('#87eff2',1),rep('#f9d85b',1),rep('#c4560c',1))
dt <- data.frame(name = names, prop = proportions)
dt$prop <- log(dt$prop)
dt$name <- factor(dt$name, levels = c('ASCL2_WT', 'ASCL2_R5K', 'ASCL2_V13R', 'ASCL2_Double'))


p <- ggplot(dt, aes(x = name, y = prop, fill = name, color = name)) +
  geom_dotplot(binaxis = 'y', stackdir = 'center', dotsize = 1) +
  scale_fill_manual(values = cols, name = "Name") +
  scale_color_manual(values = cols, name = "Name") +
  labs(
    x = "",
    y = "-ΔΔG/RT"
  ) +
  theme(legend.position = "none",
    axis.text.x = element_text(face = 'bold',color=cols, size = 12),
    panel.background = element_blank(),     # Remove panel background
    panel.grid.major = element_blank(),     # Remove major gridlines
    panel.grid.minor = element_blank(),     # Remove minor gridlines
    plot.background = element_blank(),
    panel.border = element_rect(colour = "black", fill=NA, size=0.5)# Remove plot background
  ) +
  geom_signif(comparisons = list(c("ASCL2_R5K", "ASCL2_WT"),
                                 c("ASCL2_V13R", "ASCL2_Double"),
                                 c("ASCL2_R5K", "ASCL2_V13R"),
                                 c("ASCL2_WT", "ASCL2_V13R"),
                                 c("ASCL2_WT", "ASCL2_Double"),
                                 c("ASCL2_R5K", "ASCL2_Double")
                                 ),
              map_signif_level=TRUE, color = 'black',  step_increase = 0.1)
plot(p)
```

## Compair R2 between models
```{r}
R2listPred <- c()
RMSDlistPred <- c()
R2listClose <- c()
RMSDlistClose <- c()
for(i in 1:52){
  predTrue <- SVDregression.Iterative.CV(mono_motifs[-i],bHLH_pbAlignment[-i,], pos = 'P-1',Ftest_pVal = 0.05)
  R2listPred <- c(R2listPred, groupedR2(predTrue$pred, predTrue$true, nrow(predTrue)))
  RMSDlistPred <- c(RMSDlistPred, RMSD(predTrue$pred,predTrue$true))
  CSpred <- closestSeqPred(mono_motifs[-i], bHLH_pbAlignment[-i,], pos = 'P-1')
  R2listClose <- c(R2listClose, groupedR2(CSpred$pred, CSpred$true, nrow(CSpred)))
  RMSDlistClose <- c(RMSDlistClose, RMSD(CSpred$pred,CSpred$true))
  print(i)
}

t <- t.test(R2listPred, R2listClose)
t$p.value
t.test(RMSDlistPred, RMSDlistClose)
```

# Supplemental figure 1
```{r}
mkdirs('S1')
for(i in 1:nrow(bHLH_index)){
  result <- readLines(paste0("HTdata/",
                                     bHLH_index$gene_symbol[i], '_', bHLH_index$study[i],'/result/fit.models.consensus.json'))
  mono <- JSON2Matrix(result, mode = bHLH_index$motif_model[i])
  plot <- mononucleotide_logo(mono)
  plot <- plot + ggplot2::ggtitle(paste0(gsub('_','',bHLH_index$gene_symbol[i]), '\nR^2: ', bHLH_index$quality[i]))
  png(paste0('S1/',gsub('_','',bHLH_index$gene_symbol[i]),'.png'))
  plot(plot)
  dev.off()
}

```

# Supplemental figure 2
```{r}
library(plotly)
mkdirs('S2')
pos_index <- c('P-3','P-2','P-1','P1','P2','P3')
for(i in 1:6){
  for(j in c(5,13)){
    AAPosition <- j
    motifPosition <- pos_index[i]
    bHLH_pbAlignment <- matchAliMotif(mono_motifs, bHLH_pbAlignment)
    AAbp_Matrix <- AAbpCombination(mono_motifs, bHLH_pbAlignment, AAPosition, motifPosition)
    plot <- plot_tetrahedron(AAbp_Matrix, color = T, vertex = F, size = 10)
    scene = list(camera = list(eye = list(x = 1.25, y = 1.25, z = 1.25)))
    plot <- plot%>%layout(scene=scene)
    htmlwidgets::saveWidget(as_widget(plot), paste0("S2/AA",j,"_",motifPosition,".html"))
  }
}
```

# Supplemtnal figure 3
```{r}
mkdirs('S3')
posMatrix <- gene2pos(mono_motifs, pos = 'P-1')
bHLH_pbAlignment <- matchAliMotif(mono_motifs, bHLH_pbAlignment)
svd <- matrixSVD(posMatrix)
#ANOVA test on the level of association between SVD Principal components and residue type
svd.pval <- svdANOVA(svd, bHLH_pbAlignment)
for(j in 1:56){
  for(u in 1:3){
    pos <- j
    svd <- matrixSVD(gene2pos(mono_motifs, pos = 'P-1'))
    x <- svd$u[,u]
    y <- substr(bHLH_pbAlignment$alignment, pos,pos)
    dt <- data.frame(AA = y, u = x)
    dt$col <- AAcolor(dt$AA)
    dt <- arrange(dt,u)
    colScheme <- unique(dt[,c(1,3)])
    png(paste0('S3/AA',pos,'_PC',u,'.png'))
    plot(x = dt$u, y = 1:52/52, col = dt$col, pch = 19, main = paste0('eCDF for PC',u,' at position ', pos), xlab = paste0('PC',u), ylab = 'Frequency', cex = 1, cex.lab = 1.2, cex.axis = 1.2)
    legend(x = "topleft", legend = colScheme$AA, col =colScheme$col, pch = 19, cex = 1, bty = 'n')
    legend(x = "bottomright", legend = paste0('-log(p-Value): ',round(svd.pval[pos,u],2)), bty = 'n')
    dev.off()
  }
}

```

# Supplemtnal figure 4
```{r}
#1400/400
pos_index <- c('P1','P2','P3')
for(i in 1:3){
  posMatrix <- gene2pos(mono_motifs, pos = pos_index[i])
  bHLH_pbAlignment <- matchAliMotif(mono_motifs, bHLH_pbAlignment)
  svd <- matrixSVD(posMatrix)
  #ANOVA test on the level of association between SVD Principal components and residue type
  svd.pval <- svdANOVA(svd, bHLH_pbAlignment)
  #Plot p-vals 
  plot(svd.pval$X1, pch = 1, col = "red", type = 'l', lwd = 3, xlab = 'AA Position', ylab = '-log p-val', ylim = c(0, (1.1*max(svd.pval[,1:3]))) ,main = paste0('Statistical significance of ANOVA test for PCs at DNA position -', i, '/+',i))
  lines(svd.pval$X2, pch = 1, col = "green", lwd = 3)
  lines(svd.pval$X3, pch = 1, col = "blue", lwd = 3)
  #legend(l, x = "topright", legend = c('PC1', 'PC2', 'PC3'), lwd = 2, col =c('red','green','blue'), lty=1, cex = 0.8, bty = 'n')
}

```


